{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# Double drive simulator with evolutionary stability"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "23-02-21\n",
    "\n",
    "Katie Willis\n",
    "\n",
    "katie.willis16@imperial.ac.uk\n",
    "___\n",
    "\n",
    "This file contains functions required to simulate the release of a 2 locus gene drive into a single panmictic population and monitor the population through time.\n",
    "We model 2 loci allowing for homing-associated loss-of-function of each transcriptional unit present in each genetic construct.\n",
    "The number of alleles and genotypes will therefore depend on the contents of the constructs being released. \n",
    "We begin with a matrix containing numbers of individuals of each genotype after transgenic release, monitoring males and females separately. \n",
    "To model processes such as fitness costs, homing, sex ratio bias and recombination during each generation, we multiply (element-wise) the numbers of individuals per genotype by a series of transition matrices for each process, built using user-defined parameters. \n",
    "The output is a dictionary containing time-series of allele and genotype frequencies, numbers of indididuals and the correlation between the two gene drive constructs.\n",
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "## Parameter and matrix definition"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true,
    "hidden": true
   },
   "source": [
    "### Allele and genotype lists"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "Since the number of alleles at each locus depends on the number of components in the released construct, we first define which components are present in each construct and then calculate the number of alleles/genotypes for each design (in contrast to the [baseline](DoubleDrive_Simulator_Baseline.ipynb) and [4-allele](DoubleDrive_Simulator_4allele.ipynb) models where the number of alleles and genotypes are constant).  \n",
    "Here the alleles are labelled $A_1...A_n$ at the first locus, and $B_1...B_m$ at the second locus, where $n$ and $m$ are the number of alleles at each loci.\n",
    "The functions below calculate the number of alleles at each loci and generate lists of alleles and genotypes according to the contents of the constructs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "code_folding": [
     3,
     61
    ],
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "define_string_genotypes"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Defines lists of alleles and genotypes as strings.\n",
    "\"\"\"\n",
    "function define_string_genotypes(n_alleles_locus1, #number of A alleles\n",
    "                                n_alleles_locus2)  #number of B alleles\n",
    "    \n",
    "    A_alleles = [\"A\"*string(i) for i in 1:n_alleles_locus1]\n",
    "    B_alleles = [\"B\"*string(i) for i in 1:n_alleles_locus2]\n",
    "    \n",
    "    #Define all A and B allele combinations: Creating single chromosomes\n",
    "    alleles = []\n",
    "    for A in A_alleles\n",
    "        for B in B_alleles\n",
    "            push!(alleles,([A,B]))\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    #Define all genotypes at locus A\n",
    "    A_genotypes = []\n",
    "    for i in 1:length(A_alleles)\n",
    "        allele1 = A_alleles[i]\n",
    "        for j in i:length(A_alleles)\n",
    "            allele2 = A_alleles[j]\n",
    "            push!(A_genotypes,([allele1,allele2]))\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    #Define all genotypes at locus B\n",
    "    B_genotypes = []\n",
    "    for i in 1:length(B_alleles)\n",
    "        allele1 = B_alleles[i]\n",
    "        for j in i:length(B_alleles)\n",
    "            allele2 = B_alleles[j]\n",
    "            push!(B_genotypes,([allele1,allele2]))\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    #Define all genotypes for both A and B loci combined\n",
    "    genotypes = []\n",
    "    for i in 1:length(alleles)\n",
    "        allele1 = alleles[i]\n",
    "        for j in i:length(alleles)\n",
    "            allele2 = alleles[j]\n",
    "            push!(genotypes,([allele1 allele2]))\n",
    "        end\n",
    "    end\n",
    "    #First index selects the genotype \n",
    "    #Second index selects the locus of interest \n",
    "    #third indec selects the chromosome of interest\n",
    "    \n",
    "    return(A_alleles,\n",
    "            B_alleles,\n",
    "            alleles,\n",
    "            A_genotypes,\n",
    "            B_genotypes,\n",
    "            genotypes)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Defines lists of alleles and genotypes as symbols.\n",
    "\"\"\"\n",
    "function define_symbolic_genotypes(n_alleles_locus1, #number of A alleles\n",
    "                                    n_alleles_locus2)#number of B alleles\n",
    "    A_alleles_sym = [\"A\"*string(i) for i in 1:n_alleles_locus1]\n",
    "    B_alleles_sym = [\"B\"*string(i) for i in 1:n_alleles_locus2];\n",
    "    \n",
    "    #Define all A and B allele combinations: Creating single chromosomes\n",
    "    alleles_sym = []\n",
    "    for A in A_alleles_sym\n",
    "        for B in B_alleles_sym\n",
    "            push!(alleles_sym,(A*B))\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    #Define all genotypes at locus A\n",
    "    A_genotypes_sym = []\n",
    "    for i in 1:length(A_alleles_sym)\n",
    "        allele1 = A_alleles_sym[i]\n",
    "        for j in i:length(A_alleles_sym)\n",
    "            allele2 = A_alleles_sym[j]\n",
    "            push!(A_genotypes_sym,(allele1*allele2))\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    #Define all genotypes at locus B\n",
    "    B_genotypes_sym = []\n",
    "    for i in 1:length(B_alleles_sym)\n",
    "        allele1 = B_alleles_sym[i]\n",
    "        for j in i:length(B_alleles_sym)\n",
    "            allele2 = B_alleles_sym[j]\n",
    "            push!(B_genotypes_sym,(allele1*allele2))\n",
    "        end\n",
    "    end\n",
    "\n",
    "    \n",
    "    #Define all genotypes for both A and B loci combined\n",
    "    genotypes_sym = []\n",
    "    for i in 1:length(alleles_sym)\n",
    "        allele1 = alleles_sym[i]\n",
    "        for j in i:length(alleles_sym)\n",
    "            allele2 = alleles_sym[j]\n",
    "            push!(genotypes_sym,(allele1*allele2))\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    A_alleles_sym = [Sym(i) for i in A_alleles_sym]\n",
    "    B_alleles_sym = [Sym(i) for i in B_alleles_sym]\n",
    "    alleles_sym = [Sym(a) for a in alleles_sym];\n",
    "    A_genotypes_sym = [Sym(a) for a in A_genotypes_sym]\n",
    "    B_genotypes_sym = [Sym(b) for b in B_genotypes_sym]\n",
    "    genotypes_sym = [Sym(i) for i in genotypes_sym];\n",
    "\n",
    "    return(alleles_sym,\n",
    "            genotypes_sym)\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "At a single locus, each allele has associated with it a number of components, or it is empty.\n",
    "The alleles and their contents are defined using a dictionary.\n",
    "Each dictionary key is an allele e.g. $A_1...A_n$, each value is a list of transcriptional units $∈$ {\"cas9\",\"grna\",\"xs\"}.\n",
    "\n",
    "At a single locus, each allele also has associated with it a cost due to disruption by insertion of a construct or by a cleavage-resistant allele. Each allele is assigned a fitness type: wt = no costs; disruption = costs due to insertion; resistant = costs due to resistant mutation.\n",
    "The alleles and their disruption costs are defined using a dictionary.\n",
    "Each dictionary key is an allele e.g. $A_1...A_n$, each value is a string $∈$ {\"wt\",\"disruption\",\"resistant\"}.\n",
    "\n",
    "The functions below generate the two dictionaries for each locus depending on the contents of the released constructs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "code_folding": [
     3,
     29
    ],
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Make_construct_content_dict"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Generates a list of the content of all possible mutant constructs\n",
    "\"\"\"\n",
    "function Make_construct_content_list(construct,   #List of components in construct\n",
    "                                    content_list, #List of lists, each containing the contents of eahc allele\n",
    "                                    L)             #number of components in the construct)\n",
    "    for i in 1:length(construct)\n",
    "        #removes a component of the construct\n",
    "        construct_mod = deepcopy(construct)\n",
    "        deleteat!(construct_mod,i)\n",
    "        if length(construct_mod) == 0\n",
    "            construct_mod = []\n",
    "        end\n",
    "        if construct_mod ∉ content_list\n",
    "            push!(content_list,construct_mod)\n",
    "        end\n",
    "        \n",
    "        #Repeat if there are still components in mutant construct \n",
    "        if length(construct_mod)<=(L-1)\n",
    "            content_list = Make_construct_content_list(construct_mod,content_list,L)\n",
    "        end\n",
    "    end\n",
    "    return(content_list)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Generates a dictionary of all alleles, including the WT, construct, all mutant forms\n",
    "and additional resistant allele if required.\n",
    "\"\"\"\n",
    "function Make_construct_content_dict(locus,          #A or B\n",
    "                                    construct,       #List of components in construct e.g [\"cas9\",\"Agrna\"]\n",
    "                                    resistant_locus) #true or false for resistant locus with different fitness\n",
    "    content_list = []\n",
    "    L = length(construct)\n",
    "    content_list = Make_construct_content_list(construct,content_list,L)\n",
    "    content_list = sort(content_list,by=length,rev=true)\n",
    "    \n",
    "    content_dict = Dict()\n",
    "    content_dict[locus*string(1)] = []\n",
    "    content_dict[locus*string(2)] = construct\n",
    "    \n",
    "    ir_dict = Dict()\n",
    "    ir_dict[locus*string(1)] = \"wt\"\n",
    "    ir_dict[locus*string(2)] = \"disruption\"\n",
    "    \n",
    "    count = 2\n",
    "    for i in 1:length(content_list)\n",
    "        count = count +1\n",
    "        content_dict[locus*string(count)] = content_list[i]\n",
    "        ir_dict[locus*string(count)] = \"disruption\"\n",
    "    end\n",
    "    if resistant_locus == true\n",
    "        count = count +1\n",
    "        content_dict[locus*string(count)] = []\n",
    "        ir_dict[locus*string(count)] = \"resistant\"\n",
    "    end\n",
    "    return(content_dict,ir_dict)\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true,
    "hidden": true
   },
   "source": [
    "### Parameters\n",
    "\n",
    "We define a dictionary of all parameters required to run the simulation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "code_folding": [
     12
    ],
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Initiate_baseline_parameters (generic function with 1 method)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Initiates a dictionary containing all parameters. \n",
    "All are set to zero except:\n",
    "m = 0.5\n",
    "r = 0.5\n",
    "Rm = 6 \n",
    "Nf_eq = 1000 (Number of females or males at equlibrium)\n",
    "OJ = 1/10\n",
    "therefore:\n",
    "f = (2*Rm)/OJ_\n",
    "α = f_*Nf_eq/(Rm-1)\n",
    "\"\"\"\n",
    "function Initiate_baseline_parameters()\n",
    "    \n",
    "    params = Dict{String,Float64}()\n",
    "\n",
    "    #expression costs for each possible construct component\n",
    "    ## for females\n",
    "    params[\"scasf\"] = 0\n",
    "    params[\"hcasf\"] = 0\n",
    "    params[\"sAgRNAf\"] = 0\n",
    "    params[\"hAgRNAf\"] = 0\n",
    "    params[\"sBgRNAf\"] = 0\n",
    "    params[\"hBgRNAf\"] = 0\n",
    "    params[\"sXSf\"] = 0\n",
    "    params[\"hXSf\"] = 0\n",
    "    ## for males\n",
    "    params[\"scasm\"] = 0\n",
    "    params[\"hcasm\"] = 0\n",
    "    params[\"sAgRNAm\"] = 0\n",
    "    params[\"hAgRNAm\"] = 0\n",
    "    params[\"sBgRNAm\"] = 0\n",
    "    params[\"hBgRNAm\"] = 0\n",
    "    params[\"sXSm\"] = 0\n",
    "    params[\"hXSm\"] = 0\n",
    "    \n",
    "    #disruption costs\n",
    "    ## for females\n",
    "    params[\"siaf\"] = 0\n",
    "    params[\"hiaf\"] = 0\n",
    "    params[\"sraf\"] = 0\n",
    "    params[\"hraf\"] = 0\n",
    "    params[\"sibf\"] = 0\n",
    "    params[\"hibf\"] = 0\n",
    "    params[\"srbf\"] = 0\n",
    "    params[\"hrbf\"] = 0\n",
    "    params[\"siraf\"] = 0\n",
    "    params[\"sirbf\"] = 0\n",
    "    ## for males\n",
    "    params[\"siam\"] = 0\n",
    "    params[\"hiam\"] = 0\n",
    "    params[\"sram\"] = 0\n",
    "    params[\"hram\"] = 0\n",
    "    params[\"sibm\"] = 0\n",
    "    params[\"hibm\"] = 0\n",
    "    params[\"srbm\"] = 0\n",
    "    params[\"hrbm\"] = 0\n",
    "    params[\"siram\"] = 0\n",
    "    params[\"sirbm\"] = 0\n",
    "    \n",
    "    ## Activity costs\n",
    "    params[\"sXsf\"] = 0\n",
    "    params[\"sXsm\"] = 0\n",
    "    params[\"sSaf\"] = 0\n",
    "    params[\"sSam\"] = 0\n",
    "    params[\"sSbf\"] = 0\n",
    "    params[\"sSbm\"] = 0\n",
    "    params[\"sHaf\"] = 0\n",
    "    params[\"sHam\"] = 0\n",
    "    params[\"sHbf\"] = 0\n",
    "    params[\"sHbm\"] = 0\n",
    "    \n",
    "    # homing rates\n",
    "    params[\"caf\"] = 0\n",
    "    params[\"jaf\"] = 0\n",
    "    params[\"cbf\"] = 0\n",
    "    params[\"jbf\"] = 0\n",
    "    params[\"cam\"] = 0\n",
    "    params[\"jam\"] = 0\n",
    "    params[\"cbm\"] = 0\n",
    "    params[\"jbm\"] = 0\n",
    "    \n",
    "    #recombination\n",
    "    params[\"r\"] = 0.5\n",
    "    \n",
    "    #sex ratio bias\n",
    "    params[\"m\"] = 0.5\n",
    "    \n",
    "    #population dynamics \n",
    "    Rm = 6\n",
    "    Nf_eq = 1000.0\n",
    "    params[\"OJ\"] = 1/10\n",
    "    params[\"f\"] = (2*Rm)/params[\"OJ\"]\n",
    "    params[\"α\"] = params[\"f\"]*Nf_eq/(Rm-1)\n",
    "    \n",
    "    #probability of NHEJ products being having the 'resistant' fitness type\n",
    "    params[\"paf\"] = 0\n",
    "    params[\"pam\"] = 0\n",
    "    params[\"pbf\"] = 0\n",
    "    params[\"pbm\"] = 0\n",
    "    \n",
    "    #probability of homing associated mutation of each trancriptional unit\n",
    "    params[\"daf\"] = 0\n",
    "    params[\"dam\"] = 0\n",
    "    params[\"dbf\"] = 0\n",
    "    params[\"dbm\"] = 0\n",
    "    \n",
    "    return(params)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Initiates a parameter set based on our selected baseline parameters and:\n",
    "    - the locus type of A and B (Ntrl, HI, HS)\n",
    "    - homing dependency of A and B (Constitutive, conditional, none)\n",
    "    - XS dependency (A or B)\n",
    "If applied, homing rates are pre-defined according to Kyrou et al 2018\n",
    "c = 0.971\n",
    "j = 0.019\n",
    "if applied, Xshredder rates are pre-defined according to Galizi et al 2014\n",
    "m = 0.95\n",
    "Fitness costs due to locus distruption are pre-defined to 1 based on the locus type\n",
    "Off target cleavage costs are set to 0.01 for both males and females\n",
    "\"\"\"\n",
    "function Assign_baseline_params(LOCUS_A,\n",
    "                                LOCUS_B,\n",
    "                                set_A_homing_dependency,\n",
    "                                set_B_homing_dependency,\n",
    "                                set_XS_dependency)\n",
    "    \n",
    "    cleavage_Kyrou2018 = 0.971\n",
    "    joining_Kyrou2018 = 0.019\n",
    "    \n",
    "    params = Initiate_baseline_parameters()\n",
    "    \n",
    "    if LOCUS_A == \"HS_rec\"\n",
    "        #HS gene\n",
    "        params[\"siaf\"]=1.0 \n",
    "        params[\"sraf\"]=1.0 \n",
    "        params[\"siraf\"] = 1.0\n",
    "    elseif LOCUS_A == \"HI\"\n",
    "        #HI gene\n",
    "        params[\"siaf\"]=1.0 \n",
    "        params[\"sraf\"]=1.0 \n",
    "        params[\"siraf\"] = 1.0\n",
    "        params[\"hiaf\"]=1.0 \n",
    "        params[\"hraf\"]=1.0\n",
    "    end\n",
    "\n",
    "    if LOCUS_B == \"HS_rec\"\n",
    "        #HS gene\n",
    "        params[\"sibf\"]=1.0 \n",
    "        params[\"srbf\"]=1.0 \n",
    "        params[\"sirbf\"] = 1.0\n",
    "    elseif LOCUS_B == \"HI\"\n",
    "        #HI gene\n",
    "        params[\"sibf\"]=1.0 \n",
    "        params[\"srbf\"]=1.0 \n",
    "        params[\"sirbf\"] = 1.0\n",
    "        params[\"hibf\"]=1.0 \n",
    "        params[\"hrbf\"]=1.0\n",
    "    end\n",
    "\n",
    "    if set_A_homing_dependency != \"none\"\n",
    "        params[\"caf\"] = cleavage_Kyrou2018 # homing \n",
    "        params[\"jaf\"] = joining_Kyrou2018 # homing\n",
    "        params[\"cam\"] = cleavage_Kyrou2018 # homing \n",
    "        params[\"jam\"] = joining_Kyrou2018 # homing\n",
    "        A_homing_dependency = set_A_homing_dependency\n",
    "    elseif set_A_homing_dependency == \"none\"\n",
    "        #Doesnt matter what this is as the homing rates are set to 0.0\n",
    "        A_homing_dependency = \"constitutive\"\n",
    "    end\n",
    "\n",
    "    if set_B_homing_dependency != \"none\"\n",
    "        params[\"cbf\"] = cleavage_Kyrou2018 # homing \n",
    "        params[\"jbf\"] = joining_Kyrou2018 # homing\n",
    "        params[\"cbm\"] = cleavage_Kyrou2018 # homing \n",
    "        params[\"jbm\"] = joining_Kyrou2018 # homing\n",
    "        B_homing_dependency = set_B_homing_dependency\n",
    "    elseif set_B_homing_dependency == \"none\"\n",
    "        #Doesnt matter what this is as the homing rates are set to 0.0\n",
    "        B_homing_dependency = \"constitutive\"\n",
    "    end\n",
    "\n",
    "    if set_XS_dependency != \"none\"\n",
    "        params[\"m\"] = 0.95\n",
    "        XS_dependency = set_XS_dependency\n",
    "    elseif set_XS_dependency == \"none\"\n",
    "        params[\"m\"] = 0.5\n",
    "        XS_dependency = \"none\"\n",
    "    end\n",
    "    \n",
    "    #off target cleavage\n",
    "    params[\"sHaf\"] = 0.01\n",
    "    params[\"sHam\"] = 0.01\n",
    "    params[\"sHbf\"] = 0.01\n",
    "    params[\"sHbm\"] = 0.01\n",
    "    \n",
    "    return(params)\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "### Fitness\n",
    "\n",
    "Fitness costs are calculated based on 3 different processes. The first two are locus specific, including costs due to locus disruption by either construct insertion or cleavage resistant mutation, and costs due to expression of the construct. The third is due to construct activity, which includes off-target Cas9 cleavage, somatic Cas9 cleavage and X-shredding, each of which may or may not require interaction between constructs for the cost to occur. \n",
    "Fitness costs are caculated multiplicitly: where the homozygous WT fitness cost is standardised to 1 and each process reduces the fitness by a factor.\n",
    "The functions below produce a fitness matrix containing the relative fitness of each genotype compared to the wildtype depending on the contents of each construct. \n",
    "Each row of the matrix represents a genotype, the first column contains the relative fitness of females and the second the relative fitness of males for each genotype.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "\n",
    "##### Disruption costs\n",
    "\n",
    "Each allele has associated with it costs due to disruption of the locus.\n",
    "The fitness due to disruption of function is as follows: \n",
    "\n",
    "\\begin{align}\n",
    "{W/W} & = 1 \\\\\n",
    "{W/D} & = 1-s_Ih_I \\\\\n",
    "{D/D} & = 1-s_I \\\\\n",
    "{W/R} & = 1-s_Rh_R \\\\\n",
    "{R/R} & = 1-s_R \\\\\n",
    "{D/R} & = 1-s_{IR} \\\\\n",
    "\\end{align}\n",
    "\n",
    "Where $s_I$ and $h_I$ are the selection and domcinance coefficient for disruption by insertion of the construct or any of its mutant forms (D), $s_R$ and $h_R$ are the selection and dominance coefficient for disruption by a resistance allele (R) and $s_{IR}$ is the selection coefficient for the heterozygotes carrying either a construct or a mutant construct and a cleavage resistant allele.\n",
    "Each parameter differs depending on locus and sex. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "code_folding": [
     6
    ],
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Get_disruption_fitness"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Makes the fitness matrix of female and male individuals taking into consideration disruption of \n",
    "function at a single locus.\n",
    "The output is a matrix, where each column represents the sex (female, male)\n",
    "and each row represents the genotype of a diploid individuals at a single locus.\n",
    "\"\"\"\n",
    "function Get_disruption_fitness(alleles_ir,locus_genotypes,\n",
    "                                sif::Float64,hif::Float64,srf::Float64,hrf::Float64,sirf::Float64,\n",
    "                                sim::Float64,him::Float64,srm::Float64,hrm::Float64,sirm::Float64)\n",
    "    \n",
    "    #Define dictionaries for disruption/insertion\n",
    "    component_ir_dict_f = Dict([(\"wt\", [0,0]), (\"disruption\", [sif, hif]), (\"resistant\", [srf, hrf])]);\n",
    "    component_ir_dict_m = Dict([(\"wt\", [0,0]), (\"disruption\", [sim, him]), (\"resistant\", [srm, hrm])]);\n",
    "    \n",
    "    #storage\n",
    "    fitness_f_vec = []\n",
    "    fitness_m_vec = []\n",
    "    \n",
    "    for genotype in locus_genotypes\n",
    "\n",
    "        #Collect info on alleles\n",
    "        ir_type_1 = alleles_ir[genotype[1]]\n",
    "        ir_type_2 = alleles_ir[genotype[2]]\n",
    "        #Assign fitness due to interuption of the gene\n",
    "        ##fitness costs of 2 components with the same insertion effect\n",
    "        if ir_type_1 == ir_type_2\n",
    "            fitness_f = (1-component_ir_dict_f[ir_type_1][1])\n",
    "            fitness_m = (1-component_ir_dict_m[ir_type_1][1])\n",
    "        ##fitness costs of 2 components with different insertion effects\n",
    "        ###if one is the WT we use heterozygote parameters\n",
    "        elseif (ir_type_1 == \"wt\") & (ir_type_2 != \"wt\")\n",
    "            fitness_f = (1-(component_ir_dict_f[ir_type_2][1]*component_ir_dict_f[ir_type_2][2]))\n",
    "            fitness_m = (1-(component_ir_dict_m[ir_type_2][1]*component_ir_dict_m[ir_type_2][2]))\n",
    "            #println(genotype,fitness_f,fitness_m)\n",
    "        elseif (ir_type_1 != \"wt\") & (ir_type_2 == \"wt\")\n",
    "            fitness_f = (1-(component_ir_dict_f[ir_type_1][1]*component_ir_dict_f[ir_type_1][2]))\n",
    "            fitness_m = (1-(component_ir_dict_m[ir_type_1][1]*component_ir_dict_m[ir_type_1][2]))\n",
    "            #println(genotype,fitness_f,fitness_m)\n",
    "        elseif (ir_type_1 != \"wt\") & (ir_type_2 != \"wt\")\n",
    "            fitness_f = (1-sirf)\n",
    "            fitness_m = (1-sirm)\n",
    "        else \n",
    "            print(\"missed an option\")\n",
    "        end\n",
    "        push!(fitness_f_vec,fitness_f)\n",
    "        push!(fitness_m_vec,fitness_m)\n",
    "    end\n",
    "    return(hcat(fitness_f_vec,fitness_m_vec))\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "\n",
    "##### Expression costs\n",
    "\n",
    "Each allele may or may not contain transcriptional units, each of which has expression costs associated. \n",
    "The fitness due to expression of these components is as follows: \n",
    "\n",
    "\\begin{align}\n",
    "{Hom} & = \\prod_{k=1}^n (1-s_k)\\\\\n",
    "{Het} & = \\prod_{k=1}^n (1-s_kh_k) + \\prod_{l=1}^n (1-s_lh_l) \\\\\n",
    "\\end{align}\n",
    "\n",
    "where k refers to fitness effects associated with each component present in the allele, if heterozygous k and l refer to fitness effects associated with each component present in the first and second allele reprectively.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [
     5
    ],
    "hidden": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Makes the fitness matrix of female and male individuals considering the contents of the construct at a single locus..\n",
    "The output is a matrix, where each column rperesents the sex (female, male)\n",
    "and each row represents the genotype of a diploid individuals at a single locus.\n",
    "\"\"\"\n",
    "function Get_construct_expression_fitness(alleles_content,locus_genotypes,p)\n",
    "\n",
    "    component_fitness_dict = Dict([(\"cas9\", [p[\"scasf\"],p[\"hcasf\"],p[\"scasm\"],p[\"hcasm\"]]), \n",
    "            (\"Agrna\", [p[\"sAgRNAf\"],p[\"hAgRNAf\"],p[\"sAgRNAm\"],p[\"hAgRNAm\"]]), \n",
    "            (\"Bgrna\", [p[\"sBgRNAf\"],p[\"hBgRNAf\"],p[\"sBgRNAm\"],p[\"hBgRNAm\"]]), \n",
    "            (\"xs\", [p[\"sXSf\"],p[\"hXSf\"],p[\"sXSm\"],p[\"hXSm\"]])])\n",
    "    \n",
    "    fitness_f_vec = Vector{Float64}()\n",
    "    fitness_m_vec = Vector{Float64}()\n",
    "    \n",
    "    for genotype in locus_genotypes\n",
    "    \n",
    "        #full fitness to begin with\n",
    "        fitness_f = 1\n",
    "        fitness_m = 1\n",
    "        \n",
    "        #Collect info on alleles\n",
    "        content_1 = alleles_content[genotype[1]]\n",
    "        content_2 = alleles_content[genotype[2]]\n",
    "        \n",
    "        #Assign fitness due to cost of expressing content of construct\n",
    "        ##Homozygote fitness costs of construct expression \n",
    "        if genotype[1] == genotype[2]\n",
    "            \n",
    "            #fitness due to expression of component contents\n",
    "            for component in content_1\n",
    "                fitness_f = fitness_f*(1-component_fitness_dict[component][1])\n",
    "                fitness_m = fitness_m*(1-component_fitness_dict[component][3])\n",
    "            end\n",
    "\n",
    "        ##Heterozygous fitness costs of construct expression \n",
    "        elseif genotype[1] != genotype[2]\n",
    "            for component in content_1\n",
    "                fitness_f = fitness_f*(1-(component_fitness_dict[component][1]*component_fitness_dict[component][2]))\n",
    "                fitness_m = fitness_m*(1-(component_fitness_dict[component][3]*component_fitness_dict[component][4]))\n",
    "            end\n",
    "            for component in content_2\n",
    "                fitness_f = fitness_f*(1-(component_fitness_dict[component][1]*component_fitness_dict[component][2]))\n",
    "                fitness_m = fitness_m*(1-(component_fitness_dict[component][3]*component_fitness_dict[component][4]))\n",
    "            end\n",
    "        end\n",
    "        push!(fitness_f_vec,fitness_f)\n",
    "        push!(fitness_m_vec,fitness_m)\n",
    "    end\n",
    "    return(hcat(fitness_f_vec,fitness_m_vec))\n",
    "end\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "The fitness of an individual based on the diploid genotype at a single allele is calculated by multplying the fitness due to expression of the components by the fitness due to disruption of function at the locus.\n",
    "The fitness for an individual considering expression and locus disruption at both loci is calculated by multiplying the fitness costs based on each single locus."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [
     5
    ],
    "hidden": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Combines the A and B 1-locus fitness matrices to generate a matrix where \n",
    "rows represent each 2-locus diploid genotype, for female (column 1) and male (column 2) fitness costs for locus \n",
    "disruption and construct expression. \n",
    "\"\"\"\n",
    "function Get_expression_fitness_matrix(A_fitness_matrix,B_fitness_matrix,A_genotypes,B_genotypes,genotypes)\n",
    "    \n",
    "    #Calculate the fitness of each combined genotype by multiplying the\n",
    "    # fitness for genotypes at each locus.\n",
    "\n",
    "    fitness_M_store = [] #male fitness\n",
    "    fitness_F_store = [] #female fitness\n",
    "\n",
    "    #For every combined genotype e.g. ABAB\n",
    "    for genotype in genotypes\n",
    "        \n",
    "        B_fitness = []\n",
    "        A_fitness = []\n",
    "        \n",
    "        #Check what the B genotype is e.g. BB\n",
    "        #and assign B_fitness accordingly:\n",
    "        for i in 1:length(B_genotypes) #iterate through the B genotypes\n",
    "            B_genotype = B_genotypes[i] #Split to get the two alleles\n",
    "            if ( #Check if both alleles occur in the combined genotype (in any order)\n",
    "                    ((B_genotype[1]==genotype[2,1]) & (B_genotype[2]==genotype[2,2]))\n",
    "                    |\n",
    "                    ((B_genotype[2]==genotype[2,1]) & (B_genotype[1]==genotype[2,2]))\n",
    "                )\n",
    "                #If they do assign the fitness\n",
    "                B_fitness = B_fitness_matrix[i,:]\n",
    "            end\n",
    "        end\n",
    "\n",
    "        #Check what the A genotype is e.g. AA\n",
    "        #and assign A_fitness accordingly:\n",
    "        for i in 1:length(A_genotypes) \n",
    "            A_genotype = A_genotypes[i] #iterate through the A genotypes\n",
    "            if ( #Check if both alleles occur in the combined genotype (in any order)\n",
    "                    ((A_genotype[1]==genotype[1,1]) & (A_genotype[2]==genotype[1,2]))\n",
    "                    |\n",
    "                    ((A_genotype[2]==genotype[1,1]) & (A_genotype[1]==genotype[1,2]))\n",
    "                )\n",
    "                #If they do assign the fitness\n",
    "                A_fitness = A_fitness_matrix[i,:]\n",
    "            end\n",
    "        end\n",
    "\n",
    "        #Multiply the fitnesses at A and B (Assume no epistatic effects)\n",
    "        combined_fitness = simplify.(A_fitness .* B_fitness)\n",
    "        fitness_F = combined_fitness[1]\n",
    "        fitness_M = combined_fitness[2]\n",
    "        #Store male and female fitnesses independently\n",
    "        push!(fitness_M_store,fitness_M)   \n",
    "        push!(fitness_F_store,fitness_F)   \n",
    "    end\n",
    "    \n",
    "    #Combine male and female fitnesses for each combined genotype into a matrix\n",
    "    return(hcat(fitness_F_store,fitness_M_store))\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "\n",
    "##### Activity costs\n",
    "\n",
    "Off-target cleavage reduces fitness by a factor $1 - s_{H}$ in the presence of Cas9 and gRNA and $(1 - s_{H})^2$ in the presence of Cas9 and two different gRNAs (gRNA targeting A and gRNA targeting B). \n",
    "Somatic cleavage reduces fitness by a factor $1-s_{S}$ in the presence of Cas9, gRNA and the target site of gRNA. \n",
    "X-shredding reduces fitness by a factor $1-s_{XS}$ in the presence of XS.\n",
    "Each selection coefficient can vary depending on the sex. \n",
    "Here we assume the costs associated with each activity are dominant, i.e. the presence of 1 unit results in the same fitness cost as 2 units.\n",
    "\n",
    "The final fitness for each genotype is calcuated by multiplying the per genotype fitness based on expression and locus disruption by the fitness due to synergistic effects. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [
     6
    ],
    "hidden": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Takes in the lists of construct content and list of genotypes and calculates \n",
    "the relative fitness due to interactions between components of constructs.\n",
    "Outputs a fitness matrix consisting of two columns (females and males),\n",
    "where each row is a 2-locus diploid genotype containing relative fitness due to construct activity.\n",
    "\"\"\"\n",
    "function Get_synergistic_fitness(A_alleles_content,B_alleles_content,genotypes,p)\n",
    "    \n",
    "    fitness_M_store = Vector{Float64}() #male fitness\n",
    "    fitness_F_store = Vector{Float64}() #female fitness\n",
    "    \n",
    "    for i in 1:length(genotypes)\n",
    "        \n",
    "        genotype_1 = genotypes[i][:,1]\n",
    "        genotype_2 = genotypes[i][:,2]\n",
    "        genotype_1_Acontent = A_alleles_content[genotype_1[1]]\n",
    "        genotype_1_Bcontent = B_alleles_content[genotype_1[2]]\n",
    "        genotype_2_Acontent = A_alleles_content[genotype_2[1]]\n",
    "        genotype_2_Bcontent = B_alleles_content[genotype_2[2]]\n",
    "        \n",
    "        fitness_m = 1.0\n",
    "        fitness_f = 1.0\n",
    "        \n",
    "        ## Off target cleavage and somatic leakage\n",
    "        #if there is at least one Cas9...\n",
    "        if ((\"cas9\" in genotype_1_Acontent) | (\"cas9\" in genotype_2_Acontent)) | ((\"cas9\" in genotype_1_Bcontent) | (\"cas9\" in genotype_2_Bcontent))\n",
    "            \n",
    "            #and at least one grna-A for off target cleavage\n",
    "            if ((\"Agrna\" in genotype_1_Acontent) | (\"Agrna\" in genotype_2_Acontent)) | ((\"Agrna\" in genotype_1_Bcontent) | (\"Agrna\" in genotype_2_Bcontent))\n",
    "                fitness_m = fitness_m*(1-p[\"sHam\"])\n",
    "                fitness_f = fitness_f*(1-p[\"sHaf\"])\n",
    "                #and at least one A target site for somatic leakage\n",
    "                if (\"A1\" in genotype_1) |(\"A1\" in genotype_2)\n",
    "                    fitness_m = fitness_m*(1-p[\"sSam\"])\n",
    "                    fitness_f = fitness_f*(1-p[\"sSaf\"])\n",
    "                end\n",
    "            end\n",
    "    \n",
    "            #and at least one grna-B for off target cleavage\n",
    "            if ((\"Bgrna\" in genotype_1_Acontent) | (\"Bgrna\" in genotype_2_Acontent)) | ((\"Bgrna\" in genotype_1_Bcontent) | (\"Bgrna\" in genotype_2_Bcontent))\n",
    "                fitness_m = fitness_m*(1-p[\"sHbm\"])\n",
    "                fitness_f = fitness_f*(1-p[\"sHbf\"])\n",
    "                #and at least one B target site for somatic leakage\n",
    "                if (\"B1\" in genotype_1) | (\"B1\" in genotype_2)\n",
    "                    fitness_m = fitness_m*(1-p[\"sSbm\"])\n",
    "                    fitness_f = fitness_f*(1-p[\"sSbf\"])\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "    \n",
    "        ## X-shredder cost\n",
    "        if ((\"xs\" in genotype_1_Acontent) | (\"xs\" in genotype_2_Acontent)) | ((\"xs\" in genotype_1_Bcontent) | (\"xs\" in genotype_2_Bcontent))\n",
    "            fitness_m = fitness_m*(1-p[\"sXsf\"])\n",
    "            fitness_f = fitness_f*(1-p[\"sXsm\"])\n",
    "        end\n",
    "\n",
    "        push!(fitness_M_store,fitness_m)   \n",
    "        push!(fitness_F_store,fitness_f)   \n",
    "        \n",
    "    end\n",
    "    return(hcat(fitness_F_store,fitness_M_store))\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "code_folding": [],
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Get_synergistic_fitness"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Wrapper: \n",
    "Input: \n",
    "* 2 lists of construct components, one for each construct A and B\n",
    "* 2 lists of construct component fitness types, one for each construct A and B\n",
    "* 2 lists of diploid alleles at a single locus, one for each construct A and B\n",
    "* 1 list of 2-locus diploid genotypes\n",
    "\n",
    "The function calculates the fitness for each 2-locus diploid genotype \n",
    "e.g. ABAB, taking into consideration costs due to \n",
    "1) locus disruption \n",
    "2) construct expression \n",
    "3) construct activity \n",
    "Outputs a fitness matrix containing relative fitnesses of females and males (columns) for each\n",
    "2-locus diploid genotype (row). \n",
    "\"\"\"\n",
    "function Build_fitness_matrix(A_alleles_content,A_alleles_ir,A_genotypes,B_alleles_content,B_alleles_ir,B_genotypes,genotypes,p)\n",
    "    \n",
    "    A_fitness_matrix = Get_disruption_fitness(A_alleles_ir,A_genotypes,\n",
    "                                p[\"siaf\"],p[\"hiaf\"],p[\"sraf\"],p[\"hraf\"],p[\"siraf\"],\n",
    "                                p[\"siam\"],p[\"hiam\"],p[\"sram\"],p[\"hram\"],p[\"siram\"]) .* \n",
    "                        Get_construct_expression_fitness(A_alleles_content,A_genotypes,p)\n",
    "    \n",
    "    B_fitness_matrix = Get_disruption_fitness(B_alleles_ir,B_genotypes,\n",
    "                                p[\"sibf\"],p[\"hibf\"],p[\"srbf\"],p[\"hrbf\"],p[\"sirbf\"],\n",
    "                                p[\"sibm\"],p[\"hibm\"],p[\"srbm\"],p[\"hrbm\"],p[\"sirbm\"]) .* \n",
    "                        Get_construct_expression_fitness(B_alleles_content,B_genotypes,p)\n",
    "    \n",
    "    disruption_expression_matrix = Get_expression_fitness_matrix(A_fitness_matrix, B_fitness_matrix,A_genotypes,B_genotypes,genotypes)\n",
    "    synergistic_matrix = Get_synergistic_fitness(A_alleles_content,B_alleles_content,genotypes,p)\n",
    "    \n",
    "    fitness_matrix = disruption_expression_matrix .* synergistic_matrix\n",
    "    \n",
    "    return(fitness_matrix)\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true,
    "hidden": true
   },
   "source": [
    "### Homing\n",
    "\n",
    "Homing in the germline acts to convert genotypes from one to another before segregation during meiosis.\n",
    "Homing requires the presence of Cas9, gRNA and a target alelle and involves two processes. \n",
    "The first is cleavage, which occurs with probability $c$. \n",
    "The second is repair which can either occur via NHEJ with probability $j$, converting the WT to a resistant allale, or HDR with probability $1-j$.\n",
    "If NHEJ occurs, the WT is converted into an allele containing no transcripitonal units with the \"resistant\" fitness type with probability $p$ or with the \"insertion\" fitness type with probability $1-p$, allowing for modelling of both functional and non-functional resistance.\n",
    "To study evolutionary stability, if HDR has occured each transcriptional unit of each construct has a probability $d$ of mutating to a non-functional version, whilst maintaining fitness costs due to locus disruption (the unit is mutated rather than lost).\n",
    "The functions below generate a genotype x genotype (i x j) matrix, which contain the proportion of each pre-homed genotype (i, row) converted to each post-homed genotypes (j, column) due to homing.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "code_folding": [
     3,
     17
    ],
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sort_genotype"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Sorts the allele string so order is in numerical order, where A > B\n",
    "\"\"\"\n",
    "function sort_allele(unsorted_allele)\n",
    "    n1 = parse(Int64,unsorted_allele[1][2:end])\n",
    "    n2 = parse(Int64,unsorted_allele[2][2:end])\n",
    "    if n1 > n2\n",
    "        sorted_allele = [unsorted_allele[2,:]; unsorted_allele[1,:]]\n",
    "    else\n",
    "        sorted_allele = [unsorted_allele[1,:]; unsorted_allele[2,:]]\n",
    "    end\n",
    "    return(sorted_allele)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Sorts the genotype string so order is in numerical order, where A > B\n",
    "\"\"\"\n",
    "function sort_genotype(unsorted_genotype)\n",
    "    sorted_genotype = [\"X\" \"Y\"; \"X\" \"Y\"]\n",
    "    if unsorted_genotype[1,1] != unsorted_genotype[1,2]\n",
    "        if sort_allele(unsorted_genotype[1,:]) == unsorted_genotype[1,:]\n",
    "            sorted_genotype = unsorted_genotype\n",
    "        else\n",
    "            sorted_genotype[:,1] = unsorted_genotype[:,2]\n",
    "            sorted_genotype[:,2] = unsorted_genotype[:,1]\n",
    "        end\n",
    "    else\n",
    "        if sort_allele(unsorted_genotype[2,:]) == unsorted_genotype[2,:]\n",
    "            sorted_genotype =unsorted_genotype\n",
    "        else\n",
    "            sorted_genotype[:,1] = unsorted_genotype[:,2]\n",
    "            sorted_genotype[:,2] = unsorted_genotype[:,1]\n",
    "        end\n",
    "    end\n",
    "    return(sorted_genotype)\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "code_folding": [
     3,
     21,
     52,
     90
    ],
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "relabel_transition_dict"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Wrapper to calculate the construct dictionary for a specific construct\n",
    "\"\"\"\n",
    "function get_transition_dict(alleles_ir,\n",
    "                                    construct_label,\n",
    "                                    content_dict,\n",
    "                                    c,j,d,p)\n",
    "    \n",
    "    construct_alleles = [k for (k,v) in alleles_ir if v==\"disruption\"] #list of labels of construct + derivatives\n",
    "    construct_content = content_dict[construct_label] #focal construct content\n",
    "    transition_dict = Build_transition_dict(construct_content,c,j,d,p) #transition dictionary for focal construct\n",
    "    transition_dict = relabel_transition_dict(transition_dict,content_dict,construct_alleles)\n",
    "    return(transition_dict)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Takes in a list of construct components \n",
    "Calculates all possible mutant variations of the construct \n",
    "Calculates the equations which represent the probability of each mutant construct being produced.\n",
    "Returns a dictionary where the key is the list of construct components and the value is the equation.\n",
    "\"\"\"\n",
    "function Build_transition_dict(construct,c,j,d,p)\n",
    "    #Save the length of the intact construct\n",
    "    L = length(construct)\n",
    "    #Initiate the construct dictionary\n",
    "    transition_dict = Dict()\n",
    "    #IF there are components present in the construct mutation can occur\n",
    "    if L > 0\n",
    "        #Calculate the probability of producing each mutant construct\n",
    "        transition_dict = Make_transition_equation_dict(construct,transition_dict,c,j,d,L)\n",
    "        #add in the probability a non-functional mutant can be produced through end-joining \n",
    "        ## rather than mutation\n",
    "        transition_dict[[]] = transition_dict[[]]+c*j*(1-p)\n",
    "        #add in the probability of homing of the intact construct\n",
    "        n=0\n",
    "        transition_dict[construct] = c*(1-j)*d^n*(1-d)^(L-n)\n",
    "\n",
    "    #IF there are no components in the construct, no mutation to a non-functional resistant allele occurs\n",
    "    else\n",
    "        #add in the probability the mutant being produced through end-joining\n",
    "        #add in the probability the mutant is produced through homing of the original disruption\n",
    "        transition_dict[[]] = c*j*(1-p)+c*(1-j)\n",
    "    end\n",
    "    return(transition_dict)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Takes in a construct, its length and a dictionary where the key is the list of construct components \n",
    "and the value is the equation.\n",
    "L = number of transcriptional units in the construct\n",
    "n = number of transcriptional units in the broken construct\n",
    "\"\"\"\n",
    "function Make_transition_equation_dict(construct,transition_dict,c,j,d,L)\n",
    "\n",
    "    #iterates through the length of the construct\n",
    "    for i in 1:length(construct)\n",
    "        #removes a component of the construct\n",
    "        construct_mod = deepcopy(construct)\n",
    "        deleteat!(construct_mod,i)\n",
    "        \n",
    "        #calculates the length of the mutant construct\n",
    "        n= L-length(construct_mod)\n",
    "        \n",
    "        #subs the values into the equation\n",
    "        #c = probability of cleavage\n",
    "        #1-j = probability of homing given cleavage\n",
    "        #d = probability a single transcriptional unit mutates per homing event\n",
    "        #L = number of transcriptional units in the construct\n",
    "        #n = number of transcriptional units in the broken construct\n",
    "        equation_vals = c*(1-j)*d^n*(1-d)^(L-n)\n",
    "        \n",
    "        #Assign equation values to the dictionary\n",
    "        if construct_mod in keys(transition_dict)\n",
    "            transition_dict[construct_mod] = transition_dict[construct_mod]\n",
    "        else\n",
    "            transition_dict[construct_mod] = equation_vals\n",
    "        end\n",
    "        \n",
    "        #Repeat if there are still components in mutant construct \n",
    "        if length(construct_mod)<=(L-1)\n",
    "            transition_dict = Make_transition_equation_dict(construct_mod,transition_dict,c,j,d,L)\n",
    "        end\n",
    "    end\n",
    "    return(transition_dict)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Takes in construct dictionary, content dictionary and construct alleles list and relabels the dictionary\n",
    "with construct labels rather than content.\n",
    "\"\"\"\n",
    "function relabel_transition_dict(transition_dict,content_dict,construct_alleles)\n",
    "    relabelled_transition_dict = Dict()\n",
    "    for label in construct_alleles\n",
    "        content = content_dict[label]\n",
    "        if content in keys(transition_dict)\n",
    "            equation = transition_dict[content]\n",
    "            relabelled_transition_dict[label] = equation\n",
    "        end\n",
    "    end\n",
    "    return(relabelled_transition_dict)\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "cell_style": "center",
    "code_folding": [
     4,
     29,
     73
    ],
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Build_homing_matrix"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Fills a row of the homing matrix with the probabily that the focal genotype (row) will be\n",
    "converted to each other genotype (column) when cleavage occurs only at 1 locus (A OR B).\n",
    "\"\"\"\n",
    "function fill_homing_matrix(homing,genotype,genotypes,index,alleles_ir,res_allele,construct_dict,modified_locus,modified_chrom,c,j,p)\n",
    "    #no change\n",
    "    homing[index,index] = [1-c]\n",
    "    #If functional resistance is possible:\n",
    "    if \"resistant\" in values(alleles_ir)\n",
    "        #mutating into a cleavage resistant mutant with different fitness to the disrupted site\n",
    "        mutated_genotype = deepcopy(genotype)\n",
    "        mutated_genotype[modified_locus,modified_chrom] = res_allele \n",
    "        mutated_genotype = sort_genotype(mutated_genotype)\n",
    "        homing[index,findall(x->x==mutated_genotype,genotypes)] = [c*j*p]\n",
    "    end\n",
    "    #For every cons\n",
    "    for construct in keys(construct_dict)#1:length(construct_alleles)\n",
    "        homed_mut_genotype = deepcopy(genotype)\n",
    "        homed_mut_genotype[modified_locus,modified_chrom] = construct\n",
    "        homed_mut_genotype = sort_genotype(homed_mut_genotype)\n",
    "        homing[index,findall(x->x==homed_mut_genotype,genotypes)] = [construct_dict[construct]]\n",
    "    end\n",
    "    return(homing)\n",
    "end\n",
    "\"\"\"\n",
    "Fills a row of the homing matrix with the probabily that the focal genotype (row) will be\n",
    "converted to each other genotype (column) when cleavage occurs at the B locus and \n",
    "has already occured at the A locus.\n",
    "\"\"\"\n",
    "function fill_homing_matrix_BafterA(homing,index,genotypes,ir_dict,content_dict,a_genotype,a_equation,resB_allele,B_alleles_ir,cb,jb,db,pb)  \n",
    "\n",
    "    modg1a,modg1b,modg2a,modg2b = a_genotype\n",
    "    if ir_dict[modg1b] == \"wt\"\n",
    "        modified_Bchrom = 1\n",
    "        Bconstruct_dict = get_transition_dict(B_alleles_ir,\n",
    "                                            modg2b,\n",
    "                                            content_dict,\n",
    "                                            cb,jb,db,pb)  \n",
    "    elseif ir_dict[modg2b] == \"wt\"\n",
    "        modified_Bchrom = 2\n",
    "        Bconstruct_dict = get_transition_dict(B_alleles_ir,\n",
    "                                            modg1b,\n",
    "                                            content_dict,\n",
    "                                            cb,jb,db,pb)\n",
    "    end\n",
    "\n",
    "    #No change to genotype_b genotype\n",
    "    ab_genotype = deepcopy(a_genotype)\n",
    "    ab_equation = a_equation*(1-cb)\n",
    "    homing[index,findall(x->x==ab_genotype,genotypes)] = [ab_equation]\n",
    "\n",
    "    if \"resistant\" in values(B_alleles_ir)\n",
    "        #cleavage resistant functional mutant\n",
    "        ab_genotype = deepcopy(a_genotype)\n",
    "        ab_genotype[2,modified_Bchrom] = resB_allele\n",
    "        ab_genotype = sort_genotype(ab_genotype)\n",
    "        ab_equation = a_equation*cb*jb*pb\n",
    "        homing[index,findall(x->x==ab_genotype,genotypes)] = [ab_equation]\n",
    "    end\n",
    "    #construct genotypes\n",
    "    for Bconstruct in keys(Bconstruct_dict)\n",
    "        ab_genotype = deepcopy(a_genotype)\n",
    "        ab_genotype[2,modified_Bchrom] = Bconstruct\n",
    "        ab_genotype = sort_genotype(ab_genotype)\n",
    "        homing[index,findall(x->x==ab_genotype,genotypes)] = [a_equation*Bconstruct_dict[Bconstruct]]\n",
    "    end\n",
    "    return(homing)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Iterates through all genotypes and fills the homing matrix in with transition probabilities from genotypes prior\n",
    "to homing (rows) to genotypes after homing (columns). \n",
    "\"\"\"\n",
    "function Build_homing_matrix(A_alleles,A_alleles_content,A_alleles_ir,B_alleles,B_alleles_content,B_alleles_ir,genotypes,ca,cb,ja,jb,pa,pb,da,db)\n",
    "\n",
    "    #Generate content and insertion dictionaries for each allele\n",
    "    content_dict = merge(A_alleles_content,B_alleles_content)\n",
    "    ir_dict = merge(A_alleles_ir,B_alleles_ir)\n",
    "    \n",
    "    #find allele label for functional resistant loci is present\n",
    "    if \"resistant\" in values(A_alleles_ir)\n",
    "        pa=pa\n",
    "        resA_allele = findall(x->x==\"resistant\",A_alleles_ir)[1]\n",
    "    else\n",
    "        resA_allele = \" \"\n",
    "        pa=0.0\n",
    "    end      \n",
    "    \n",
    "    if \"resistant\" in values(B_alleles_ir)\n",
    "        pb=pb\n",
    "        resB_allele = findall(x->x==\"resistant\",B_alleles_ir)[1]\n",
    "    else\n",
    "        resB_allele = \" \"\n",
    "        pb=0.0\n",
    "    end\n",
    "\n",
    "    #Initiate the homing construct\n",
    "    homing = Array{Float64}(undef,length(genotypes),length(genotypes))\n",
    "    fill!(homing, 0.0)\n",
    "\n",
    "    #We begin by assuming no homing, therefore all diagonals are set to 1.0\n",
    "        #This is just so we dont have to assign all genotypes where no homing occurs.\n",
    "    homing[diagind(homing, 0)] .= 1.0\n",
    "\n",
    "    count = 0\n",
    "    for genotype in genotypes\n",
    "\n",
    "        count = count+1\n",
    "        index = findall(x->x==genotype,genotypes) \n",
    "        \n",
    "        #Store the allele labels for each allele at A and B sites\n",
    "        g1a,g1b,g2a,g2b = genotype\n",
    "        #Get the transition probability dictionaries for each allele depending on its content\n",
    "        g1a_construct_dict = get_transition_dict(A_alleles_ir,\n",
    "                                                        g1a,\n",
    "                                                        content_dict,\n",
    "                                                        ca,ja,da,pa)\n",
    "        g2a_construct_dict = get_transition_dict(A_alleles_ir,\n",
    "                                                        g2a,\n",
    "                                                        content_dict,\n",
    "                                                        ca,ja,da,pa)\n",
    "        g1b_construct_dict = get_transition_dict(B_alleles_ir,\n",
    "                                                        g1b,\n",
    "                                                        content_dict,\n",
    "                                                        cb,jb,db,pb)\n",
    "        g2b_construct_dict = get_transition_dict(B_alleles_ir,\n",
    "                                                        g2b,\n",
    "                                                        content_dict,\n",
    "                                                        cb,jb,db,pb)  \n",
    "        \n",
    "        #IS THERE A CAS9 present at least once:  \n",
    "        if ((\"cas9\" in content_dict[g1a]) | (\"cas9\" in content_dict[g2a]) | (\"cas9\" in content_dict[g1b]) | (\"cas9\" in content_dict[g2b]))\n",
    "            \n",
    "            #Cleavage is possible at locus A only\n",
    "                #gRNA-A is present at least once\n",
    "            if (((\"Agrna\" in content_dict[g1a]) | (\"Agrna\" in content_dict[g2a]) | (\"Agrna\" in content_dict[g1b]) | (\"Agrna\" in content_dict[g2b]))\n",
    "                #AND gRNA-B is absent OR B-WT is absent OR B-WT is present twice\n",
    "                & (((\"Bgrna\" ∉ content_dict[g1a]) & (\"Bgrna\" ∉ content_dict[g2a]) & (\"Bgrna\" ∉ content_dict[g1b]) & (\"Bgrna\" ∉ content_dict[g2b]))\n",
    "                | ((ir_dict[g1b]!=\"wt\") & (ir_dict[g2b]!=\"wt\")) \n",
    "                        | ((ir_dict[g1b]==\"wt\") & (ir_dict[g2b]==\"wt\")) ))\n",
    "                \n",
    "                 modified_locus = 1 #editing occurs only at the A locus\n",
    "            \n",
    "                #Check whether there is a target site at A\n",
    "                #2.1 #if there is one WT, homing will occur. \n",
    "                if ((ir_dict[g1a]==\"wt\") & ((ir_dict[g2a]==\"disruption\")|(ir_dict[g2a]==\"resistant\"))) \n",
    "                    modified_chrom = 1\n",
    "                    construct_dict = g2a_construct_dict\n",
    "                    homing = fill_homing_matrix(homing,genotype,genotypes,index,A_alleles_ir,resA_allele,construct_dict,modified_locus,modified_chrom,ca,ja,pa)   \n",
    "                #2.2 #if there is one WT, homing will occur. \n",
    "                elseif ((ir_dict[g2a]==\"wt\") & ((ir_dict[g1a]==\"disruption\")|(ir_dict[g1a]==\"resistant\")))   \n",
    "                    modified_chrom = 2\n",
    "                    construct_dict = g1a_construct_dict\n",
    "                    homing = fill_homing_matrix(homing,genotype,genotypes,index,A_alleles_ir,resA_allele,construct_dict,modified_locus,modified_chrom,ca,ja,pa) \n",
    "                #2.3 #if there are 2 wt, mutation will occur.\n",
    "                elseif ((ir_dict[g1a]==\"wt\") & (ir_dict[g2a]==\"wt\"))\n",
    "                    #TO BE COMLPETED.\n",
    "                #2 end\n",
    "                end\n",
    "            \n",
    "\n",
    "                \n",
    "            #Cleavage is possible at locus B only\n",
    "                #gRNA-B is present at least once\n",
    "            elseif (((\"Bgrna\" in content_dict[g1a]) | (\"Bgrna\" in content_dict[g2a]) | (\"Bgrna\" in content_dict[g1b]) | (\"Bgrna\" in content_dict[g2b]))\n",
    "                #AND gRNA-A is absent OR A-WT is absent OR A-WT is present twice\n",
    "                & (((\"Agrna\" ∉ content_dict[g1a]) & (\"Agrna\" ∉ content_dict[g2a]) & (\"Agrna\" ∉ content_dict[g1b]) & (\"Agrna\" ∉ content_dict[g2b]))\n",
    "                | ((ir_dict[g1a]!=\"wt\") & (ir_dict[g2a]!=\"wt\"))\n",
    "                        | ((ir_dict[g1a]==\"wt\") & (ir_dict[g2a]==\"wt\"))))\n",
    "                \n",
    "                modified_locus = 2 #editing occurs only at the A locus\n",
    "            \n",
    "                #Check whether there is a target site \n",
    "                #3.1 #if there is one WT, homing will occur. \n",
    "                if ((ir_dict[g1b]==\"wt\") & ((ir_dict[g2b]==\"disruption\")|(ir_dict[g2b]==\"resistant\")))\n",
    "                    modified_chrom = 1\n",
    "                    construct_dict = g2b_construct_dict\n",
    "                    homing = fill_homing_matrix(homing,genotype,genotypes,index,B_alleles_ir,resB_allele,construct_dict,modified_locus,modified_chrom,cb,jb,pb)\n",
    "                #3.2 #if there is one WT, homing will occur. \n",
    "                elseif ((ir_dict[g2b]==\"wt\") & ((ir_dict[g1b]==\"disruption\")|(ir_dict[g1b]==\"resistant\")))\n",
    "                    modified_chrom = 2\n",
    "                    construct_dict = g1b_construct_dict\n",
    "                    homing = fill_homing_matrix(homing,genotype,genotypes,index,B_alleles_ir,resB_allele,construct_dict,modified_locus,modified_chrom,cb,jb,pb)\n",
    "                #3.3 #if there are 2 wt, mutation will occur.\n",
    "                elseif ((ir_dict[g1b]==\"wt\") & (ir_dict[g2b]==\"wt\"))\n",
    "                    #TO BE COMLPETED.\n",
    "                #3 end\n",
    "                end\n",
    "            #Cleavage is possible at A and B loci\n",
    "                #both gRNA-A and gRNA-B are present at least once\n",
    "            elseif (((\"Agrna\" in content_dict[g1a]) | (\"Agrna\" in content_dict[g2a]) | (\"Agrna\" in content_dict[g1b]) | (\"Agrna\" in content_dict[g2b]))\n",
    "                &\n",
    "                ((\"Bgrna\" in content_dict[g1a]) | (\"Bgrna\" in content_dict[g2a]) | (\"Bgrna\" in content_dict[g1b]) | (\"Bgrna\" in content_dict[g2b])))\n",
    "\n",
    "                # if there is one WT at A and 1 WT at B = homing possible at both \n",
    "                if ((((ir_dict[g1a]==\"wt\") & ((ir_dict[g2a]==\"disruption\")|(ir_dict[g2a]==\"resistant\"))) \n",
    "                   | ((ir_dict[g2a]==\"wt\") & ((ir_dict[g1a]==\"disruption\")|(ir_dict[g1a]==\"resistant\"))))\n",
    "                   &\n",
    "                   (((ir_dict[g1b]==\"wt\") & ((ir_dict[g2b]==\"disruption\")|(ir_dict[g2b]==\"resistant\"))) \n",
    "                   | ((ir_dict[g2b]==\"wt\") & ((ir_dict[g1b]==\"disruption\")|(ir_dict[g1b]==\"resistant\"))))\n",
    "                    )\n",
    "                    \n",
    "                    #Define which are the donor chromosomes and which are the wt (modified)\n",
    "                    if ir_dict[g1a] == \"wt\"\n",
    "                        modified_Achrom = 1\n",
    "                        Aconstruct_dict = g2a_construct_dict\n",
    "                    elseif ir_dict[g2a] == \"wt\"\n",
    "                        modified_Achrom = 2\n",
    "                        Aconstruct_dict = g1a_construct_dict\n",
    "                    end \n",
    "                    \n",
    "                    #Double homing\n",
    "                    a_genotype = deepcopy(genotype)\n",
    "                    a_equation = 1-ca\n",
    "                    homing = fill_homing_matrix_BafterA(homing,index,genotypes,ir_dict,content_dict,a_genotype,a_equation,resB_allele,B_alleles_ir,cb,jb,db,pb)\n",
    "                    \n",
    "                    #attempted home and change into functional resistant mutant\n",
    "                    if \"resistant\" in values(A_alleles_ir)\n",
    "                        #cleavage resistant functional mutant\n",
    "                        a_genotype = deepcopy(genotype)\n",
    "                        a_genotype[1,modified_Achrom] = resA_allele\n",
    "                        a_genotype = sort_genotype(a_genotype)\n",
    "                        a_equation = ca*ja*pa\n",
    "                        homing = fill_homing_matrix_BafterA(homing,index,genotypes,ir_dict,content_dict,a_genotype,a_equation,resB_allele,B_alleles_ir,cb,jb,db,pb)\n",
    "                    end\n",
    "                    #attempted home and change into mutant construct \n",
    "                    for Aconstruct in keys(Aconstruct_dict)\n",
    "                        a_genotype = deepcopy(genotype)\n",
    "                        a_genotype[1,modified_Achrom] = Aconstruct\n",
    "                        a_genotype = sort_genotype(a_genotype)\n",
    "                        a_equation = Aconstruct_dict[Aconstruct]\n",
    "                        homing = fill_homing_matrix_BafterA(homing,index,genotypes,ir_dict,content_dict,a_genotype,a_equation,resB_allele,B_alleles_ir,cb,jb,db,pb)\n",
    "                    end\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "    return(homing)                                                                                                                                                                                                \n",
    "end\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true,
    "hidden": true
   },
   "source": [
    "### Sex bias\n",
    "\n",
    "The probability of a male producing an X- or Y-bearing gamete depends on the presence of an X-shredder. The functions below generate a matrix containing the proportion of X- (columns 1) or Y- (column 2) bearing sperm depending on the genotype (row). m is the proporion of Y-bearing sperm produced for a male carrying at least one copy of the X-shredder. We assume that x-shredding function is dominant, thus heterozygotes and homozygotes for the xshredder create the same sex bias. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [
     3
    ],
    "hidden": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "sex distortion is dependent on the the presence of an \"xs\" in the constuct\n",
    "\"\"\"\n",
    "function Build_sex_ratio_matrix(A_alleles,A_alleles_content,B_alleles,B_alleles_content,genotypes,p)\n",
    "    \n",
    "    m = p[\"m\"]\n",
    "    content_dict = merge(A_alleles_content,B_alleles_content)\n",
    "    #Calculate the sex ratio of each genotype\n",
    "    sex_ratio_storeM = Vector{Float64}()\n",
    "    sex_ratio_storeF = Vector{Float64}()\n",
    "    #For each combined genotype e.g. ABAB\n",
    "    for genotype in genotypes\n",
    "        sex_ratio = []\n",
    "        #Check if each the genotype carries an X shredder and assign sex ratio accordingly:\n",
    "        if ((\"xs\" in content_dict[genotype[1,1]])|(\"xs\" in content_dict[genotype[1,2]])|(\"xs\" in content_dict[genotype[2,1]])|(\"xs\" in content_dict[genotype[2,2]]))\n",
    "            sex_ratio = m #WT \n",
    "            push!(sex_ratio_storeF,1-sex_ratio)   \n",
    "            push!(sex_ratio_storeM,sex_ratio) \n",
    "        else\n",
    "            push!(sex_ratio_storeF,0.5)   \n",
    "            push!(sex_ratio_storeM,0.5) \n",
    "        end\n",
    "    end\n",
    "    sex_ratio_matrix = hcat(sex_ratio_storeF,sex_ratio_storeM);\n",
    "    return(sex_ratio_matrix)\n",
    "end\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true,
    "hidden": true
   },
   "source": [
    "### Recombination\n",
    "\n",
    "Recombination can occur between the two loci with probability r during gamete production. \n",
    "The function generates a matrix of the proportion of each gamete (column) produced from each genotype (row), \n",
    "where each row refers to a diploid genotype (individual) and each column to a haploid genotype (gamete) at the autosomal loci only.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "code_folding": [
     7
    ],
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Build_recombination_matrix"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Calculate the probability of each genotype producing each gamete (allele)\n",
    "taking into consideration the recombination probability r between A and B loci.\n",
    "Returns matrix of the proportion of each gamete (columns) produced by each parental genotype (row)\n",
    "\"\"\"\n",
    "function Build_recombination_matrix(genotypes,alleles,p::Dict{String,Float64})\n",
    "    r = p[\"r\"]\n",
    "    \n",
    "    recombination_matrix = Array{Float64}(undef,(length(genotypes),length(alleles)))\n",
    "\n",
    "    empty_row = [0.0 for i in 1:length(alleles)]\n",
    "    empty_row = reshape(empty_row, 1, length(empty_row))\n",
    "\n",
    "    for a in enumerate(genotypes)\n",
    "        local genotype = a[2]\n",
    "        local i = a[1]\n",
    "        \n",
    "        #for genotype in genotypes\n",
    "\n",
    "        recombination_vec = deepcopy(empty_row)\n",
    "\n",
    "        #Screen whether each allele (gamete) is present in each genotype\n",
    "        nonrecombinant_mask = [(a == genotype[:,1])|(a == genotype[:,2]) for a in alleles]\n",
    "        #if 2 alleles are present (the genotype is heterozygous)\n",
    "\n",
    "        if sum(nonrecombinant_mask) == 2\n",
    "            #Assign the 2 matched alleles non-recombinant inheritence (1-r)/2\n",
    "            recombination_vec[nonrecombinant_mask] = collect([(1-r)/2, (1-r)/2])\n",
    "            #Assign the rest of the gametes that are not produced and inheritance of 0\n",
    "            recombination_vec[nonrecombinant_mask .== false] = empty_row[:,1:end-2]\n",
    "        #if 1 allele is present (the genotype is homozygous)\n",
    "        elseif sum(nonrecombinant_mask) == 1\n",
    "            #Assign the 1 matched allele non-recombinant inheritance (1-r)\n",
    "            recombination_vec[nonrecombinant_mask] = collect([(1-r)])\n",
    "            #Assign the rest of the gametes that are not produced and inheritance of 0\n",
    "            recombination_vec[nonrecombinant_mask .== false] = empty_row[:,1:end-1]\n",
    "        end\n",
    "\n",
    "        #Recombine the genotype\n",
    "        genotype_recomb = [\"X\" \"Y\";\"X\" \"Y\"]\n",
    "        genotype_recomb[1,1] = genotype[1,1]\n",
    "        genotype_recomb[1,2] = genotype[1,2]\n",
    "        genotype_recomb[2,1] = genotype[2,2]\n",
    "        genotype_recomb[2,2] = genotype[2,1]\n",
    "\n",
    "        #Screen whether each allele (gamete) is present in each genotype\n",
    "        recombinant_mask = [(a == genotype_recomb[:,1])|(a == genotype_recomb[:,2]) for a in alleles]\n",
    "        #if 2 alleles are present (the genotype is heterozygous)\n",
    "        if sum(recombinant_mask) == 2\n",
    "            #Assign the 2 matched alleles recombinant inheritence r/2\n",
    "            recombination_vec[recombinant_mask] = collect(recombination_vec[recombinant_mask]+[r/2,r/2])\n",
    "            #if 1 allele is present (the genotype is homozygous)\n",
    "            elseif sum(recombinant_mask) == 1\n",
    "                #Assign the 1 matched alleles recombinant r\n",
    "                recombination_vec[recombinant_mask] = collect(recombination_vec[recombinant_mask]+[r])\n",
    "        end\n",
    "        recombination_matrix[i,:]=recombination_vec\n",
    "    end\n",
    "    return(recombination_matrix)\n",
    "end\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "## Simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "code_folding": [
     5
    ],
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Simulation_wrapper"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Defines starting population and matrices, \n",
    "runs simulation and \n",
    "calculates additional metrics from output.\n",
    "\"\"\"\n",
    "function Simulation_wrapper(params,             #Dictionary of parameters\n",
    "                            A_alleles_content,  #Dictionary of components in each A allele\n",
    "                            A_alleles_ir,       #Dictionary of disruption type in each A allele\n",
    "                            B_alleles_content,  #Dictionary of components in each B llele\n",
    "                            B_alleles_ir,       #Dictionary of disruption type in each B allele\n",
    "                            fitness_effect,     #adult or larvae\n",
    "                            transgenic_genotype,#e.g. [\"A1\" \"A2\"; \"B1\" \"B2\"]\n",
    "                            release_freq,       #Frequency of released males (as)\n",
    "                            resistant_allele, #e.g. \"A2\"\n",
    "                            resistant_freq,     #Frequency of resistant allele\n",
    "                            t)                  #number of generations\n",
    "    \n",
    "    #Define lists of alleles and genotypes based on the number of A and B alleles\n",
    "    A_alleles,\n",
    "    B_alleles,\n",
    "    alleles,\n",
    "    A_genotypes,\n",
    "    B_genotypes,\n",
    "    genotypes = define_string_genotypes(length(A_alleles_content),length(B_alleles_content));\n",
    "    \n",
    "    #Initiate the population\n",
    "    starting_population_post_release = release_transgenic_males(\n",
    "        Initiate_prerelease_population_3lifestages(resistant_allele, resistant_freq, params, genotypes),\n",
    "        transgenic_genotype,\n",
    "        release_freq);\n",
    "    \n",
    "    #Define the matrices\n",
    "    matrices = make_matrices(params,\n",
    "                        genotypes,\n",
    "                        alleles,\n",
    "                        A_alleles_content,\n",
    "                        A_alleles_ir,\n",
    "                        A_genotypes,\n",
    "                        B_alleles_content,\n",
    "                        B_alleles_ir,\n",
    "                        B_genotypes);\n",
    "    \n",
    "    #Run the simulation\n",
    "    output = Simulate_timeseries(\n",
    "                    params,\n",
    "                    matrices,\n",
    "                    fitness_effect,\n",
    "                    t,\n",
    "                    starting_population_post_release,\n",
    "                    genotypes,\n",
    "                    alleles)\n",
    "\n",
    "    data_dict = Simulation_processing_wrapper(output,A_alleles,B_alleles,alleles,genotypes)\n",
    "    \n",
    "    return(data_dict)\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "\n",
    "###### Defining the population upon release\n",
    "\n",
    "We begin with a Nx6 matrix of numbers of individuals of each genotype after transgenic release, where N is the number of genotypes. Each row refers to a genotype and each column to the following catagories or organism: Female zygotes, males zygotes, female pupae, male pupae, female adults and male adults.\n",
    "The pre-release population may contain only WT individuals (ABAB) or may contain a resistant allele, assumed to be at Hardy-Weinberg equilibrium at a user-define allele frequency. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "code_folding": [
     3,
     17,
     52,
     83
    ],
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "release_transgenic_males"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Initiates a dictionary of genotypes frequencies, where each genotype freq has an initial value of 0. \n",
    "\"\"\"\n",
    "function Initiate_genotype_freq(genotypes)\n",
    "    freq_genotype = Dict()\n",
    "    for i in 1:length(genotypes)\n",
    "       freq_genotype[genotypes[i]] = 0.0\n",
    "    end\n",
    "    return(freq_genotype)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Returns 2 dictionaries containing the numbers of each genotype, one for females and one for males.\n",
    "The total number of males or females is defined by the variable params: OJ, f and α and refers to the equilibrium population size of each population.\n",
    "The population can contain both WT individuals (ABAB) and genotypes containing a resistant allele.\n",
    "The population is assumed to be at hardy weinberg equilibrium, with the resistant allele at a frequency defined by the variable \"resistant_freq\".\n",
    "\"\"\"\n",
    "function Initiate_prerelease_population(resistant_allele, resistant_freq,params,genotypes)\n",
    "    \n",
    "    Rm = (params[\"f\"]*params[\"OJ\"])/2\n",
    "    Nf = params[\"α\"]*(Rm-1)/params[\"f\"]\n",
    "\n",
    "    #Define each frequency initially as 0\n",
    "    genotype_freqf = Initiate_genotype_freq(genotypes)\n",
    "    genotype_freqm = Initiate_genotype_freq(genotypes)\n",
    "\n",
    "    #homozygous genotype\n",
    "    if resistant_allele[1] == 'A'\n",
    "        het = [\"A1\" resistant_allele; \"B1\" \"B1\"]\n",
    "        hom = [resistant_allele resistant_allele; \"B1\" \"B1\"]\n",
    "    elseif resistant_allele[1] == 'B'\n",
    "        het = [\"A1\" \"A1\"; \"B1\" resistant_allele]\n",
    "        hom = [\"A1\" \"A1\"; resistant_allele resistant_allele]\n",
    "    end\n",
    "    \n",
    "    genotype_freqf[[\"A1\" \"A1\"; \"B1\" \"B1\"]] = (1-resistant_freq)^2*Nf\n",
    "    genotype_freqf[hom] = resistant_freq^2*Nf\n",
    "    genotype_freqf[het] = 2 * resistant_freq * (1-resistant_freq)*Nf\n",
    "\n",
    "    genotype_freqm[[\"A1\" \"A1\"; \"B1\" \"B1\"]] = (1-resistant_freq)^2*Nf\n",
    "    genotype_freqm[hom] = resistant_freq^2*Nf\n",
    "    genotype_freqm[het] = 2 * resistant_freq * (1-resistant_freq)*Nf\n",
    "\n",
    "    return(genotype_freqf,genotype_freqm)\n",
    "end\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "Initiates the prerelease population of adults. \n",
    "Back-calculates the number of zygotes and pupae for a given adult population and parameter values. \n",
    "Returns a 6xN matrix, where N is the number of genotypes, of numbers of zygote, pupae and adults for females and males, with column headers labelled by genotype.\n",
    "\"\"\"\n",
    "function Initiate_prerelease_population_3lifestages(resistant_allele, resistant_freq, params, genotypes)\n",
    "    \n",
    "    genotype_Nf,genotype_Nm = Initiate_prerelease_population(resistant_allele, resistant_freq, params, genotypes)\n",
    "\n",
    "    starting_population = Array{Float64}(undef,(6,length(genotypes)))\n",
    " \n",
    "    #Modify data type to arrays in the the order of genotypes\n",
    "    starting_population[5,:] = [genotype_Nf[i] for i in genotypes]\n",
    "    starting_population[6,:] = [genotype_Nm[i] for i in genotypes]\n",
    "    \n",
    "    #Assuming the population is at equilibrium AND the population is pre-release\n",
    "    #No. zygotes\n",
    "    starting_population[1,:] =  starting_population[5,:] .* params[\"f\"] /2\n",
    "    starting_population[2,:] =  starting_population[6,:] .* params[\"f\"] /2\n",
    "\n",
    "    #No. zygotes after density dependent mortality (pupae)\n",
    "    n_zygotes_sum = sum(starting_population[1,:])+sum(starting_population[2,:])\n",
    "    juvenile_survival_prob = params[\"OJ\"]*(params[\"α\"]/(params[\"α\"]+n_zygotes_sum))\n",
    "    starting_population[3,:]= starting_population[5,:] .* params[\"f\"] /2 * juvenile_survival_prob\n",
    "    starting_population[4,:] = starting_population[6,:] .* params[\"f\"] /2 * juvenile_survival_prob\n",
    "    genotypes_sym = [join(t) for t in genotypes]\n",
    "    N_individuals = DataFrame(starting_population,Symbol.(genotypes_sym))\n",
    "    \n",
    "    return(N_individuals)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Takes in a 6xN matrix, where N is the number of genotypes, defining the numbers of genotypes for a population pre-release and \n",
    "returns a Nx6 matrix of numbers of each genotype after release of males. \n",
    "The frequency and genotype of the release males is defined by the input.\n",
    "\"\"\"\n",
    "function release_transgenic_males(starting_population,transgenic_genotype,release_freq)\n",
    "    #Here we make the assumption that all released individuals are fully fit\n",
    "    transgenic_genotype = Symbol(join(transgenic_genotype))\n",
    "    starting_population[6,transgenic_genotype] = sum(starting_population[6,:])*release_freq\n",
    "    \n",
    "    starting_population = transpose(Array(starting_population))\n",
    "    \n",
    "    return(starting_population)\n",
    "end\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "\n",
    "###### Time series simulation\n",
    "\n",
    "Parameter values and the allele content (transcriptional units) and fitness type for each locus are used to produce matrices (using the functions in the Model Definition section) for applying fitness, homing, sex ratio and recombination processes to the population.\n",
    "\n",
    "During each generation:\n",
    "* Eggs and sperm are produced, taking into consideration homing, recombination and sex ratio bias. \n",
    "* Zygotes are procuded assuming random mating and that males are not limiting. \n",
    "* Fitness effects are applied at either the zygote or adult stage. \n",
    "* Density dependence is applied at the zygote stage after fitness effects (if applicable to zygotes). \n",
    "* Censusing of both the zygote and adults are taken after any fitness effects are applied.\n",
    "\n",
    "The ouput is a dictionary of time-series for allele or genotype frequencies, numbers of individuals and correlation between constructs.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "code_folding": [
     7,
     43,
     78,
     111,
     160,
     186
    ],
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Make_dataframe (generic function with 1 method)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Builds the matrices to be used based on the \n",
    "parameters, \n",
    "construct contents, \n",
    "fitness types\n",
    "and numbers of alleles/genotypes. \n",
    "\"\"\"\n",
    "function make_matrices(params,            #Dictionary of parameters\n",
    "                        genotypes,        #List of genotypes, each entry = [\"A1\" \"A2\"; \"B1\" \"B2\"]\n",
    "                        alleles,          #List of alleles, each entry = [\"A1\", \"B1\"]\n",
    "                        A_alleles_content,#Dictionary of components in each A allele\n",
    "                        A_alleles_ir,     #Dictionary of disruption type in each A allele\n",
    "                        A_genotypes,      #List of A genotypes, each entry = [\"A1\", \"A1\"]\n",
    "                        B_alleles_content,#Dictionary of components in each B allele\n",
    "                        B_alleles_ir,     #Dictionary of disruption type in each B allele\n",
    "                        B_genotypes)      #List of B genotypes, each entry = [\"B1\", \"B1\"]\n",
    "    \n",
    "    A_alleles = keys(A_alleles_content)\n",
    "    B_alleles = keys(B_alleles_content)\n",
    "    \n",
    "    fitness_matrix = Build_fitness_matrix(A_alleles_content,A_alleles_ir,A_genotypes,B_alleles_content,B_alleles_ir,B_genotypes,genotypes,params)\n",
    "    homing_matrix_f = Build_homing_matrix(A_alleles,A_alleles_content,A_alleles_ir,B_alleles,B_alleles_content,B_alleles_ir,genotypes,params[\"caf\"],params[\"cbf\"],params[\"jaf\"],params[\"jbf\"],params[\"paf\"],params[\"pbf\"],params[\"daf\"],params[\"dbf\"]);\n",
    "    homing_matrix_m = Build_homing_matrix(A_alleles,A_alleles_content,A_alleles_ir,B_alleles,B_alleles_content,B_alleles_ir,genotypes,params[\"cam\"],params[\"cbm\"],params[\"jam\"],params[\"jbm\"],params[\"pam\"],params[\"pbm\"],params[\"dam\"],params[\"dbm\"]);\n",
    "    sex_ratio_matrix = Build_sex_ratio_matrix(A_alleles,A_alleles_content,B_alleles,B_alleles_content,genotypes,params)\n",
    "    recombination_matrix = Build_recombination_matrix(genotypes,alleles,params);\n",
    "\n",
    "    return(fitness_matrix,\n",
    "            homing_matrix_f,\n",
    "            homing_matrix_m,\n",
    "            recombination_matrix,\n",
    "            sex_ratio_matrix)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Takes in the number of females adults, female homing matrix, recombination matrix and parameter values. \n",
    "The number of eggs produced by each genotype is calculated. \n",
    "Homing is applied by multiplying the vector of egg numbers per genotype by the female homing matrix, \n",
    "and summing over the columns. \n",
    "Recomnbination is applied and the number of each gamete produced per genotype is calculated (45x9 matrix) \n",
    "by multiplyng the number of eggs per genotype after homing by the recombination matrix. \n",
    "The matrix is summed over each column to find the number of eggs produced of each gamete genotype, \n",
    "returning a vector with a length equal to the number of gametes.\n",
    "\"\"\"\n",
    "function make_eggs(genotype_Nf_adult,          #list of numbers of adult females of each genotype\n",
    "                    homing_matrix_f_vals,      #female homing matrix\n",
    "                    recombination_matrix_vals, #recombination matrix\n",
    "                    params)                    #Dictionary of parameters\n",
    "\n",
    "                #Calulate the number of eggs produced by each reproducing adult female\n",
    "                neggs_per_genotype = genotype_Nf_adult * params[\"f\"];\n",
    "\n",
    "                #Homing:\n",
    "                #Calculate the numbers of eggs produced from each genotype after homing\n",
    "                neggs_per_genotype_per_genotype_homing = neggs_per_genotype.*homing_matrix_f_vals\n",
    "                neggs_per_genotype_homing = sum(neggs_per_genotype_per_genotype_homing,dims=1)[:]\n",
    "\n",
    "                #recombination\n",
    "                #Calulate the number of each gamete produced by each genotype after recombination\n",
    "                neggs_per_genotype_per_gamete = neggs_per_genotype_homing .*recombination_matrix_vals;\n",
    "                # Calculate the number of each gamate produced by the female population in total\n",
    "                neggs_per_gamete = sum(neggs_per_genotype_per_gamete,dims=1)[:];\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Takes in the number of male adults, male homing matrix, recombination matrix and sex ratio matrix. \n",
    "Males are assumed to be not limiting, so the number of males is converted to frequency. \n",
    "Homing is applied by multiplying the frequency of males per genotype by the male homing matrix, \n",
    "and summing over the columns. \n",
    "Recomnbination is applied and the frequency of each gamete produced per genotype is calculated (45x9 matrix) \n",
    "by multiplyng the frequency of males per genotype after homing by the recombination matrix. \n",
    "Sex ratio bias is applied by multiplying the 45x9 matrix containing the frequency of each gamete per genotype\n",
    "by each column of the the sex ratio matrix separately, providing 2 matrices, one for the proportion of each Y-bearing gamete\n",
    "genotype per adult male and another for the X-bearing equivalent.\n",
    "The columns of each X-bearing and Y-bearing sperm matrix are summed to find the frequency of \n",
    "each gamete genotype produced.\n",
    "Returns a Gx2 matrix, where G is the number of gametes, containing the frequency of X-bearing sperm (column 1) and Y-bearing sperm (column 2)\n",
    "of each gamate genotype (rows).\n",
    "\"\"\"\n",
    "function make_sperm(genotype_Nm_adult,        #list of numbers of adult males of each genotype\n",
    "                    homing_matrix_m_vals,     #male homing matrix\n",
    "                    recombination_matrix_vals,#recombination matrix\n",
    "                    sex_ratio_matrix_vals)    #sex ration matrix\n",
    "\n",
    "                #conver male numbers to frequency. \n",
    "                genotype_freqm_fitness = genotype_Nm_adult./sum(genotype_Nm_adult)\n",
    "\n",
    "                #Homing:\n",
    "                genotype_freqm_per_genotype_homing = genotype_freqm_fitness.*homing_matrix_m_vals\n",
    "                genotype_freqm_homing = sum(genotype_freqm_per_genotype_homing,dims=1)[:]\n",
    "\n",
    "                #recombination\n",
    "                freqsperm_per_genotype = genotype_freqm_homing.*recombination_matrix_vals;\n",
    "\n",
    "                #sex ratio\n",
    "                freqsperm_per_genotype_per_gamete_X = freqsperm_per_genotype.*sex_ratio_matrix_vals[:,1]\n",
    "                freqsperm_per_genotype_per_gamete_Y = freqsperm_per_genotype.*sex_ratio_matrix_vals[:,2]\n",
    "\n",
    "                freqsperm_per_gamete_X = sum(freqsperm_per_genotype_per_gamete_X,dims=1)[:]\n",
    "                freqsperm_per_gamete_Y = sum(freqsperm_per_genotype_per_gamete_Y,dims=1)[:]\n",
    "                freqsperm_per_gamete = vcat(freqsperm_per_gamete_X,freqsperm_per_gamete_Y)\n",
    "\n",
    "                return(hcat(freqsperm_per_gamete_X,freqsperm_per_gamete_Y))\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Takes in a Gx3 matrix, where G is the number of gametes, containing egg and sperm (X and Y) \n",
    "numbers or frequencies and calculates the proportion of \n",
    "each female and male zygote produced of each diploid genotype. \n",
    "Assumes random mating and that males are not limiting. \n",
    "Returns as Nx2 matrix, where N is the number of diploid genotypes, containing numbers of female (column 1) and male (column 2) genotypes. \n",
    "\"\"\"\n",
    "function Find_nzygotes(gametes,   #matrix containing number of eggs, X-bearing sperm and Y-bearing sperm of each gamete genotype\n",
    "                        alleles,  #list of alleles, each entry = [\"A1\", \"B1\"]\n",
    "                        genotypes)#list of genotypes, each entry = [\"A1\" \"A2\"; \"B1\" \"B2\"]\n",
    "    \n",
    "    #Taking in matrices of the number of eggs and sperm produced by each genotype:\n",
    "    #Calculates the proportion of each female and male zygote of each genotype produced.\n",
    "    \n",
    "    neggs_genotype = gametes[:,1]\n",
    "    freqsperm_genotype_X = gametes[:,2]\n",
    "    freqsperm_genotype_Y = gametes[:,3]\n",
    "    \n",
    "    #Female zygote genotype \n",
    "    genotype_freq_F_vec = []\n",
    "    #Male zygote genotype \n",
    "    genotype_freq_M_vec = []\n",
    "    \n",
    "    #For each potential zygote genotype\n",
    "    for genotype in genotypes\n",
    "        #Check if homozygote\n",
    "        if genotype[:,1] == genotype[:,2]\n",
    "            #Find the index of the gamete present in the genotype\n",
    "            index = findall(x->x==genotype[:,1],alleles)[1]\n",
    "            #multiply the number of eggs by the frequency X sperm of the specified gamete\n",
    "            genotype_freq_F = neggs_genotype[index]*freqsperm_genotype_X[index]\n",
    "            #multiply the number of eggs by the frequency Y sperm of the specified gamete\n",
    "            genotype_freq_M = neggs_genotype[index]*freqsperm_genotype_Y[index]\n",
    "        else\n",
    "            #Find the index of each allele in the genotype\n",
    "            index1 = findall(x->x==genotype[:,1],alleles)[1]\n",
    "            index2 = findall(x->x==genotype[:,2],alleles)[1]\n",
    "            #multiply the number of eggs by the frequency X sperm of the specified gametes in both combinations\n",
    "            genotype_freq_F = ((neggs_genotype[index1]*freqsperm_genotype_X[index2])) + ((freqsperm_genotype_X[index1]*neggs_genotype[index2]))\n",
    "            #multiply the number of eggs by the frequency Y sperm of the specified gametes in both combinations\n",
    "            genotype_freq_M = ((neggs_genotype[index1]*freqsperm_genotype_Y[index2])) + ((freqsperm_genotype_Y[index1]*neggs_genotype[index2]))\n",
    "\n",
    "        end\n",
    "        push!(genotype_freq_F_vec,genotype_freq_F)\n",
    "        push!(genotype_freq_M_vec,genotype_freq_M)\n",
    "    end\n",
    "    \n",
    "    \n",
    "    return(hcat(genotype_freq_F_vec,genotype_freq_M_vec))\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Takes in a set of parameters and a Nx2 matrix, where N is the number of diploid genotypes, of numbers per genotype. \n",
    "Applies the density dependent function to the population. \n",
    "Returns a Nx2 matrix of numbers of individuals per genotype after density dependence has been applied. \n",
    "\"\"\"\n",
    "function apply_density_dependence(params, #Dictionary of parameters\n",
    "                                n_zygotes)#matrix containing male and female zygote numbers for each genotype\n",
    "    n_zygotes_sum = sum(n_zygotes)\n",
    "    juvenile_survival_prob = params[\"OJ\"]*(params[\"α\"]/(params[\"α\"]+n_zygotes_sum))\n",
    "    n_zygote_dd = n_zygotes * juvenile_survival_prob\n",
    "\n",
    "    return(n_zygote_dd)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Simulation of population through time, in discrete non-overlapping generations. \n",
    "The function takes in: parameters, matrixes, fitness effect, a starting population (post release) and \n",
    "the number of generations required.\n",
    "Eggs and sperm are produced, taking into consideration homing, recombination and sex ratio bias. \n",
    "Zygotes are procuded assumin random mating and that males are not limiting. \n",
    "fitness effects are applied according to the fitness effect input(zygote or adult)\n",
    "Density dependence is applied at the zygote stage before zygotes develop into pupae. \n",
    "Zygote mortality (if applied) is applied before density dependence and the zygote population is censused after. \n",
    "The adult population is also censused after fitness effects are applied (if applicable).\n",
    "Zygote, pupae and adult population numbers for each sex are saved in a 45x6xt matrix, \n",
    "and egg numbers and sperm frequencies are saved in a 9x3xt matrix, where t is the number of generations. \n",
    "To enable intuitive data access, the output matrices are processed into a dictionary of dataframes\n",
    "containing txN matrices for each diploid population, a txG matrix for egg numbes and a tx18 for sperm frequency,\n",
    "where N is the number of diploid genotypes and G is the number of gametes,\n",
    "where columns are accessible by labelled genotypes.\n",
    "\"\"\"\n",
    "function Simulate_timeseries(params,        #Dictionary of parameters\n",
    "                                        matrices,      #list of matrices\n",
    "                                        fitness_effect,#adult or larvae\n",
    "                                        t,             #number of generations\n",
    "                                        starting_population_post_release,#starting population matrix\n",
    "                                        genotypes,     #list of genotypes, each entry = [\"A1\" \"A2\"; \"B1\" \"B2\"]\n",
    "                                        alleles)       #list of alleles, each entry = [\"A1\", \"B1\"]\n",
    "    \n",
    "    #Unpack list of pre-defined matrices\n",
    "    fitness_matrix_vals,\n",
    "    homing_matrix_m_vals,\n",
    "    homing_matrix_f_vals,\n",
    "    recombination_matrix_vals,\n",
    "    sex_ratio_matrix_vals = matrices\n",
    "    \n",
    "    #define (45x6xt) storage for n. individuals (f zygote, m zygote, f pupae, m pupae, f adult, m adult)\n",
    "    n = Array{Float64}(undef,(length(genotypes),6,t+1))\n",
    "    #define (9x3xt) storage for n. gametes (eggs, X-bearing sperm, Y-bearing sperm))\n",
    "    g = Array{Float64}(undef,(length(alleles),3,t+1))\n",
    "\n",
    "    #save starting population from input\n",
    "    n[:,:,1]=starting_population_post_release \n",
    "    \n",
    "    #same no. adults in starting pop\n",
    "    genotype_Nf_adult=n[:,5,1]\n",
    "    genotype_Nm_adult=n[:,6,1]\n",
    "\n",
    "    for gen in 1:t\n",
    "        \n",
    "        #female adults produce eggs \n",
    "        neggs_per_gamete =  make_eggs(genotype_Nf_adult,homing_matrix_f_vals,recombination_matrix_vals,params)\n",
    "        g[:,1,gen+1] = neggs_per_gamete\n",
    "        #male adults produce sperm\n",
    "        freqsperm_per_gamete = make_sperm(genotype_Nm_adult,homing_matrix_m_vals,recombination_matrix_vals,sex_ratio_matrix_vals)\n",
    "        g[:,2:3,gen+1] = freqsperm_per_gamete\n",
    "        #zygotes are generated from eggs and sperm\n",
    "        n_zygotes = Find_nzygotes(g[:,:,gen+1],alleles,genotypes)\n",
    "        #fitess costs are applied to zygotes if applicable\n",
    "        fitness_effect == \"zygote\" ? n_zygotes = n_zygotes .*fitness_matrix_vals : n_zygotes = n_zygotes\n",
    "        n[:,1:2,gen+1] = n_zygotes\n",
    "        #density dependent mortality is applied\n",
    "        n_pupae = apply_density_dependence(params,n_zygotes)\n",
    "        n[:,3:4,gen+1] = n_pupae\n",
    "        #fitess costs are applied to zygotes if applicable\n",
    "        fitness_effect == \"adult\" ? n_adults = n_pupae .*fitness_matrix_vals : n_adults = n_pupae\n",
    "        n[:,5:6,gen+1] = n_adults\n",
    "        #Number of adults are updated for generation t+1\n",
    "        genotype_Nf_adult=n_adults[:,1]\n",
    "        genotype_Nm_adult=n_adults[:,2]\n",
    "    end \n",
    "    \n",
    "    #Store as dataframes in dictionary for intuitive labelled access \n",
    "    genotypes_sym = [join(t) for t in genotypes]\n",
    "    alleles_sym = [join(t) for t in alleles]\n",
    "    header = Symbol.(genotypes_sym)\n",
    "    output = Dict(\n",
    "        \"genotypenumber_adult_female\" => DataFrame(transpose(n[:,5,:]), header),\n",
    "        \"genotypenumber_adult_male\" => DataFrame(transpose(n[:,6,:]), header),\n",
    "        \"genotypenumber_pupae_female\" => DataFrame(transpose(n[:,3,:]), header),\n",
    "        \"genotypenumber_pupae_male\" => DataFrame(transpose(n[:,4,:]), header),\n",
    "        \"genotypenumber_zygote_female\" => DataFrame(transpose(n[:,1,:]), header),\n",
    "        \"genotypenumber_zygote_male\" => DataFrame(transpose(n[:,2,:]), header),\n",
    "        \"eggnumber\" => DataFrame(transpose(g[:,1,:]), Symbol.(alleles_sym)),\n",
    "        \"spermfreq\" => DataFrame(hcat(transpose(g[:,2,:]),transpose(g[:,3,:])),Symbol.([string.(alleles_sym).*\"X\" ; string.(alleles_sym).*\"Y\"]))\n",
    "    )\n",
    "\n",
    "\n",
    "    return(output)\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "\n",
    "###### Post simulation processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "code_folding": [
     4,
     74,
     85,
     110,
     137
    ],
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Find_correlation_df (generic function with 1 method)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Takes in a simulation output and calculates various summary metrics from the numbers of individuals/gametes. \n",
    "Returns a dictionary containing the same output data with additional entries. \n",
    "\"\"\"\n",
    "function Simulation_processing_wrapper(output,A_alleles,B_alleles,alleles,genotypes)\n",
    "\n",
    "    #Calculate additional values of interest from the time series output\n",
    "    total_Nf = hcat(sum.(eachrow(output[\"genotypenumber_pupae_female\"])),sum.(eachrow(output[\"genotypenumber_adult_female\"])))\n",
    "    total_Nf = DataFrame(total_Nf,[\"pupae\",\"adult\"])\n",
    "    relative_Nf = Find_relative_Nf(total_Nf);\n",
    "\n",
    "    #For the zygotes (regardless of reproductive fitness)\n",
    "    genotype_freqzf_store = freq_from_no_df(output[\"genotypenumber_zygote_female\"])\n",
    "    genotype_freqzm_store = freq_from_no_df(output[\"genotypenumber_zygote_male\"])\n",
    "    genotype_freqz_store = (genotype_freqzf_store.+genotype_freqzm_store)./2;\n",
    "    allele_freqz = Find_allele_frequencies_from_genotypes_df(genotypes,genotype_freqz_store,A_alleles,B_alleles);\n",
    "    correlationz = Find_correlation_df(genotypes,alleles,genotype_freqz_store,A_alleles,B_alleles,\"A2\",\"B2\")\n",
    "\n",
    "    #For all pupae (regardless of reproductive fitness)\n",
    "    genotype_freqpf_store = freq_from_no_df(output[\"genotypenumber_pupae_female\"])\n",
    "    genotype_freqpm_store = freq_from_no_df(output[\"genotypenumber_pupae_male\"])\n",
    "    genotype_freqp_store = (genotype_freqpf_store.+genotype_freqpm_store)./2;\n",
    "    allele_freqp = Find_allele_frequencies_from_genotypes_df(genotypes,genotype_freqp_store,A_alleles,B_alleles);\n",
    "    correlationp = Find_correlation_df(genotypes,alleles,genotype_freqp_store,A_alleles,B_alleles,\"A2\",\"B2\")\n",
    "\n",
    "    #For the reproductive adults (regardless of reproductive fitness)\n",
    "    genotype_freqaf_store = freq_from_no_df(output[\"genotypenumber_adult_female\"])\n",
    "    genotype_freqam_store = freq_from_no_df(output[\"genotypenumber_adult_male\"])\n",
    "    genotype_freqa_store = (genotype_freqaf_store.+genotype_freqam_store)./2;\n",
    "    allele_freqa = Find_allele_frequencies_from_genotypes_df(genotypes,genotype_freqa_store,A_alleles,B_alleles);\n",
    "    correlationa = Find_correlation_df(genotypes,alleles,genotype_freqa_store,A_alleles,B_alleles,\"A2\",\"B2\")\n",
    "\n",
    "    #save in a dictionary\n",
    "    data_dict = Dict(\n",
    "        \n",
    "        \"genotypenumber_adult_female\" => output[\"genotypenumber_adult_female\"],\n",
    "        \"genotypenumber_adult_male\" => output[\"genotypenumber_adult_male\"],\n",
    "        \"genotypenumber_pupae_female\" => output[\"genotypenumber_pupae_female\"],\n",
    "        \"genotypenumber_pupae_male\" => output[\"genotypenumber_pupae_male\"],\n",
    "        \"genotypenumber_zygote_female\" => output[\"genotypenumber_zygote_female\"],\n",
    "        \"genotypenumber_zygote_male\" => output[\"genotypenumber_zygote_male\"],\n",
    "        \n",
    "        \"number_females\" => total_Nf,\n",
    "        \"relative_number_females\" => relative_Nf,\n",
    "        \n",
    "        \"genotypefreq_adult_female\" => genotype_freqaf_store,\n",
    "        \"genotypefreq_adult_male\" => genotype_freqam_store,\n",
    "        \"genotypefreq_adult\" => genotype_freqa_store,\n",
    "        \n",
    "        \"genotypefreq_pupae_female\" => genotype_freqpf_store,\n",
    "        \"genotypefreq_pupae_male\" => genotype_freqpm_store,\n",
    "        \"genotypefreq_pupae\" => genotype_freqp_store,\n",
    "        \n",
    "        \"genotypefreq_zygote_female\" => genotype_freqzf_store,\n",
    "        \"genotypefreq_zygote_male\" => genotype_freqzm_store,\n",
    "        \"genotypefreq_zygote\" => genotype_freqz_store,\n",
    "        \n",
    "        \"allelefreq_adult\" => allele_freqa,\n",
    "        \"allelefreq_pupae\" => allele_freqp,\n",
    "        \"allelefreq_zygote\" => allele_freqz,\n",
    "        \n",
    "        \"correlation_adult\" => correlationa,\n",
    "        \"correlation_pupae\" => correlationp,\n",
    "        \"correlation_zygote\" => correlationz\n",
    "    );\n",
    "    return(data_dict)\n",
    "end\n",
    "\n",
    "\"\"\"Calculate the row-wise frequency\"\"\"\n",
    "function freq_from_no_df(genotype_N_store)\n",
    "    genotype_N_store./[sum(row) for row = eachrow(genotype_N_store)]\n",
    "end\n",
    "\n",
    "\"\"\"Calculate the fraction of females relative to generation zero\"\"\"\n",
    "function Find_relative_Nf(Nf_store)\n",
    "    pupae = (Nf_store.pupae)/(Nf_store.pupae[1])\n",
    "    adult = (Nf_store.adult)/(Nf_store.adult[1]);\n",
    "    relative_Nf = DataFrame(pupae = float.(pupae),\n",
    "                    adult=float.(adult))\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Takes in the genotype frequencies of a population at a single time point and \n",
    "calculated the allele frequency.\n",
    "\"\"\"\n",
    "function Find_allele_frequencies_from_genotypes(genotypes,freq_genotype,A_alleles,B_alleles)\n",
    "    \"\"\"Extract the allele frequencies from the genotype frequencies\"\"\"\n",
    "    allele_freqs = Vector{Float64}()\n",
    "    for allele in vcat(A_alleles,B_alleles)\n",
    "        allele_freq = 0\n",
    "        for i in 1:length(genotypes)\n",
    "            if (allele in genotypes[i][:,1]) & (allele in genotypes[i][:,2])\n",
    "                allele_freq = allele_freq+freq_genotype[i]\n",
    "            elseif (allele in genotypes[i][:,1]) | (allele in genotypes[i][:,2])\n",
    "                allele_freq = allele_freq+(freq_genotype[i]/2)\n",
    "            end\n",
    "        end\n",
    "       push!(allele_freqs,allele_freq)\n",
    "    end\n",
    "    return(allele_freqs)\n",
    "end\n",
    "function Find_allele_frequencies_from_genotypes_df(genotypes,genotype_freq_store,A_alleles,B_alleles)\n",
    "    header = Symbol.(vcat(A_alleles,B_alleles))\n",
    "    Make_dataframe([Find_allele_frequencies_from_genotypes(genotypes,row,A_alleles,B_alleles) for row = eachrow(genotype_freq_store)],header)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Takes in the genotype frequencies of a population at a single time point and \n",
    "calculated the frequency of each haploid genotype (chromosome).\n",
    "\"\"\"\n",
    "function Find_chromosome_frequencies_from_genotypes(freq_genotype,genotypes,alleles)\n",
    "    \"\"\"Extract the allele frequencies from the genotype frequencies\"\"\"\n",
    "    allele_freqs = []\n",
    "    for allele in alleles\n",
    "        allele_freq = 0\n",
    "        for i in 1:length(genotypes)\n",
    "            if ((allele == genotypes[i][:,1]) | (allele == genotypes[i][:,2]))\n",
    "                if genotypes[i][:,1] == genotypes[i][:,2]\n",
    "                    allele_freq = allele_freq+(freq_genotype[i])\n",
    "                else\n",
    "                    allele_freq = allele_freq+(freq_genotype[i]/ 2)\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "       append!(allele_freqs,allele_freq)\n",
    "    end\n",
    "    return(allele_freqs)\n",
    "end\n",
    "function Find_chromosome_frequencies_from_genotypes_df(genotype_freq_store,genotypes,alleles)\n",
    "    header = Symbol.(alleles)\n",
    "    Make_dataframe([Find_chromosome_frequencies_from_genotypes(row,genotypes,alleles) for row = eachrow(genotype_freq_store)],header)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Takes in the genotype frequencies of a population at a single time point and \n",
    "calculated the correlation bewteen the constructs a and b.\n",
    "\"\"\"\n",
    "function Find_correlation(genotypes,alleles,genotype_freq,A_alleles,B_alleles,A,B)\n",
    "    \n",
    "    chrom_freq = Find_chromosome_frequencies_from_genotypes(genotype_freq,genotypes,alleles)\n",
    "    ab_freq = chrom_freq[findall(x->x==[A,B],alleles)[1]]\n",
    "    \n",
    "    allele_freq = Find_allele_frequencies_from_genotypes(genotypes,genotype_freq,A_alleles,B_alleles)\n",
    "    AB_alleles = vcat(A_alleles,B_alleles)\n",
    "    a_freq = allele_freq[findall(x->x==A,AB_alleles)[1]]\n",
    "    b_freq = allele_freq[findall(x->x==B,AB_alleles)[1]]\n",
    "    \n",
    "    if (float(b_freq) > 0.0)\n",
    "        cor = (ab_freq - (a_freq*b_freq))/sqrt(float(a_freq*(1-a_freq)*b_freq*(1-b_freq)))\n",
    "    else\n",
    "        cor = 0.0\n",
    "    end\n",
    "end\n",
    "function Find_correlation_df(genotypes,alleles,genotype_freq_store,A_alleles,B_alleles,A,B)\n",
    "    header = [\"Correlation\"]\n",
    "    Make_dataframe([Find_correlation(genotypes,alleles,row,A_alleles,B_alleles,A,B) for row = eachrow(genotype_freq_store)],header)\n",
    "end\n",
    "\n",
    "\"\"\"Processes a matrix into a dataframe\"\"\"\n",
    "function Make_dataframe(store, header)\n",
    "    df = DataFrame(transpose(hcat(store...)))\n",
    "    rename!(df, header)\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "###### Input/Output functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [
     3,
     27
    ],
    "hidden": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Writes output dictionary to file, with each entry of the dictionary being written to a different txt file.\n",
    "\"\"\"\n",
    "function Sim_IO_save(file_prefix,\n",
    "                            params,\n",
    "                            data_dict,\n",
    "                            t)\n",
    "    \n",
    "    #Write to info file the parameters used to run the sumulation\n",
    "    fn = file_prefix*\"_info.txt\"\n",
    "    io = open(fn, \"w\");\n",
    "    write(io,\"sim_length \\t\"*string(t)*\"\\n\")\n",
    "    p_label = keys(params)\n",
    "    for p in p_label\n",
    "        write(io,p*\"\\t\"*string(params[p])*\"\\n\")\n",
    "    end\n",
    "    close(io)\n",
    "    \n",
    "    #Write to file the data according to the dictionary label\n",
    "    for i in [keys(data_dict)...]\n",
    "        CSV.write(file_prefix*\"_\"*i*\".csv\", data_dict[i])\n",
    "    end\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Reads output files and stores in dictionary.\n",
    "\"\"\"\n",
    "function Sim_IO_read(file_prefix,\n",
    "                               data_labels)\n",
    "    data_dict = Dict()\n",
    "    for label in data_labels\n",
    "        data_dict[label] = CSV.read(file_prefix*\"_\"*label*\".csv\",DataFrame)\n",
    "    end\n",
    "    return(data_dict)\n",
    "end"
   ]
  }
 ],
 "metadata": {
  "@webio": {
   "lastCommId": null,
   "lastKernelId": null
  },
  "kernelspec": {
   "display_name": "Julia 1.5.2",
   "language": "julia",
   "name": "julia-1.5"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.5.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
