{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# Baseline double drive simulator\n",
    "\n",
    "23-02-21\n",
    "\n",
    "Katie Willis\n",
    "\n",
    "katie.willis16@imperial.ac.uk\n",
    "\n",
    "___\n",
    "\n",
    "This file contains functions required to simulate the release of a 2 locus gene drive into a single panmictic population and monitor the population through time. We model 2 loci, each with 3 alleles, and therefore 9 haploid genotypes and 45 diploid genotypes. We begin with a matrix containing numbers of individuals of each genotype after transgenic release, monitoring males and females separately. To model processes such as fitness costs, homing, sex ratio bias and recombination during each generation, we multiply (element-wise) the numbers of individuals per genotype by a series of transition matrices for each process, built using user-defined parameters. The output is a dictionary containing time-series of allele and genotype frequencies, numbers of indididuals and the correlation between the two gene drive constructs.\n",
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parameter and matrix definition"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Allele and genotype lists\n",
    "\n",
    "We model 2 loci, each with 3 alleles, and therefore 9 haploid genotypes and 45 diploid genotypes.\n",
    "\n",
    "* First locus: \n",
    "    * A = WT\n",
    "    * a = Transgene\n",
    "    * α = Resistant\n",
    "* Second locus: \n",
    "    * B = Target site \n",
    "    * b = Transgene\n",
    "    * β = Resistant\n",
    "    \n",
    "*Note that labels differ compared to publication, where in the publication a and b refer to resistant loci and α and β refer to the construct."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "code_folding": [
     3,
     51
    ]
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "define_alleles_and_genotypes (generic function with 1 method)"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Defines lists of alleles and genotypes as strings.\n",
    "\"\"\"\n",
    "function define_alleles_and_genotypes()\n",
    "    \n",
    "    A_alleles = ['A','a','α']\n",
    "    B_alleles = ['B','b','β'];\n",
    "    \n",
    "    #Define all A and B allele combinations: Creating haploid genotypes (e.g. AB, Ab etc.)\n",
    "    alleles = []\n",
    "    for A in A_alleles\n",
    "        for B in B_alleles\n",
    "            push!(alleles,(A*B))\n",
    "        end\n",
    "    end;\n",
    "    \n",
    "    #Define all diploid genotypes at the first locus (e.g. AA, Aa etc.)\n",
    "    A_genotypes = []\n",
    "    for i in 1:length(A_alleles)\n",
    "        allele1 = A_alleles[i]\n",
    "        for j in i:length(A_alleles)\n",
    "            allele2 = A_alleles[j]\n",
    "            push!(A_genotypes,(allele1*allele2))\n",
    "        end\n",
    "    end;\n",
    "\n",
    "    #Define all diploid genotypes at the second locus (e.g. BB, Bb etc.)\n",
    "    B_genotypes = []\n",
    "    for i in 1:length(B_alleles)\n",
    "        allele1 = B_alleles[i]\n",
    "        for j in i:length(B_alleles)\n",
    "            allele2 = B_alleles[j]\n",
    "            push!(B_genotypes,(allele1*allele2))\n",
    "        end\n",
    "    end;\n",
    "    \n",
    "    #Define all genotypes for both A and B loci combined (e.g. ABAB, ABAb etc.)\n",
    "    genotypes = []\n",
    "    for i in 1:length(alleles)\n",
    "        allele1 = alleles[i]\n",
    "        for j in i:length(alleles)\n",
    "            allele2 = alleles[j]\n",
    "            push!(genotypes,([allele1 allele2]))\n",
    "        end\n",
    "    end;\n",
    "    \n",
    "    return(alleles,A_alleles,B_alleles,A_genotypes,B_genotypes,genotypes)\n",
    "end\n",
    "\"\"\"\n",
    "Converts lists of alleles and genotypes into lists of symbols for labelling of data structures.\n",
    "\"\"\"\n",
    "function define_symbolic_alleles_and_genotypes(alleles,A_alleles,B_alleles,A_genotypes,B_genotypes,genotypes)\n",
    "    ab, aB, aβ, Ab, AB, Aβ, αb, αB, αβ = symbols(\"ab, aB, aβ, Ab, AB, Aβ, αb, αB, αβ\");\n",
    "    alleles_sym = [Sym(a) for a in alleles]\n",
    "    A_genotypes_sym = [Sym(a) for a in A_genotypes]\n",
    "    B_genotypes_sym = [Sym(a) for a in B_genotypes]\n",
    "    genotypes_sym = [sympify(i[1]*i[2]) for i in genotypes];\n",
    "    return(alleles_sym,A_genotypes_sym,B_genotypes_sym,genotypes_sym)\n",
    "end\n",
    "\n",
    "#alleles and genotype labels are fixed and defined as global variables\n",
    "global alleles,A_alleles,B_alleles,A_genotypes,B_genotypes,genotypes = define_alleles_and_genotypes();\n",
    "global alleles_sym,A_genotypes_sym,B_genotypes_sym,genotypes_sym = define_symbolic_alleles_and_genotypes(alleles,A_alleles,B_alleles,A_genotypes,B_genotypes,genotypes);\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "### Parameters\n",
    "\n",
    "We define a dictionary of all parameters required to run the simulation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "code_folding": [
     12,
     112
    ],
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Assign_baseline_params"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Initiates a dictionary containing all parameters. \n",
    "All are set to zero except:\n",
    "m = 0.5\n",
    "r = 0.5\n",
    "Rm = 6 \n",
    "Nf_eq = 1000 (Number of females or males at equlibrium)\n",
    "OJ = 1/10\n",
    "therefore:\n",
    "f = (2*Rm)/OJ_\n",
    "α = f_*Nf_eq/(Rm-1)\n",
    "\"\"\"\n",
    "function Initiate_baseline_parameters()\n",
    "    \n",
    "    params = Dict{String,Float64}()\n",
    "\n",
    "    #expression costs for each possible construct component\n",
    "    ## for females\n",
    "    params[\"scasf\"] = 0\n",
    "    params[\"hcasf\"] = 0\n",
    "    params[\"sgRNAf\"] = 0\n",
    "    params[\"hgRNAf\"] = 0\n",
    "    params[\"sXSf\"] = 0\n",
    "    params[\"hXSf\"] = 0\n",
    "    ## for males\n",
    "    params[\"scasm\"] = 0\n",
    "    params[\"hcasm\"] = 0\n",
    "    params[\"sgRNAm\"] = 0\n",
    "    params[\"hgRNAm\"] = 0\n",
    "    params[\"sXSm\"] = 0\n",
    "    params[\"hXSm\"] = 0\n",
    "    \n",
    "    #disruption costs\n",
    "    ## for females\n",
    "    params[\"siaf\"] = 0\n",
    "    params[\"hiaf\"] = 0\n",
    "    params[\"sraf\"] = 0\n",
    "    params[\"hraf\"] = 0\n",
    "    params[\"sibf\"] = 0\n",
    "    params[\"hibf\"] = 0\n",
    "    params[\"srbf\"] = 0\n",
    "    params[\"hrbf\"] = 0\n",
    "    params[\"siraf\"] = 0\n",
    "    params[\"sirbf\"] = 0\n",
    "    ## for males\n",
    "    params[\"siam\"] = 0\n",
    "    params[\"hiam\"] = 0\n",
    "    params[\"sram\"] = 0\n",
    "    params[\"hram\"] = 0\n",
    "    params[\"sibm\"] = 0\n",
    "    params[\"hibm\"] = 0\n",
    "    params[\"srbm\"] = 0\n",
    "    params[\"hrbm\"] = 0\n",
    "    params[\"siram\"] = 0\n",
    "    params[\"sirbm\"] = 0\n",
    "    \n",
    "    # Activity costs\n",
    "    ##X-shredder costs\n",
    "    params[\"sXsf\"] = 0\n",
    "    params[\"sXsm\"] = 0\n",
    "    ##Somatic expression costs\n",
    "    params[\"sSaf\"] = 0\n",
    "    params[\"sSam\"] = 0\n",
    "    params[\"sSbf\"] = 0\n",
    "    params[\"sSbm\"] = 0\n",
    "    ##Off target cleavage costs\n",
    "    params[\"sHaf\"] = 0\n",
    "    params[\"sHam\"] = 0\n",
    "    params[\"sHbf\"] = 0\n",
    "    params[\"sHbm\"] = 0\n",
    "\n",
    "    # homing rates\n",
    "    ## in females\n",
    "    params[\"caf\"] = 0\n",
    "    params[\"jaf\"] = 0\n",
    "    params[\"cbf\"] = 0\n",
    "    params[\"jbf\"] = 0\n",
    "    ## in males\n",
    "    params[\"cam\"] = 0\n",
    "    params[\"jam\"] = 0\n",
    "    params[\"cbm\"] = 0\n",
    "    params[\"jbm\"] = 0\n",
    "    \n",
    "    #recombination\n",
    "    params[\"r\"] = 0.5\n",
    "    \n",
    "    #sex ratio bias\n",
    "    params[\"m\"] = 0.5\n",
    "    \n",
    "    #population dynamics \n",
    "    Rm = 6\n",
    "    Nf_eq = 1000.0\n",
    "    params[\"OJ\"] = 1/10\n",
    "    params[\"f\"] = (2*Rm)/params[\"OJ\"]\n",
    "    params[\"α\"] = params[\"f\"]*Nf_eq/(Rm-1)\n",
    "\n",
    "    return(params)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Initiates a parameter set based on our selected baseline parameters and:\n",
    "    - the locus type of A and B (Ntrl, HI, HS)\n",
    "    - homing dependency of A and B (Constitutive, conditional, none)\n",
    "    - XS dependency (A or B)\n",
    "If applied, homing rates are pre-defined according to Kyrou et al 2018\n",
    "c = 0.971\n",
    "j = 0.019\n",
    "if applied, Xshredder rates are pre-defined according to Galizi et al 2014\n",
    "m = 0.95\n",
    "Fitness costs due to locus distruption are pre-defined to 1 based on the locus type\n",
    "Off target cleavage costs are set to 0.01 for both males and females\n",
    "\"\"\"\n",
    "function Assign_baseline_params(LOCUS_A,\n",
    "                                LOCUS_B,\n",
    "                                set_A_homing_dependency,\n",
    "                                set_B_homing_dependency,\n",
    "                                set_XS_dependency\n",
    "                                )\n",
    "    \n",
    "    cleavage_Kyrou2018 = 0.971\n",
    "    joining_Kyrou2018 = 0.019\n",
    "\n",
    "    p = Initiate_baseline_parameters()\n",
    "    \n",
    "    if LOCUS_A == \"HS_rec\"\n",
    "        p[\"siaf\"]=1.0 \n",
    "        p[\"sraf\"]=1.0 \n",
    "        p[\"siraf\"] = 1.0\n",
    "    elseif LOCUS_A == \"HI\"\n",
    "        p[\"siaf\"]=1.0 \n",
    "        p[\"sraf\"]=1.0 \n",
    "        p[\"siraf\"] = 1.0\n",
    "        p[\"hiaf\"]=1.0 \n",
    "        p[\"hraf\"]=1.0\n",
    "    end\n",
    "\n",
    "    if LOCUS_B == \"HS_rec\"\n",
    "        p[\"sibf\"]=1.0 \n",
    "        p[\"srbf\"]=1.0 \n",
    "        p[\"sirbf\"] = 1.0\n",
    "    elseif LOCUS_B == \"HI\"\n",
    "        p[\"sibf\"]=1.0 \n",
    "        p[\"srbf\"]=1.0 \n",
    "        p[\"sirbf\"] = 1.0\n",
    "        p[\"hibf\"]=1.0 \n",
    "        p[\"hrbf\"]=1.0\n",
    "    end\n",
    "\n",
    "    if set_A_homing_dependency != \"none\"\n",
    "        p[\"caf\"] = cleavage_Kyrou2018 # homing \n",
    "        p[\"jaf\"] = joining_Kyrou2018 # homing\n",
    "        p[\"cam\"] = cleavage_Kyrou2018 # homing \n",
    "        p[\"jam\"] = joining_Kyrou2018 # homing\n",
    "        A_homing_dependency = set_A_homing_dependency\n",
    "    elseif set_A_homing_dependency == \"none\"\n",
    "        #Doesnt matter what this is as the homing rates are set to 0.0\n",
    "        A_homing_dependency = \"constitutive\" #re-define for selection of homing matrix\n",
    "    end\n",
    "\n",
    "    if set_B_homing_dependency != \"none\"\n",
    "        p[\"cbf\"] = cleavage_Kyrou2018 # homing \n",
    "        p[\"jbf\"] = joining_Kyrou2018 # homing\n",
    "        p[\"cbm\"] = cleavage_Kyrou2018 # homing \n",
    "        p[\"jbm\"] = joining_Kyrou2018 # homing\n",
    "        B_homing_dependency = set_B_homing_dependency\n",
    "    elseif set_B_homing_dependency == \"none\"\n",
    "        #Doesnt matter what this is as the homing rates are set to 0.0\n",
    "        B_homing_dependency = \"constitutive\" #re-define for selection of homing matrix\n",
    "    end\n",
    "\n",
    "    if set_XS_dependency != \"none\"\n",
    "        p[\"m\"] = 0.95\n",
    "        XS_dependency = set_XS_dependency\n",
    "    elseif set_XS_dependency == \"none\"\n",
    "        p[\"m\"] = 0.5\n",
    "        XS_dependency = \"none\"\n",
    "    end\n",
    "    \n",
    "    #off target cleavage\n",
    "    par = 0.01\n",
    "    p[\"sHaf\"] = par\n",
    "    p[\"sHam\"] = par\n",
    "    p[\"sHbf\"] = par\n",
    "    p[\"sHbm\"] = par\n",
    "    \n",
    "    return(p,A_homing_dependency,B_homing_dependency,XS_dependency)\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "### Fitness\n",
    "\n",
    "Fitness costs are calculated based on 3 different processes. The first two are locus specific, including costs due to locus disruption by either construct insertion or cleavage resistant mutation, and costs due to expression of the construct. The third is due to construct activity, which includes off-target Cas9 cleavage, somatic Cas9 cleavage and X-shredding, each of which may or may not depend on interaction between transcriptional units from different constructs. \n",
    "Fitness costs are caculated multiplicitly: where the homozygous WT fitness cost is standardised to 1 and each process reduces the fitness by a factor.\n",
    "The functions below produce a fitness matrix containing the relative fitness of each genotype compared to the wildtype. This depends on the contents of each construct, the homing dependency (constitutive, conditional or no homing of the construct at each locus) and location of the X-shredder (A or B) if present. \n",
    "Each row of the matrix represents a genotype, the first column contains the relative fitness of females and the second the relative fitness of males for each genotype."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "\n",
    "###### Disruption costs\n",
    "At a single locus, where the fitness of the homoxygous wildtype is standardised to 1, the fitness of individuals homozygous for the construct is $1-{s_I}$; homozygous for the cleavage resisitant allele $1-{s_R}$; heterozygous for the construct allele $1-{h_Is_I}$; heterozygous for the cleavage resistant allele $1-{h_Rs_R}$ or heterozygous carrying both the construct and cleavage resistant allele $1-s_{IR}$, where $s_I$, $s_R$ and $s_{IR}$ are selection coefficients and $h_I$ and $h_R$ are dominance coefficients, each of which differ depending on the locus and sex. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [
     6
    ],
    "hidden": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Makes the fitness matrix for female and male individuals taking into consideration disruption of \n",
    "function at a single locus.\n",
    "The output is a matrix, where each column represents the sex (female, male)\n",
    "and each row represents the genotype of diploid individuals at a single locus e.g. AA, Aa etc.\n",
    "\"\"\"\n",
    "function Get_disruption_fitness(sif::Float64,hif::Float64,srf::Float64,hrf::Float64,sirf::Float64,\n",
    "                                sim::Float64,him::Float64,srm::Float64,hrm::Float64,sirm::Float64)\n",
    "    #genotypes are in the order of A_genotypes and B_genotypes vectors defined as global variables\n",
    "    fitness_matrix = Array{Float64}(undef,(6,2))\n",
    "    fitness_matrix[:,1] = [1,1-(sif*hif),1-(srf*hrf),1-sif,1-sirf,1-srf]\n",
    "    fitness_matrix[:,2] = [1,1-(sim*him),1-(srm*hrm),1-sim,1-sirm,1-srm]\n",
    "    return(fitness_matrix)\n",
    "end\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "\n",
    "###### Expression costs\n",
    "\n",
    "For individuals carrying a construct at either locus, there may be additional costs due to expression of the transcription units present (one or more of Cas9, gRNA or X-shredder). We assume expression of each transcription unit has the same cost when expressed at either locus. Where the fitness of the homozygous is standardised to 1, the fitness is reduced due to construct expression by $\\prod_{k=1}^n (1-s_{k})$ in homozygoyes and $\\prod_{k=1}^n (1-s_{k}h_{k})$ in heterozygotes, where $k$ refers to each component (Cas9, gRNA, XS) in the construct, $s$ is the selection coefficient and $h$ the dominance coefficient, each of which can vary depending on the sex.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [
     5
    ],
    "hidden": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Makes the fitness matrix for female and male individuals considering the contents of the construct at a single locus..\n",
    "The output is a matrix, where each column rperesents the sex (female, male)\n",
    "and each row represents the genotype of a diploid individuals at a single locus e.g. AA, Aa etc.\n",
    "\"\"\"\n",
    "function Get_construct_expression_fitness(construct,p::Dict{String,Float64})\n",
    "    \n",
    "    #Define the dictionary which assignes parameters to names of construct components\n",
    "    fitness_dict = Dict([(\"cas9\", [p[\"scasf\"],p[\"hcasf\"],p[\"scasm\"],p[\"hcasm\"]]), \n",
    "            (\"grna\", [p[\"sgRNAf\"],p[\"hgRNAf\"],p[\"sgRNAm\"],p[\"hgRNAm\"]]), \n",
    "            (\"xs\", [p[\"sXSf\"],p[\"hXSf\"],p[\"sXSm\"],p[\"hXSm\"]])])\n",
    "\n",
    "    #The initial fitness of an empty construct is 1\n",
    "    hom_f = 1\n",
    "    het_f = 1\n",
    "    hom_m = 1\n",
    "    het_m = 1\n",
    "    \n",
    "    #For every component in the construct the fitness is multiplied by 1-cost\n",
    "    for component in construct\n",
    "        hom_f = hom_f*(1-fitness_dict[component][1])\n",
    "        het_f = het_f*(1-(fitness_dict[component][1]*fitness_dict[component][2]))\n",
    "        hom_m = hom_m*(1-fitness_dict[component][3])\n",
    "        het_m = het_m*(1-(fitness_dict[component][3]*fitness_dict[component][4]))\n",
    "    end\n",
    "    \n",
    "    #returns a matrix where each row is a genotype and the two columns refer to females and males. \n",
    "    #genotypes are in the order of A_genotypes and B_genotypes vectors previously defined as global variables\n",
    "    fitness_matrix = Array{Float64}(undef,(length(A_genotypes),2))\n",
    "    fitness_matrix[:,1] = [1,het_f,1,hom_f,1,1]\n",
    "    fitness_matrix[:,2] = [1,het_m,1,hom_m,1,1]\n",
    "    return(fitness_matrix)\n",
    "end\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "\n",
    "###### Activity costs\n",
    "\n",
    "Off-target cleavage reduces fitness by a factor $1 - s_{H}$ in the presence of Cas9 and gRNA and $(1 - s_{H})^2$ in the presence of Cas9 and two different gRNAs (gRNA targeting A and gRNA targeting B). \n",
    "Somatic cleavage reduces fitness by a factor $1-s_{S}$ in the presence of Cas9, gRNA and the target site of gRNA. \n",
    "X-shredding reduces fitness by a factor $1-s_{XS}$ in the presence of XS.\n",
    "Each selection coefficient can vary depending on the sex. \n",
    "Here we assume the costs associated with each activity are dominant, i.e. the presence of one of each transcriptional unit required has the same fitness cost as when there are two.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [
     9
    ],
    "hidden": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Takes in the fitness matrices for each locus (A and B) and the homing/X-shredder dependencies.\n",
    "Calculates the fitness for each 2-locus diploid genotype e.g. ABAB, taking into consideration costs due to \n",
    "1) locus disruption (from single locus fitness matrix input)\n",
    "2) construct expression (from single locus fitness matrix input)\n",
    "3) construct activity (calculated for each locus in this function, based on homing/X-shredder dependencies)\n",
    "Outputs a fitness matrix consisting of two columns (females and males),\n",
    "where each row is a 2-locus diploid genotype e.g. ABAB, ABAb etc.\n",
    "\"\"\"\n",
    "function Get_fitness_matrix(A_fitness_matrix, B_fitness_matrix,A_homing_dependency, B_homing_dependency, XS_dependency,p::Dict{String,Float64})\n",
    "    \n",
    "    #____________________________________________________________\n",
    "    #\n",
    "    #Fitness of genotype (before synergistic costs are considered)\n",
    "    #____________________________________________________________\n",
    "    \n",
    "    #Calculate the fitness of each combined genotype by multiplying the fitness for genotypes at each locus.\n",
    "    fitness_M_store = Vector{Float64}() #male fitness\n",
    "    fitness_F_store = Vector{Float64}() #female fitness\n",
    "\n",
    "    #For every combined genotype e.g. ABAB\n",
    "    for genotype in genotypes\n",
    "        B_fitness = Vector{Float64}()\n",
    "        A_fitness = Vector{Float64}()\n",
    "\n",
    "        #Check what the B genotype is e.g. BB\n",
    "        #and assign B_fitness accordingly:\n",
    "        for i in 1:length(B_genotypes) #iterate through the B genotypes\n",
    "            B_genotype = B_genotypes[i] #Split to get the two alleles\n",
    "            B_genotype = split(B_genotype,\"\")\n",
    "            if ( #Check if both alleles occur in the combined genotype (in any order)\n",
    "                    (occursin(B_genotype[1],genotype[1])) & ((occursin(B_genotype[2],genotype[2])))\n",
    "                    |\n",
    "                    (occursin(B_genotype[1],genotype[2])) & ((occursin(B_genotype[2],genotype[1])))\n",
    "                )\n",
    "                #If they do assign the fitness\n",
    "                B_fitness = B_fitness_matrix[i,:]\n",
    "            end\n",
    "        end\n",
    "\n",
    "        #Check what the A genotype is e.g. AA\n",
    "        #and assign A_fitness accordingly:\n",
    "        for i in 1:length(A_genotypes) \n",
    "            A_genotype = A_genotypes[i] #iterate through the A genotypes\n",
    "            A_genotype = split(A_genotype,\"\") #Split to get the two alleles\n",
    "            if ( #Check if both alleles occur in the combined genotype (in any order)\n",
    "                    (occursin(A_genotype[1],genotype[1])) & ((occursin(A_genotype[2],genotype[2])))\n",
    "                    |\n",
    "                    (occursin(A_genotype[1],genotype[2])) & ((occursin(A_genotype[2],genotype[1])))\n",
    "                )\n",
    "                #If they do assign the fitness\n",
    "                A_fitness = A_fitness_matrix[i,:]\n",
    "            end\n",
    "        end\n",
    "\n",
    "        #Multiply the fitnesses at A and B (Assume no epistatic effects)\n",
    "        combined_fitness = simplify.(A_fitness .* B_fitness)\n",
    "        fitness_F = combined_fitness[1]\n",
    "        fitness_M = combined_fitness[2]\n",
    "        #Store male and female fitnesses independently\n",
    "        push!(fitness_M_store,fitness_M)   \n",
    "        push!(fitness_F_store,fitness_F)   \n",
    "    end\n",
    "    \n",
    "    #____________________________________________________________\n",
    "    #\n",
    "    # Applying activit costs\n",
    "    #____________________________________________________________\n",
    "    \n",
    "    for i in 1:length(genotypes)\n",
    "        genotype_1 = split(genotypes[i][1],\"\")\n",
    "        genotype_2 = split(genotypes[i][2],\"\")\n",
    "\n",
    "        #a) off-target cleavage\n",
    "        \n",
    "        #IF homing at a is condtional on b, a and b are required for off-target cleavage by a\n",
    "        if A_homing_dependency == \"conditional\"\n",
    "            #a and b are required for off-target cleavage by a\n",
    "            if (((genotype_1[1]==\"a\") | (genotype_2[1]==\"a\")) & ((genotype_1[2]==\"b\") | (genotype_2[2]==\"b\")))\n",
    "                fitness_F_store[i] = fitness_F_store[i]*(1-p[\"sHaf\"])\n",
    "                fitness_M_store[i] = fitness_M_store[i]*(1-p[\"sHam\"])\n",
    "            end\n",
    "        elseif A_homing_dependency == \"constitutive\"\n",
    "            #a is required for somatic leakage by a\n",
    "            if ((genotype_1[1]==\"a\") | (genotype_2[1]==\"a\"))\n",
    "                fitness_F_store[i] = fitness_F_store[i]*(1-p[\"sHaf\"])\n",
    "                fitness_M_store[i] = fitness_M_store[i]*(1-p[\"sHam\"])\n",
    "            end\n",
    "        end\n",
    "        if B_homing_dependency == \"conditional\"\n",
    "            #a and b are required for off-target cleavage by b\n",
    "            if (((genotype_1[1]==\"a\") | (genotype_2[1]==\"a\")) & ((genotype_1[2]==\"b\") | (genotype_2[2]==\"b\")))\n",
    "                fitness_F_store[i] = fitness_F_store[i]*(1-p[\"sHbf\"])\n",
    "                fitness_M_store[i] = fitness_M_store[i]*(1-p[\"sHbm\"])\n",
    "            end\n",
    "        elseif B_homing_dependency == \"constitutive\"\n",
    "            #b is required for somatic leakage by b\n",
    "            if ((genotype_1[2]==\"b\") | (genotype_2[2]==\"b\"))\n",
    "                fitness_F_store[i] = fitness_F_store[i]*(1-p[\"sHbf\"])\n",
    "                fitness_M_store[i] = fitness_M_store[i]*(1-p[\"sHbm\"])\n",
    "            end\n",
    "        end \n",
    "       \n",
    "        # b) somatic cleavage\n",
    "        \n",
    "        ## if a and b and WT are present, multiple by 1-se\n",
    "        #This represents somatic leakage of Cas9 leading to cleavage of the target site and some dominant fitness cost\n",
    "        if A_homing_dependency == \"conditional\"\n",
    "            #a, b and A are required for off-target cleavage by a\n",
    "            if ((((genotype_1[1]==\"a\") & (genotype_2[1]==\"A\")) |\n",
    "                ((genotype_1[1]==\"A\") & (genotype_2[1]==\"a\"))) & ((genotype_1[2]==\"b\") | (genotype_2[2]==\"b\")))\n",
    "                fitness_F_store[i] = fitness_F_store[i]*(1-p[\"sSaf\"])\n",
    "                fitness_M_store[i] = fitness_M_store[i]*(1-p[\"sSam\"])\n",
    "            end\n",
    "        elseif A_homing_dependency == \"constitutive\"\n",
    "            #a and A is required for somatic leakage by a\n",
    "            if (((genotype_1[1]==\"a\") & (genotype_2[1]==\"A\")) |\n",
    "                ((genotype_1[1]==\"A\") & (genotype_2[1]==\"a\")))\n",
    "                fitness_F_store[i] = fitness_F_store[i]*(1-p[\"sSaf\"])\n",
    "                fitness_M_store[i] = fitness_M_store[i]*(1-p[\"sSam\"])\n",
    "            end\n",
    "        end\n",
    "\n",
    "        if B_homing_dependency == \"conditional\"\n",
    "            #a, b and B are required for off-target cleavage by b\n",
    "            if ((((genotype_1[2]==\"b\") & (genotype_2[2]==\"B\")) |\n",
    "                ((genotype_1[2]==\"B\") & (genotype_2[2]==\"b\"))) & ((genotype_1[1]==\"a\") | (genotype_2[1]==\"a\")))\n",
    "                fitness_F_store[i] = fitness_F_store[i]*(1-p[\"sSbf\"])\n",
    "                fitness_M_store[i] = fitness_M_store[i]*(1-p[\"sSbm\"])\n",
    "            end\n",
    "        elseif B_homing_dependency == \"constitutive\"\n",
    "            #b and B is required for somatic leakage by b\n",
    "            #a, b and B are required for off-target cleavage by b\n",
    "            if (((genotype_1[2]==\"b\") & (genotype_2[2]==\"B\")) |\n",
    "                ((genotype_1[2]==\"B\") & (genotype_2[2]==\"b\")))\n",
    "                fitness_F_store[i] = fitness_F_store[i]*(1-p[\"sSbf\"])\n",
    "                fitness_M_store[i] = fitness_M_store[i]*(1-p[\"sSbm\"])\n",
    "            end\n",
    "        end\n",
    "        \n",
    "        # c) X-shredder\n",
    "        \n",
    "        if XS_dependency == \"conditional\"\n",
    "            if (((genotype_1[1]==\"a\") | (genotype_2[1]==\"a\")) & ((genotype_1[2]==\"b\") | (genotype_2[2]==\"b\")))\n",
    "                fitness_F_store[i] = fitness_F_store[i]*(1-p[\"sXsf\"])\n",
    "                fitness_M_store[i] = fitness_M_store[i]*(1-p[\"sXsm\"])\n",
    "            end\n",
    "        elseif XS_dependency == \"constitutive_A\"\n",
    "            if ((genotype_1[1]==\"a\") | (genotype_2[1]==\"a\"))\n",
    "                fitness_F_store[i] = fitness_F_store[i]*(1-p[\"sXsf\"])\n",
    "                fitness_M_store[i] = fitness_M_store[i]*(1-p[\"sXsm\"])\n",
    "            end     \n",
    "        elseif XS_dependency == \"constitutive_B\"\n",
    "            if ((genotype_1[2]==\"b\") | (genotype_2[2]==\"b\"))\n",
    "                fitness_F_store[i] = fitness_F_store[i]*(1-p[\"sXsf\"])\n",
    "                fitness_M_store[i] = fitness_M_store[i]*(1-p[\"sXsm\"])\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "     \n",
    "    #Combine male and female fitnesses for each combined genotype into a matrix\n",
    "    fitness_matrix = hcat(fitness_F_store,fitness_M_store);\n",
    "    \n",
    "    return(fitness_matrix)\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "code_folding": [
     22
    ],
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Get_fitness_matrix (generic function with 2 methods)"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Wrapper function to generate the fitness matrix: \n",
    "Input: \n",
    "* 2 lists containing construct components for A and B from \"cas9\",\"grna\" or \"xs\"\n",
    "* Homing dependency of locus A and B from \"conditional\" or \"constitutive\"\n",
    "* X-shredder dependency from \"conditional_A\" or \"conditional_B\"\n",
    "\n",
    "The function calculates the fitness for each 2-locus diploid genotype \n",
    "e.g. ABAB, taking into consideration costs due to \n",
    "1) locus disruption \n",
    "2) construct expression \n",
    "3) construct activity \n",
    "Outputs a fitness matrix consisting of two columns (females and males),\n",
    "where each row is a 2-locus diploid genotype. \n",
    "\n",
    "It is possible to input empty constructs (in the form of an empty vector) to allow for simulation of non-specific configerations \n",
    "of constructs, where activity is determined solely on homing and X-shredder dependencies. \n",
    "If so, the function returns fitness matrix considering costs due to: \n",
    "1) locus disruption\n",
    "2) construct activity\n",
    "where expression costs are considered to be zero.\n",
    "\"\"\"\n",
    "function Build_fitness_matrix(A_construct,B_construct,A_homing_dependency,B_homing_dependency,XS_dependency,p)\n",
    "    \n",
    "    A_fitness_matrix = Get_disruption_fitness(p[\"siaf\"],p[\"hiaf\"],p[\"sraf\"],p[\"hraf\"],p[\"siraf\"],\n",
    "                                            p[\"siam\"],p[\"hiam\"],p[\"sram\"],p[\"hram\"],p[\"siram\"]) .* \n",
    "                    Get_construct_expression_fitness(A_construct,p)\n",
    "\n",
    "    B_fitness_matrix = Get_disruption_fitness(p[\"sibf\"],p[\"hibf\"],p[\"srbf\"],p[\"hrbf\"],p[\"sirbf\"],\n",
    "                                            p[\"sibm\"],p[\"hibm\"],p[\"srbm\"],p[\"hrbm\"],p[\"sirbm\"]) .* \n",
    "                        Get_construct_expression_fitness(B_construct,p)\n",
    "\n",
    "    fitness_matrix = Get_fitness_matrix(A_fitness_matrix, B_fitness_matrix, A_homing_dependency, B_homing_dependency, XS_dependency,p)\n",
    "    return(fitness_matrix)\n",
    "end\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Homing\n",
    "\n",
    "Homing in the germline acts to convert genotypes from one to another before segregation during meiosis.\n",
    "Homing requires the presence of Cas9, gRNA and a target alelle and involves two processes. \n",
    "The first is cleavage, which occurs with probability c. \n",
    "The second is repair which can either occur via NHEJ with probability j, converting the WT to a resistant allale, or HDR with probability 1-j, converting the WT to transgene.\n",
    "The functions below generate a genotype x genotype (i x j) matrix, which contain the proportion of each pre-homed genotype (i, row) converted to each post-homed genotypes (j, column) due to homing.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "code_folding": [
     4,
     32,
     72,
     123,
     187,
     259,
     330
    ]
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Build_homing_matrix"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Selects which matrix to buid depending on the homing dependencies \n",
    "\"conditional\" or \"constitutive\" at locus A or B\n",
    "\"\"\"\n",
    "function Build_homing_matrix(A_homing_dependency,B_homing_dependency,p)\n",
    "    caf,jaf,cbf,jbf = p[\"caf\"],p[\"jaf\"],p[\"cbf\"],p[\"jbf\"]\n",
    "    cam,jam,cbm,jbm = p[\"cam\"],p[\"jam\"],p[\"cbm\"],p[\"jbm\"]\n",
    "    \n",
    "    if ((A_homing_dependency  == \"constitutive\") | (A_homing_dependency  == \"none\"))\n",
    "        if ((B_homing_dependency == \"constitutive\") | (B_homing_dependency  == \"none\"))\n",
    "            homing_matrix_f = Build_homing_matrix_Acons_Bcons(caf,jaf,cbf,jbf)\n",
    "            homing_matrix_m = Build_homing_matrix_Acons_Bcons(cam,jam,cbm,jbm)\n",
    "        elseif B_homing_dependency  == \"conditional\"\n",
    "            homing_matrix_f = Build_homing_matrix_Acons_Bcond(caf,jaf,cbf,jbf)\n",
    "            homing_matrix_m = Build_homing_matrix_Acons_Bcond(cam,jam,cbm,jbm)\n",
    "        end\n",
    "    elseif A_homing_dependency  == \"conditional\"\n",
    "        if ((B_homing_dependency == \"constitutive\") | (B_homing_dependency  == \"none\"))\n",
    "            homing_matrix_f = Build_homing_matrix_Acond_Bcons(caf,jaf,cbf,jbf)\n",
    "            homing_matrix_m = Build_homing_matrix_Acond_Bcons(cam,jam,cbm,jbm)\n",
    "        elseif B_homing_dependency  == \"conditional\"\n",
    "            homing_matrix_f = Build_homing_matrix_Acond_Bcond(caf,jaf,cbf,jbf)\n",
    "            homing_matrix_m = Build_homing_matrix_Acond_Bcond(cam,jam,cbm,jbm)\n",
    "        end\n",
    "    end\n",
    "    return(homing_matrix_f,homing_matrix_m)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Takes in a genotype string and modifies the \"non-construct\" allele to the allele input\n",
    "If one of the alleles (A or B) is unchanged, \"none\" is entered\n",
    "\"\"\"\n",
    "function get_modified_genotype(original_genotype,mod_a_allele,mod_b_allele)\n",
    "\n",
    "    genotype = deepcopy(original_genotype)\n",
    "    genotype1 = split(genotype[1],\"\")\n",
    "    genotype2 = split(genotype[2],\"\")\n",
    "\n",
    "    if genotype1[1]==\"A\"\n",
    "        awt = 1\n",
    "    else \n",
    "        awt = 2\n",
    "    end\n",
    "\n",
    "    if genotype1[2]==\"B\"\n",
    "        bwt = 1\n",
    "    else \n",
    "        bwt = 2\n",
    "    end\n",
    "\n",
    "    x = genotype[awt]\n",
    "    x = split(x,\"\")\n",
    "    if mod_a_allele != \"none\"\n",
    "        x[1] = mod_a_allele\n",
    "    end\n",
    "    genotype[awt] = join(x,\"\")\n",
    "\n",
    "    x = genotype[bwt]\n",
    "    x = split(x,\"\")\n",
    "    if mod_b_allele != \"none\"\n",
    "        x[2] = mod_b_allele\n",
    "    end\n",
    "    genotype[bwt] = join(x,\"\")\n",
    "\n",
    "    return(sort_genotype(genotype))\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Sorts the genotype string so order is maintained: \n",
    "A, a, α\n",
    "B, b, β\n",
    "\"\"\"\n",
    "function sort_genotype(genotype)\n",
    "    \n",
    "    genotype1 = split(genotype[1],\"\")\n",
    "    genotype2 = split(genotype[2],\"\")\n",
    "\n",
    "    first_genotype = genotype1\n",
    "    second_genotype = genotype2\n",
    "    \n",
    "\n",
    "    A_alleles_s = string.(A_alleles)\n",
    "    B_alleles_s = string.(B_alleles)\n",
    "    a1 = findall(x->x==genotype1[1],A_alleles_s )\n",
    "    a2 = findall(x->x==genotype2[1],A_alleles_s )\n",
    "    b1 = findall(x->x==genotype1[2],B_alleles_s )\n",
    "    b2 = findall(x->x==genotype2[2],B_alleles_s )\n",
    "    \n",
    "    if a1 != a2\n",
    "        #sort by a\n",
    "        if a1 == a2\n",
    "            first_genotype = genotype1\n",
    "            second_genotype = genotype2\n",
    "        elseif a1 < a2\n",
    "            first_genotype = genotype1\n",
    "            second_genotype = genotype2\n",
    "        elseif a1 > a2\n",
    "            first_genotype = genotype2\n",
    "            second_genotype = genotype1\n",
    "        end\n",
    "    else\n",
    "        #sort by b\n",
    "        if b1 == b2\n",
    "            first_genotype = genotype1\n",
    "            second_genotype = genotype2\n",
    "        elseif b1 < b2\n",
    "            first_genotype = genotype1\n",
    "            second_genotype = genotype2\n",
    "        elseif b1 > b2\n",
    "            first_genotype = genotype2\n",
    "            second_genotype = genotype1\n",
    "        end\n",
    "    end\n",
    "    genotype = [join(first_genotype) join(second_genotype)]\n",
    "    \n",
    "    return(genotype)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Homing at A and B is constitutive so occurs in the presence of: \n",
    "* a (drive) and A (target)\n",
    "* b (drive) and B (target)\n",
    "\"\"\"\n",
    "function Build_homing_matrix_Acons_Bcons(ca::Float64,\n",
    "                                        ja::Float64,\n",
    "                                        cb::Float64,\n",
    "                                        jb::Float64)\n",
    "\n",
    "    homing = Array{Float64}(undef,length(genotypes),length(genotypes))\n",
    "    fill!(homing, 0.0)\n",
    "\n",
    "    #We begin by assuming no homing, therefore all diagonals are set to 1.0\n",
    "        #This is just so we dont have to assign all genotypes where no homing occurs.\n",
    "    homing[diagind(homing, 0)] .= 1.0\n",
    "\n",
    "    #count = 0\n",
    "    for genotype in genotypes\n",
    "\n",
    "        local i=findall(x->x==genotype,genotypes)\n",
    "        original_genotype = genotype\n",
    "\n",
    "        #count = count+1\n",
    "        genotype1 = split(genotype[1],\"\")\n",
    "        genotype2 = split(genotype[2],\"\")\n",
    "\n",
    "        #IF a and A is present AND b and B is present\n",
    "        if ((((genotype1[1]==\"A\") | (genotype2[1]==\"A\")) & ((genotype1[1]==\"a\") | (genotype2[1]==\"a\")))\n",
    "            &\n",
    "            (((genotype1[2]==\"B\") | (genotype2[2]==\"B\")) & ((genotype1[2]==\"b\") | (genotype2[2]==\"b\"))))\n",
    "\n",
    "            homing[i,i] = [(1-ca)*(1-cb)] #No change\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"a\",\"b\")],genotypes)] = [(ca*(1-ja))*(cb*(1-jb))] #Both homing events occur in either order\n",
    "\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"a\",\"none\")],genotypes)] = [(ca*(1-ja))*(1-cb)] #Homing of A only\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"a\",\"β\")],genotypes)] = [(ca*(1-ja))*(cb*jb)] #homing of A and mutation of B\n",
    "\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"none\",\"b\")],genotypes)] = [(cb*(1-jb))*(1-ca)] #Homing of B only\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"α\",\"b\")],genotypes)] = [(cb*(1-jb))*(ca*ja)] #homing of B and mutation of A\n",
    "\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"α\",\"β\")],genotypes)] = [(cb*jb)*(ca*ja)] #mutation of B and A \n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"α\",\"none\")],genotypes)] = [(ca*ja)*(1-cb)] #Mutation of A only\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"none\",\"β\")],genotypes)] = [(cb*jb)*(1-ca)] #Mutation of B only\n",
    "\n",
    "        #IF only a and A is present \n",
    "        elseif (((genotype1[1]==\"A\") | (genotype2[1]==\"A\")) & ((genotype1[1]==\"a\") | (genotype2[1]==\"a\")))\n",
    "\n",
    "            homing[i,i] = [1.0-ca] #No change\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"a\",\"none\")],genotypes)] = [ca*(1-ja)] #Homing of A only\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"α\",\"none\")],genotypes)] = [ca*ja] #Mutation of A only\n",
    "\n",
    "        #IF only b and B is present \n",
    "        elseif (((genotype1[2]==\"B\") | (genotype2[2]==\"B\")) & ((genotype1[2]==\"b\") | (genotype2[2]==\"b\")))\n",
    "\n",
    "            homing[i,i] = [1.0-cb] #No change\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"none\",\"b\")],genotypes)] = [cb*(1-jb)] #Homing of B only\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"none\",\"β\")],genotypes)] = [cb*jb] #Mutation of B only\n",
    "\n",
    "        end\n",
    "    end\n",
    "    return(homing)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Homing at constitutive at A and conditional at B, so occurs in the presence of:  \n",
    "* a (drive) and A (target)\n",
    "* b (drive) and B (target) and a (drive)\n",
    "\"\"\"\n",
    "function Build_homing_matrix_Acons_Bcond(ca::Float64,\n",
    "                                        ja::Float64,\n",
    "                                        cb::Float64,\n",
    "                                        jb::Float64)\n",
    "    \n",
    "    homing = Array{Float64}(undef,length(genotypes),length(genotypes))\n",
    "    fill!(homing, 0.0)\n",
    "\n",
    "    #We begin by assuming no homing, therefore all diagonals are set to 1.0\n",
    "        #This is just so we dont have to assign all genotypes where no homing occurs.\n",
    "    homing[diagind(homing, 0)] .= 1.0\n",
    "\n",
    "    #count = 0\n",
    "    for genotype in genotypes\n",
    "\n",
    "        local i=findall(x->x==genotype,genotypes)\n",
    "        original_genotype = genotype\n",
    "\n",
    "        #count = count+1\n",
    "        genotype1 = split(genotype[1],\"\")\n",
    "        genotype2 = split(genotype[2],\"\")\n",
    "\n",
    "        #IF a and A is present AND b and B is present \n",
    "        #Since A is constitutive A drives, since B is conditional and there is an 'a' present B drives()\n",
    "        if ((((genotype1[1]==\"A\") | (genotype2[1]==\"A\")) & ((genotype1[1]==\"a\") | (genotype2[1]==\"a\")))\n",
    "            &\n",
    "            (((genotype1[2]==\"B\") | (genotype2[2]==\"B\")) & ((genotype1[2]==\"b\") | (genotype2[2]==\"b\"))))\n",
    "\n",
    "            homing[i,i] = [(1-ca)*(1-cb)] #No change\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"a\",\"b\")],genotypes)] = [(ca*(1-ja))*(cb*(1-jb))] #Both homing events occur in either order\n",
    "\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"a\",\"none\")],genotypes)] = [(ca*(1-ja))*(1-cb)] #Homing of A only\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"a\",\"β\")],genotypes)] = [(ca*(1-ja))*(cb*jb)] #homing of A and mutation of B\n",
    "\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"none\",\"b\")],genotypes)] = [(cb*(1-jb))*(1-ca)] #Homing of B only\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"α\",\"b\")],genotypes)] = [(cb*(1-jb))*ca*ja] #homing of B and mutation of A into non-functional allele\n",
    "\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"α\",\"β\")],genotypes)] = [cb*jb*ca*ja] #mutation of B and A into non-functional allele\n",
    "\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"α\",\"none\")],genotypes)] = [(ca*ja)*(1-cb)] #Mutation of A only into non-functional allele\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"none\",\"β\")],genotypes)] = [(cb*jb)*(1-ca)] #Mutation of B only\n",
    "\n",
    "            #println(count,\" A and B \",genotype)\n",
    "\n",
    "        #IF only a and A is present (constitutive homing)\n",
    "        elseif (((genotype1[1]==\"A\") | (genotype2[1]==\"A\")) & ((genotype1[1]==\"a\") | (genotype2[1]==\"a\")))\n",
    "\n",
    "            homing[i,i] = [1.0-ca] #No change\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"a\",\"none\")],genotypes)] = [ca*(1-ja)] #Homing of A only\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"α\",\"none\")],genotypes)] = [ca*ja] #Mutation of A only into non-functional allele\n",
    "\n",
    "            #println(count,\" A \",genotype)\n",
    "\n",
    "        #IF only b and B is present and a least one a (to allow for conditional homing)\n",
    "        elseif (((genotype1[2]==\"B\") | (genotype2[2]==\"B\")) & ((genotype1[2]==\"b\") | (genotype2[2]==\"b\")) & ((genotype1[1]==\"a\") | (genotype2[1]==\"a\")))\n",
    "\n",
    "            homing[i,i] = [1.0-cb] #No change\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"none\",\"b\")],genotypes)] = [cb*(1-jb)] #Homing of B only\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"none\",\"β\")],genotypes)] = [cb*jb] #Mutation of B only\n",
    "\n",
    "            #println(count,\" B \",genotype)\n",
    "        end\n",
    "    end\n",
    "\n",
    "    return(homing)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Homing at conditional at A and constitutive at B, so occurs in the presence of:  \n",
    "* a (drive) and A (target) and b (drive)\n",
    "* b (drive) and B (target)\n",
    "\"\"\"\n",
    "function Build_homing_matrix_Acond_Bcons(ca::Float64,\n",
    "                                        ja::Float64,\n",
    "                                        cb::Float64,\n",
    "                                        jb::Float64)\n",
    "    \n",
    "    homing = Array{Float64}(undef,length(genotypes),length(genotypes))\n",
    "    fill!(homing, 0.0)\n",
    "\n",
    "    #We begin by assuming no homing, therefore all diagonals are set to 1.0\n",
    "        #This is just so we dont have to assign all genotypes where no homing occurs.\n",
    "    homing[diagind(homing, 0)] .= 1.0\n",
    "\n",
    "    #count = 0\n",
    "    for genotype in genotypes\n",
    "\n",
    "        local i=findall(x->x==genotype,genotypes)\n",
    "        original_genotype = genotype\n",
    "\n",
    "        #count = count+1\n",
    "        genotype1 = split(genotype[1],\"\")\n",
    "        genotype2 = split(genotype[2],\"\")\n",
    "\n",
    "        #IF a and A is present AND b and B is present \n",
    "        #Since B is constitutive B drives, since A is conditional and there is an 'b' present A drives()\n",
    "        if ((((genotype1[1]==\"A\") | (genotype2[1]==\"A\")) & ((genotype1[1]==\"a\") | (genotype2[1]==\"a\")))\n",
    "            &\n",
    "            (((genotype1[2]==\"B\") | (genotype2[2]==\"B\")) & ((genotype1[2]==\"b\") | (genotype2[2]==\"b\"))))\n",
    "\n",
    "            homing[i,i] = [(1-ca)*(1-cb)] #No change\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"a\",\"b\")],genotypes)] = [(ca*(1-ja))*(cb*(1-jb))] #Both homing events occur in either order\n",
    "\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"a\",\"none\")],genotypes)] = [(ca*(1-ja))*(1-cb)] #Homing of A only\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"a\",\"β\")],genotypes)] = [(ca*(1-ja))*(cb*jb)] #homing of A and mutation of B\n",
    "\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"none\",\"b\")],genotypes)] = [(cb*(1-jb))*(1-ca)] #Homing of B only\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"α\",\"b\")],genotypes)] = [(cb*(1-jb))*(ca*ja)] #homing of B and mutation of A into non-functional allele\n",
    "\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"α\",\"β\")],genotypes)] = [(cb*jb)*(ca*ja)] #mutation of B and A into non-functional allele\n",
    "\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"α\",\"none\")],genotypes)] = [(ca*ja)*(1-cb)] #Mutation of A only into non-functional allele\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"none\",\"β\")],genotypes)] = [(cb*jb)*(1-ca)] #Mutation of B only\n",
    "\n",
    "            #println(count,\" A and B \",genotype)\n",
    "\n",
    "        #IF only b and B is present (constitutive homing)\n",
    "        elseif (((genotype1[2]==\"B\") | (genotype2[2]==\"B\")) & ((genotype1[2]==\"b\") | (genotype2[2]==\"b\")))\n",
    "\n",
    "            homing[i,i] = [1.0-cb] #No change\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"none\",\"b\")],genotypes)] = [cb*(1-jb)] #Homing of B only\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"none\",\"β\")],genotypes)] = [cb*jb] #Mutation of B only\n",
    "\n",
    "            #println(count,\" A \",genotype)\n",
    "\n",
    "        #IF only a and A is present and at least one b (to allow for conditional homing)\n",
    "        elseif (((genotype1[1]==\"A\") | (genotype2[1]==\"A\")) & ((genotype1[1]==\"a\") | (genotype2[1]==\"a\")) & ((genotype1[2]==\"b\") | (genotype2[2]==\"b\")))\n",
    "\n",
    "            homing[i,i] = [1.0-ca] #No change\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"a\",\"none\")],genotypes)] = [ca*(1-ja)] #Homing of A only\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"α\",\"none\")],genotypes)] = [ca*ja] #Mutation of A only into non-functional allele\n",
    "\n",
    "            #println(count,\" B \",genotype)\n",
    "        end\n",
    "    end\n",
    "    return(homing)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Homing at A and B is conditional, so occurs in the presence of:  \n",
    "* a (drive) and A (target) and b (drive)\n",
    "* b (drive) and B (target) and a (drive)\n",
    "\"\"\"\n",
    "function Build_homing_matrix_Acond_Bcond(ca::Float64,\n",
    "                                        ja::Float64,\n",
    "                                        cb::Float64,\n",
    "                                        jb::Float64)\n",
    "    \n",
    "    homing = Array{Float64}(undef,length(genotypes),length(genotypes))\n",
    "    fill!(homing, 0.0)\n",
    "\n",
    "    #We begin by assuming no homing, therefore all diagonals are set to 1.0\n",
    "        #This is just so we dont have to assign all genotypes where no homing occurs.\n",
    "    homing[diagind(homing, 0)] .= 1.0\n",
    "\n",
    "    #count = 0\n",
    "    for genotype in genotypes\n",
    "\n",
    "        local i=findall(x->x==genotype,genotypes)\n",
    "        original_genotype = genotype\n",
    "\n",
    "        #count = count+1\n",
    "        genotype1 = split(genotype[1],\"\")\n",
    "        genotype2 = split(genotype[2],\"\")\n",
    "\n",
    "        #IF a and A is present AND b and B is present \n",
    "        #Since A is conditional and 'b' is present A drives, since B is conditional and there is an 'a' present B drives()\n",
    "        if ((((genotype1[1]==\"A\") | (genotype2[1]==\"A\")) & ((genotype1[1]==\"a\") | (genotype2[1]==\"a\")))\n",
    "            &\n",
    "            (((genotype1[2]==\"B\") | (genotype2[2]==\"B\")) & ((genotype1[2]==\"b\") | (genotype2[2]==\"b\"))))\n",
    "\n",
    "            homing[i,i] = [(1-ca)*(1-cb)] #No change\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"a\",\"b\")],genotypes)] = [(ca*(1-ja))*(cb*(1-jb))] #Both homing events occur in either order\n",
    "\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"a\",\"none\")],genotypes)] = [(ca*(1-ja))*(1-cb)] #Homing of A only\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"a\",\"β\")],genotypes)] = [(ca*(1-ja))*(cb*jb)] #homing of A and mutation of B\n",
    "\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"none\",\"b\")],genotypes)] = [(cb*(1-jb))*(1-ca)] #Homing of B only\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"α\",\"b\")],genotypes)] = [(cb*(1-jb))*(ca*ja)] #homing of B and mutation of A into non-functional allele\n",
    "\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"α\",\"β\")],genotypes)] = [cb*jb*(ca*ja)] #mutation of B and A into non-functional allele\n",
    "\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"α\",\"none\")],genotypes)] = [ca*ja*(1-cb)] #Mutation of A only into non-functional allele\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"none\",\"β\")],genotypes)] = [(cb*jb)*(1-ca)] #Mutation of B only\n",
    "\n",
    "            #println(count,\" A and B \",genotype)\n",
    "\n",
    "        #IF only a, A and at least one b is present (conditional homing of a)\n",
    "        elseif (((genotype1[1]==\"A\") | (genotype2[1]==\"A\")) & ((genotype1[1]==\"a\") | (genotype2[1]==\"a\")) & ((genotype1[2]==\"b\") | (genotype2[2]==\"b\")))\n",
    "\n",
    "            homing[i,i] = [1.0-ca] #No change\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"a\",\"none\")],genotypes)] = [ca*(1-ja)] #Homing of A only\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"α\",\"none\")],genotypes)] = [ca*ja] #Mutation of A only into non-functional allele\n",
    "\n",
    "            #println(count,\" A \",genotype)\n",
    "\n",
    "        #IF only b, B and at least one a is present (conditional homing of b)\n",
    "        elseif (((genotype1[2]==\"B\") | (genotype2[2]==\"B\")) & ((genotype1[2]==\"b\") | (genotype2[2]==\"b\")) & ((genotype1[1]==\"a\") | (genotype2[1]==\"a\")))\n",
    "\n",
    "            homing[i,i] = [1.0-cb] #No change\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"none\",\"b\")],genotypes)] = [cb*(1-jb)] #Homing of B only\n",
    "            homing[i,findall(x->[x]==[get_modified_genotype(original_genotype,\"none\",\"β\")],genotypes)] = [cb*jb] #Mutation of B only\n",
    "\n",
    "            #println(count,\" B \",genotype)\n",
    "        end\n",
    "    end\n",
    "    return(homing)\n",
    "end\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sex bias\n",
    "\n",
    "The probability of a male producing an X- or Y-bearing gamete depends on the presence of an X-shredder. The functions below generate a matrix containing the proportion of X- (columns 1) or Y- (column 2) bearing sperm depending on the genotype (row). m is the proporion of Y-bearing sperm produced for a male carrying at least one copy of the X-shredder. We assume that x-shredding function is dominant, thus heterozygotes and homozygotes for the xshredder create the same sex bias. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "code_folding": [
     4,
     26,
     42,
     77
    ]
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Build_sex_ratio_matrix"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Selects which matrix to generate depending on the sex ratio dependencies \n",
    "\"constitutive_A\" or \"constitutive_B\". \n",
    "\"\"\"\n",
    "function Build_sex_ratio_matrix(XS_dependency,p)\n",
    "    m=p[\"m\"]\n",
    "    ## Pick the male bias matrix according to the XS dependency\n",
    "    sex_ratio_matrix = []\n",
    "    if XS_dependency == \"none\"\n",
    "        ## If none, use constitutive_A as default matrix and ensure m = 0.5\n",
    "        sex_ratio_matrix = Build_sexratio_matrix_constitutive_A(m)\n",
    "    elseif XS_dependency  == \"constitutive_A\"\n",
    "        sex_ratio_matrix = Build_sexratio_matrix_constitutive_A(m)\n",
    "    elseif XS_dependency  == \"constitutive_B\"\n",
    "        sex_ratio_matrix = Build_sexratio_matrix_constitutive_B(m)\n",
    "    elseif XS_dependency  == \"conditional\"\n",
    "        sex_ratio_matrix = Build_sexratio_matrix_conditional(m)\n",
    "    else\n",
    "        print(\"Model not available\")\n",
    "    end\n",
    "    return(sex_ratio_matrix)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Defines the sex ratio for a single locus\n",
    "\"\"\"\n",
    "function define_locus_sex_ratio_matrix(m)\n",
    "    #order is as defined by the A_genotypes and B_genotypes\n",
    "    #each construct has the ability to induce a sex ratio\n",
    "    locus_sex_ratio = [\n",
    "        0.5 0.5 ; \n",
    "        1-m m;\n",
    "        0.5 0.5;\n",
    "        1-m m;  \n",
    "        1-m m;\n",
    "        0.5 0.5]\n",
    "    return(locus_sex_ratio)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "sex distortion is dependent on the B construct\n",
    "\"\"\"\n",
    "function Build_sexratio_matrix_constitutive_B(m::Float64)\n",
    "    \n",
    "    locus_sex_ratio = define_locus_sex_ratio_matrix(m)\n",
    "    \n",
    "    #Calculate the sex ratio of each genotype\n",
    "    sex_ratio_storeM = Vector{Float64}()\n",
    "    sex_ratio_storeF = Vector{Float64}()\n",
    "    #For each combined genotype e.g. ABAB\n",
    "    for genotype in genotypes\n",
    "        sex_ratio = []\n",
    "        #Check what the B genotype is and assign sex ratio accordingly:\n",
    "        for i in 1:length(B_genotypes) #For each B genotype\n",
    "            B_genotype = B_genotypes[i]\n",
    "            B_genotype = split(B_genotype,\"\") #Split the B genotype into alleles e.g. B, B\n",
    "            if ( #If both alleles occur in the combined genotype\n",
    "                    (occursin(B_genotype[1],genotype[1])) & ((occursin(B_genotype[2],genotype[2])))\n",
    "                    |\n",
    "                    (occursin(B_genotype[1],genotype[2])) & ((occursin(B_genotype[2],genotype[1])))\n",
    "                )\n",
    "                #Assign sec ratio of that B genotype\n",
    "                sex_ratio = locus_sex_ratio[i,:] #WT \n",
    "            end\n",
    "        end\n",
    "        push!(sex_ratio_storeF,sex_ratio[1])   \n",
    "        push!(sex_ratio_storeM,sex_ratio[2])   \n",
    "\n",
    "    end\n",
    "\n",
    "    sex_ratio_matrix_constitutive_B = hcat(sex_ratio_storeF,sex_ratio_storeM);\n",
    "    return(sex_ratio_matrix_constitutive_B)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "sex distortion is dependent on the A construct\n",
    "\"\"\"\n",
    "function Build_sexratio_matrix_constitutive_A(m::Float64)\n",
    "    locus_sex_ratio = define_locus_sex_ratio_matrix(m)\n",
    "    \n",
    "    #Calculate the sex ratio of each genotype\n",
    "    sex_ratio_storeM = Vector{Float64}()\n",
    "    sex_ratio_storeF = Vector{Float64}()\n",
    "    #For each combined genotype e.g. ABAB\n",
    "    for genotype in genotypes\n",
    "        sex_ratio = []\n",
    "        #Check what the A genotype is and assign sex ratio accordingly:\n",
    "        for i in 1:length(A_genotypes) #For each B genotype\n",
    "            A_genotype = A_genotypes[i]\n",
    "            A_genotype = split(A_genotype,\"\") #Split the A genotype into alleles e.g. A, A\n",
    "            if ( #If both alleles occur in the combined genotype\n",
    "                    (occursin(A_genotype[1],genotype[1])) & ((occursin(A_genotype[2],genotype[2])))\n",
    "                    |\n",
    "                    (occursin(A_genotype[1],genotype[2])) & ((occursin(A_genotype[2],genotype[1])))\n",
    "                )\n",
    "                #Assign sec ratio of that A genotype\n",
    "                sex_ratio = locus_sex_ratio[i,:] #WT \n",
    "            end\n",
    "        end\n",
    "        push!(sex_ratio_storeF,sex_ratio[1])   \n",
    "        push!(sex_ratio_storeM,sex_ratio[2])   \n",
    "\n",
    "    end\n",
    "\n",
    "    sex_ratio_matrix_constitutive_A = hcat([sex_ratio_storeF,sex_ratio_storeM]...);\n",
    "    return(sex_ratio_matrix_constitutive_A)\n",
    "end\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "### Recombination\n",
    "Recombination can occur between the two loci with probability r during gamete production. \n",
    "The function generates a matrix containing the proportion of each gamete (column) produced from each genotype (row), \n",
    "where each row refers to a diploid genotype (individual) and each column to a haploid genotype (gamete) at the autosomal loci only.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "code_folding": [
     6
    ],
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Build_recombination_matrix"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Calculate the probability of each genotype generating each gamete (allele)\n",
    "taking into consideration the recombination probability r between \n",
    "A and B loci.\n",
    "Returns matrix of the proportion of each gamete (columns) prouced by each parental genotype (row)\n",
    "\"\"\"\n",
    "function Build_recombination_matrix(p::Dict{String,Float64})\n",
    "    \n",
    "    r=p[\"r\"]\n",
    "\n",
    "    recombination_matrix = Array{Float64}(undef,(length(genotypes),9))\n",
    "    #For each combined genotype\n",
    "    for a in enumerate(genotypes)\n",
    "        local genotype = a[2]\n",
    "        local i = a[1]\n",
    "        #Initiate a row of the matrix\n",
    "        recombination_vec = [Sym(0.0) Sym(0.0) Sym(0.0) Sym(0.0) Sym(0.0) Sym(0.0) Sym(0.0) Sym(0.0) Sym(0.0)]\n",
    "\n",
    "        #Screen whether each allele (gamete) is present in each genotype\n",
    "        nonrecombinant_mask = [a in genotype for a in alleles]\n",
    "        #if 2 alleles are present (the genotype is heterozygous)\n",
    "        if sum(nonrecombinant_mask) == 2\n",
    "            #Assign the 2 matched alleles non-recombinant inheritence (1-r)/2\n",
    "            recombination_vec[nonrecombinant_mask] = collect([(1-r)/2, (1-r)/2])\n",
    "            #Assign the rest of the gametes that are not produced and inheritance of 0\n",
    "            recombination_vec[nonrecombinant_mask .== false] = [0,0,0,0,0,0,0]\n",
    "            #if 1 allele is present (the genotype is homozygous)\n",
    "            elseif sum(nonrecombinant_mask) == 1\n",
    "                #Assign the 1 matched allele non-recombinant inheritance (1-r)\n",
    "                recombination_vec[nonrecombinant_mask] = collect([(1-r)])\n",
    "                #Assign the rest of the gametes that are not produced and inheritance of 0\n",
    "                recombination_vec[nonrecombinant_mask .== false] = [0,0,0,0,0,0,0,0]\n",
    "        end\n",
    "        #Recombine the genotype\n",
    "        A1 = split(genotype[1],\"\")\n",
    "        A2 = split(genotype[2],\"\")\n",
    "        genotype_recomb = [A1[1]*A2[2],A2[1]*A1[2]]\n",
    "        #Screen whether each allele (gamete) is present in each genotype\n",
    "        recombinant_mask = [a in genotype_recomb for a in alleles]\n",
    "        #if 2 alleles are present (the genotype is heterozygous)\n",
    "        if sum(recombinant_mask) == 2\n",
    "            #Assign the 2 matched alleles recombinant inheritence r/2\n",
    "            recombination_vec[recombinant_mask] = collect(recombination_vec[recombinant_mask]+[r/2,r/2])\n",
    "            #if 1 allele is present (the genotype is homozygous)\n",
    "            elseif sum(recombinant_mask) == 1\n",
    "                #Assign the 1 matched alleles recombinant r\n",
    "                recombination_vec[recombinant_mask] = collect(recombination_vec[recombinant_mask]+[r])\n",
    "        end\n",
    "        recombination_matrix[i,:]=recombination_vec\n",
    "    end\n",
    "    return(recombination_matrix)\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "code_folding": [
     6,
     51
    ]
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Simulation_wrapper_separate_release"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Defines starting population and matrices,\n",
    "runs simulation\n",
    "and calculates additional metrics from output.\n",
    "\"\"\"\n",
    "function Simulation_wrapper(params,\n",
    "                            A_construct,\n",
    "                            B_construct,\n",
    "                            A_homing_dependency,\n",
    "                            B_homing_dependency,\n",
    "                            XS_dependency,\n",
    "                            fitness_effect,\n",
    "                            transgenic_genotype,\n",
    "                            release_freq, \n",
    "                            resistant_allele, \n",
    "                            resistant_freq,\n",
    "                            t)\n",
    "    \n",
    "    #Define starting population\n",
    "    starting_population_post_release = release_transgenic_males(\n",
    "        Initiate_prerelease_population_3lifestages(resistant_allele, resistant_freq, params),\n",
    "        Symbol(transgenic_genotype),\n",
    "        release_freq)\n",
    "\n",
    "    #Define matrices\n",
    "    matrices = make_matrices(params,\n",
    "                    A_construct,\n",
    "                    B_construct,\n",
    "                    A_homing_dependency,\n",
    "                    B_homing_dependency,\n",
    "                    XS_dependency);\n",
    "    \n",
    "    #simulate time series\n",
    "    output = Simulate_timeseries(\n",
    "                params,\n",
    "                matrices,\n",
    "                fitness_effect,\n",
    "                t,\n",
    "                starting_population_post_release)\n",
    "    \n",
    "    #process simulation\n",
    "    data_dict = Simulation_processing_wrapper(output)\n",
    "\n",
    "    return(data_dict)\n",
    "end\n",
    "\"\"\"\n",
    "Defines starting population (with constructs released in separate males) and matrices, \n",
    "runs simulation and \n",
    "calculates additional metrics from output.\n",
    "\"\"\"\n",
    "function Simulation_wrapper_separate_release(params,\n",
    "                            A_construct,\n",
    "                            B_construct,\n",
    "                            A_homing_dependency,\n",
    "                            B_homing_dependency,\n",
    "                            XS_dependency,\n",
    "                            fitness_effect,\n",
    "                            transgenic_genotype1,\n",
    "                            transgenic_genotype2,\n",
    "                            release_freq1,\n",
    "                            release_freq2, \n",
    "                            resistant_allele, \n",
    "                            resistant_freq,\n",
    "                            t)\n",
    "    \n",
    "    #Define starting population\n",
    "    starting_population_post_release = release_2transgenic_males(\n",
    "        Initiate_prerelease_population_3lifestages(resistant_allele, resistant_freq, params),\n",
    "        Symbol(transgenic_genotype1),\n",
    "        Symbol(transgenic_genotype2),\n",
    "        release_freq1,\n",
    "        release_freq2)\n",
    "\n",
    "    #Define matrices\n",
    "    matrices = make_matrices(params,\n",
    "                    A_construct,\n",
    "                    B_construct,\n",
    "                    A_homing_dependency,\n",
    "                    B_homing_dependency,\n",
    "                    XS_dependency);\n",
    "    \n",
    "    #simulate time series\n",
    "    output = Simulate_timeseries(\n",
    "                params,\n",
    "                matrices,\n",
    "                fitness_effect,\n",
    "                t,\n",
    "                starting_population_post_release)\n",
    "    \n",
    "    #process simulation\n",
    "    data_dict = Simulation_processing_wrapper(output)\n",
    "\n",
    "    return(data_dict)\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "###### Defining the population upon release\n",
    "\n",
    "We define a 45x6 matrix of numbers of individuals of each genotype after transgenic release. Each row refers to a genotype and each column to the following catagories or organism: Female zygotes, males zygotes, female pupae, male pupae, female adults and male adults. The pre-release population may contain only WT individuals (ABAB) or may contain a resistant allele, assumed to be at Hardy-Weinberg equilibrium at a user-define allele frequency. We provide functions to simulate the release of either a single type of transgenic male, or separate releases of males with two different genotypes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "code_folding": [
     3,
     17,
     58,
     90,
     105
    ]
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "release_2transgenic_males (generic function with 1 method)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Initiates a dictionary containing the number of individuals for each genotype, each of which has an initial value of 0. \n",
    "\"\"\"\n",
    "function Initiate_genotype_freq()\n",
    "    freq_genotype = Dict()\n",
    "    for i in 1:length(genotypes_sym)\n",
    "       freq_genotype[string(genotypes_sym[i])] = 0.0\n",
    "    end\n",
    "    return(freq_genotype)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Returns 2 dictionaries containing the numbers of each genotype, one for females and one for males.\n",
    "The total number of males or females is defined by the variable params: OJ, f and α and refers to the equilibrium population size of each population.\n",
    "The population can contain both WT individuals (ABAB) and genotypes containing a resistant allele.\n",
    "The population is assumed to be at hardy weinberg equilibrium, with the resistant allele at a frequency defined by the variable \"resistant_freq\".\n",
    "\"\"\"\n",
    "function Initiate_prerelease_population(resistant_allele, resistant_freq, params)\n",
    "    \n",
    "    #Assuming population is at equilibrium \n",
    "    Rm = (params[\"f\"]*params[\"OJ\"])/2\n",
    "    Nf = params[\"α\"]*(Rm-1)/params[\"f\"]\n",
    "    # notes: \n",
    "    #f=(2*Rm)/OJ \n",
    "    #J=(2*Rm)/f \n",
    "    #a = f*Nf/(Rm-1)\n",
    "    #Rm = (OJ*f)/2\n",
    "    #Nf = a*(Rm-1)/f\n",
    "    \n",
    "    #Define each frequency initially as 0\n",
    "    genotype_Nf = Initiate_genotype_freq()\n",
    "    genotype_Nm = Initiate_genotype_freq()\n",
    "\n",
    "    #homozygous genotype\n",
    "    if resistant_allele ∈ [\"A\",\"a\",\"α\"]\n",
    "        het = \"AB\"*resistant_allele*\"B\"\n",
    "        hom = resistant_allele*\"B\"*resistant_allele*\"B\"\n",
    "    elseif resistant allele ∈ [\"B\",\"b\",\"β\"]\n",
    "        het = \"ABA\"*resistant_allele\n",
    "        hom = \"A\"*resistant_allele*\"A\"*resistant_allele\n",
    "    end\n",
    "\n",
    "    genotype_Nf[\"ABAB\"] = (1-resistant_freq)^2*Nf\n",
    "    genotype_Nf[hom] = resistant_freq^2*Nf\n",
    "    genotype_Nf[het] = 2 * resistant_freq * (1-resistant_freq)*Nf\n",
    "\n",
    "    genotype_Nm[\"ABAB\"] = (1-resistant_freq)^2*Nf\n",
    "    genotype_Nm[hom] = resistant_freq^2*Nf\n",
    "    genotype_Nm[het] = 2 * resistant_freq * (1-resistant_freq)*Nf\n",
    "\n",
    "    return(genotype_Nf,genotype_Nm)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Initiates the prerelease population of adults.\n",
    "Back-calculates the number of zygotes and pupae for a given adult population and parameter values. \n",
    "Returns a (6x45 matrix) of numbers of zygote, pupae and adults for females and males, with column headers labelled by genotype.\n",
    "\"\"\"\n",
    "function Initiate_prerelease_population_3lifestages(resistant_allele, resistant_freq, params)\n",
    "    \n",
    "    genotype_Nf,genotype_Nm = Initiate_prerelease_population(resistant_allele, resistant_freq, params)\n",
    "\n",
    "    starting_population = Array{Float64}(undef,(6,length(genotypes)))\n",
    " \n",
    "    #Modify data type to arrays in the the order of genotypes\n",
    "    starting_population[5,:] = [genotype_Nf[i] for i in string.(genotypes_sym)]\n",
    "    starting_population[6,:] = [genotype_Nm[i] for i in string.(genotypes_sym)]\n",
    "    \n",
    "    #Assuming the population is at equilibrium AND the population is pre-release\n",
    "    #No. zygotes\n",
    "    starting_population[1,:] =  starting_population[5,:] .* params[\"f\"] /2\n",
    "    starting_population[2,:] =  starting_population[6,:] .* params[\"f\"] /2\n",
    "\n",
    "    #No. zygotes after density dependent mortality (pupae)\n",
    "    n_zygotes_sum = sum(starting_population[1,:])+sum(starting_population[2,:])\n",
    "    juvenile_survival_prob = params[\"OJ\"]*(params[\"α\"]/(params[\"α\"]+n_zygotes_sum))\n",
    "    starting_population[3,:]= starting_population[5,:] .* params[\"f\"] /2 * juvenile_survival_prob\n",
    "    starting_population[4,:] = starting_population[6,:] .* params[\"f\"] /2 * juvenile_survival_prob\n",
    "    \n",
    "    N_individuals = DataFrame(starting_population,Symbol.(genotypes_sym))\n",
    "    \n",
    "    return(N_individuals)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Takes in a 6x45 matrix defining the numbers of genotypes for a population pre-release and \n",
    "returns a 45x6 matrix of numbers of each genotype after release of adult males. \n",
    "The frequency and genotype of the release males is defined by the inputs: \n",
    "transgenic_genotype and release_freq.\n",
    "\"\"\"\n",
    "function release_transgenic_males(starting_population,transgenic_genotype,release_freq)\n",
    "    #Here we make the assumption that all released individuals are fully fit\n",
    "    starting_population[6,transgenic_genotype] = sum(starting_population[6,:])*release_freq\n",
    "    \n",
    "    starting_population = transpose(Array(starting_population))\n",
    "    \n",
    "    return(starting_population)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Takes in a 6x45 matrix defining the numbers of genotypes for a population pre-release and \n",
    "returns a 45x6 matrix of numbers of each genotype after release of two different types of adult males. \n",
    "The frequency and genotype of the release males is defined by the inputs: \n",
    "transgenic_genotype1, transgenic_genotype2, release_freq1 and release_freq2\n",
    "\"\"\"\n",
    "function release_2transgenic_males(starting_population,transgenic_genotype1,transgenic_genotype2,release_freq1,release_freq2)\n",
    "    #Here we make the assumption that all released individuals are fully fit\n",
    "    starting_population[6,transgenic_genotype1] = sum(starting_population[6,:])*release_freq1\n",
    "    starting_population[6,transgenic_genotype2] = sum(starting_population[6,:])*release_freq2\n",
    "    \n",
    "    starting_population = transpose(Array(starting_population))\n",
    "    \n",
    "    return(starting_population)\n",
    "end\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "###### Time series simulation\n",
    "\n",
    "Parameter values and information on loci type, homing conditions and the x-shredder location are used to produce matrices (using the functions in the Model Definition section) for applying fitness, homing, sex ratio and recombination processes to the population during each generation.\n",
    "\n",
    "During each generation:\n",
    "* Eggs and sperm are produced, taking into consideration homing, recombination and sex ratio bias. \n",
    "* Zygotes are procuded assuming random mating and that males are not limiting. \n",
    "* Fitness effects are applied at either the zygote or adult stage. \n",
    "* Density dependence is applied at the zygote stage after fitness effects (if applicable to zygotes). \n",
    "* Censusing of both the zygote and adults are taken after any fitness effects are applied.\n",
    "\n",
    "The ouput is a dictionary of time-series for allele or genotype frequencies, numbers of individuals and correlation between constructs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "code_folding": [
     34,
     66,
     95,
     140,
     164
    ]
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "apply_density_dependence (generic function with 1 method)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Builds the matrices to be used based on the construct contents, homing dependencies, XS dependencies and parameters.\n",
    "If construct contents are left empty ([]), expression costs are assumed to be zero, but activity can be defined \n",
    "for a generic construct structure based on homing and xshredder dependencies alone. \n",
    "\"\"\"\n",
    "function make_matrices(p,\n",
    "                        A_construct,\n",
    "                        B_construct,\n",
    "                        A_homing_dependency,\n",
    "                        B_homing_dependency,\n",
    "                        XS_dependency)\n",
    "\n",
    "    fitness_matrix = Build_fitness_matrix(A_construct, B_construct,A_homing_dependency, B_homing_dependency, XS_dependency,p)\n",
    "    homing_matrix_f,homing_matrix_m = Build_homing_matrix(A_homing_dependency,B_homing_dependency,p)\n",
    "    sex_ratio_matrix = Build_sex_ratio_matrix(XS_dependency,p)\n",
    "    recombination_matrix = Build_recombination_matrix(p)\n",
    "\n",
    "    return(fitness_matrix,\n",
    "            homing_matrix_f,\n",
    "            homing_matrix_m,\n",
    "            recombination_matrix,\n",
    "            sex_ratio_matrix)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Takes in the number of females adults, female homing matrix, recombination matrix and parameter values. \n",
    "The number of eggs produced by each genotype is calculated. \n",
    "Homing is applied by multiplying (element-wise) the vector of egg numbers per genotype by the female homing matrix, \n",
    "and summing over the columns. \n",
    "Recombination is applied and the number of each gamete produced per genotype is calculated (45x9 matrix) \n",
    "by multiplying (element-wise) the number of eggs per genotype after homing by the recombination matrix.\n",
    "The matrix is summed over each column to find the number of eggs produced of each gamete genotype, \n",
    "returning a vector of length 9.\n",
    "\"\"\"\n",
    "function make_eggs(genotype_Nf_adult,homing_matrix_f_vals,recombination_matrix_vals,params)\n",
    "\n",
    "    #Calulate the number of eggs produced by each reproducing adult female\n",
    "    neggs_per_genotype = genotype_Nf_adult * params[\"f\"];\n",
    "    \n",
    "    #Homing:\n",
    "    #Calculate the numbers of eggs produced from each genotype after homing\n",
    "    neggs_per_genotype_per_genotype_homing = neggs_per_genotype.*homing_matrix_f_vals\n",
    "    neggs_per_genotype_homing = sum(neggs_per_genotype_per_genotype_homing,dims=1)[:]\n",
    "\n",
    "    #recombination\n",
    "    #Calulate the number of each gamete produced by each genotype after recombination\n",
    "    neggs_per_genotype_per_gamete = neggs_per_genotype_homing .*recombination_matrix_vals;\n",
    "    # Calculate the number of each gamate produced by the female population in total\n",
    "    neggs_per_gamete = sum(neggs_per_genotype_per_gamete,dims=1)[:];\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Takes in the number of male adults, male homing matrix, recombination matrix and sex ratio matrix. \n",
    "Males are assumed to be not limiting, so the number of males is converted to frequency. \n",
    "Homing is applied by multiplying (element-wise) the frequency of males per genotype by the male homing matrix, \n",
    "and summing over the columns. \n",
    "Recomnbination is applied and the frequency of each gamete produced per genotype is calculated (45x9 matrix) \n",
    "by multiplyng (element-wise) the frequency of males per genotype after homing by the recombination matrix. \n",
    "Sex ratio bias is applied by multiplying (element-wise) the 45x9 matrix containing the frequency of each gamete per genotype\n",
    "by each column of the the sex ratio matrix separately, providing 2 matrices, one for the proportion of each Y-bearing gamete\n",
    "genotype per adult male and another for the X-bearing equivalent.\n",
    "The columns of each X-bearing and Y-bearing sperm matrix are summed to find the frequency of \n",
    "each gamete genotype produced.\n",
    "Returns a (9x2) matrix, containing the frequency of X-bearing sperm (column 1) and Y-bearing sperm (column 2)\n",
    "of each gamate genotype (rows).\n",
    "\"\"\"\n",
    "function make_sperm(genotype_Nm_adult,homing_matrix_m_vals,recombination_matrix_vals,sex_ratio_matrix_vals)\n",
    "    \n",
    "    #conver male numbers to frequency. \n",
    "    genotype_freqm_fitness = genotype_Nm_adult./sum(genotype_Nm_adult)\n",
    "\n",
    "    #Homing:\n",
    "    genotype_freqm_per_genotype_homing = genotype_freqm_fitness.*homing_matrix_m_vals\n",
    "    genotype_freqm_homing = sum(genotype_freqm_per_genotype_homing,dims=1)[:]\n",
    "\n",
    "    #recombination\n",
    "    freqsperm_per_genotype = genotype_freqm_homing.*recombination_matrix_vals;\n",
    "\n",
    "    #sex ratio\n",
    "    freqsperm_per_genotype_per_gamete_X = freqsperm_per_genotype.*sex_ratio_matrix_vals[:,1]\n",
    "    freqsperm_per_genotype_per_gamete_Y = freqsperm_per_genotype.*sex_ratio_matrix_vals[:,2]\n",
    "\n",
    "    freqsperm_per_gamete_X = sum(freqsperm_per_genotype_per_gamete_X,dims=1)[:]\n",
    "    freqsperm_per_gamete_Y = sum(freqsperm_per_genotype_per_gamete_Y,dims=1)[:]\n",
    "    freqsperm_per_gamete = vcat(freqsperm_per_gamete_X,freqsperm_per_gamete_Y)\n",
    "\n",
    "    return(hcat(freqsperm_per_gamete_X,freqsperm_per_gamete_Y))\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Takes in a 9x3 matrix containing egg numbers and sperm (X and Y) frequencies and calculates the proportion of \n",
    "each female and male zygote produced of each diploid genotype. \n",
    "Assumes random mating and that males are not limiting. \n",
    "Returns as 78x2 matrix containing numbers of female (column 1) and male (column 2) zygote genotypes. \n",
    "\"\"\"\n",
    "function Find_nzygotes(gametes)\n",
    "    #Taking in matrices of the number of eggs and sperm produced by each genotype:\n",
    "    #Calculates the proportion of each female and male zygote of each genotype produced.\n",
    "    \n",
    "    neggs_genotype = gametes[:,1]\n",
    "    freqsperm_genotype_X = gametes[:,2]\n",
    "    freqsperm_genotype_Y = gametes[:,3]\n",
    "    \n",
    "    #Female zygote genotype \n",
    "    genotype_freq_F_vec = []\n",
    "    #Male zygote genotype \n",
    "    genotype_freq_M_vec = []\n",
    "    \n",
    "    #For each potential zygote genotype\n",
    "    for genotype in genotypes\n",
    "        #Check if homozygote\n",
    "        if genotype[1] == genotype[2]\n",
    "            #Find the index of the gamete present in the genotype\n",
    "            index = findall(x->x==genotype[1],alleles)[1]\n",
    "            #multiply the number of eggs by the frequency X sperm of the specified gamete\n",
    "            genotype_freq_F = neggs_genotype[index]*freqsperm_genotype_X[index]\n",
    "            #multiply the number of eggs by the frequency Y sperm of the specified gamete\n",
    "            genotype_freq_M = neggs_genotype[index]*freqsperm_genotype_Y[index]\n",
    "        else\n",
    "            #Find the index of each allele in the genotype\n",
    "            index1 = findall(x->x==genotype[1],alleles)[1]\n",
    "            index2 = findall(x->x==genotype[2],alleles)[1]\n",
    "            #multiply the number of eggs by the frequency X sperm of the specified gametes in both combinations\n",
    "            genotype_freq_F = ((neggs_genotype[index1]*freqsperm_genotype_X[index2])) + ((freqsperm_genotype_X[index1]*neggs_genotype[index2]))\n",
    "            #multiply the number of eggs by the frequency Y sperm of the specified gametes in both combinations\n",
    "            genotype_freq_M = ((neggs_genotype[index1]*freqsperm_genotype_Y[index2])) + ((freqsperm_genotype_Y[index1]*neggs_genotype[index2]))\n",
    "\n",
    "        end\n",
    "        push!(genotype_freq_F_vec,genotype_freq_F)\n",
    "        push!(genotype_freq_M_vec,genotype_freq_M)\n",
    "    end\n",
    "    \n",
    "    return(hcat(genotype_freq_F_vec,genotype_freq_M_vec))\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Takes in a set of parameters and a 45x2 matrix of numbers per genotype. \n",
    "Applies the density dependent function to the population. \n",
    "Returns a 45x2 matrix of numbers of individuals per genotype after density dependence has been applied. \n",
    "\"\"\"\n",
    "function apply_density_dependence(params,n_zygotes)\n",
    "    n_zygotes_sum = sum(n_zygotes)\n",
    "    juvenile_survival_prob = params[\"OJ\"]*(params[\"α\"]/(params[\"α\"]+n_zygotes_sum))\n",
    "    n_zygote_dd = n_zygotes * juvenile_survival_prob\n",
    "\n",
    "    return(n_zygote_dd)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Simulation of population through time, in discrete non-overlapping generations. \n",
    "The function takes in: parameters, matrixes, fitness effect, a starting population (post release) and \n",
    "the number of generations required.\n",
    "Eggs and sperm are produced, taking into consideration homing, recombination and sex ratio bias. \n",
    "Zygotes are procuded assumin random mating and that males are not limiting. \n",
    "fitness effects are applied according to the fitness effect input(zygote or adult)\n",
    "Density dependence is applied at the zygote stage before zygotes develop into pupae. \n",
    "Zygote mortality (if applied) is applied before density dependence and the zygote population is censused after. \n",
    "The adult population is also censused after fitness effects are applied (if applicable).\n",
    "Zygote, pupae and adult population numbers for each sex are saved in a 45x6xt matrix, \n",
    "and egg numbers and sperm frequencies are saved in a 9x3xt matrix, where t is the number of generations. \n",
    "To enable intuitive data access, the output matrices are processed into a dictionary of dataframes\n",
    "containing tx45 matrices for each diploid population, a tx9 matrix for egg numbes and a tx18 for sperm frequency\n",
    "with columns accessible by labelled genotypes.\n",
    "\"\"\"\n",
    "function Simulate_timeseries(\n",
    "        params, #dictionary of parameters\n",
    "        matrices, #list of matrices\n",
    "        fitness_effect, #string defining when fitness costs are applied (zygote or adult)\n",
    "        t, #number of generations to run the simulation\n",
    "        starting_population_post_release #the starting population (45 x 6 matrix)\n",
    "        )\n",
    "\n",
    "    #Unpack list of pre-defined matrices\n",
    "    fitness_matrix_vals,\n",
    "    homing_matrix_m_vals,\n",
    "    homing_matrix_f_vals,\n",
    "    recombination_matrix_vals,\n",
    "    sex_ratio_matrix_vals = matrices\n",
    "\n",
    "    #define (45x6xt) storage for n. individuals (f zygote, m zygote, f pupae, m pupae, f adult, m adult)\n",
    "    n = Array{Float64}(undef,(length(genotypes),6,t+1))\n",
    "    #define (9x3xt) storage for n. gametes (eggs, X-bearing sperm, Y-bearing sperm))\n",
    "    g = Array{Float64}(undef,(length(alleles),3,t+1))\n",
    "\n",
    "    #save starting population from input\n",
    "    n[:,:,1]=starting_population_post_release \n",
    "\n",
    "    #same no. adults in starting pop\n",
    "    genotype_Nf_adult=n[:,5,1]\n",
    "    genotype_Nm_adult=n[:,6,1]\n",
    "\n",
    "    for gen in 1:t\n",
    "        #female adults produce eggs \n",
    "        neggs_per_gamete =  make_eggs(genotype_Nf_adult,homing_matrix_f_vals,recombination_matrix_vals,params)\n",
    "        g[:,1,gen+1] = neggs_per_gamete\n",
    "        #male adults produce sperm\n",
    "        freqsperm_per_gamete = make_sperm(genotype_Nm_adult,homing_matrix_m_vals,recombination_matrix_vals,sex_ratio_matrix_vals)\n",
    "        g[:,2:3,gen+1] = freqsperm_per_gamete\n",
    "        #zygotes are generated from eggs and sperm\n",
    "        n_zygotes = Find_nzygotes(g[:,:,gen+1])\n",
    "        #fitess costs are applied to zygotes if applicable\n",
    "        fitness_effect == \"zygote\" ? n_zygotes = n_zygotes .*fitness_matrix_vals : n_zygotes = n_zygotes\n",
    "        n[:,1:2,gen+1] = n_zygotes\n",
    "        #density dependent mortality is applied\n",
    "        n_pupae = apply_density_dependence(params,n_zygotes)\n",
    "        n[:,3:4,gen+1] = n_pupae\n",
    "        #fitess costs are applied to zygotes if applicable\n",
    "        fitness_effect == \"adult\" ? n_adults = n_pupae .*fitness_matrix_vals : n_adults = n_pupae\n",
    "        n[:,5:6,gen+1] = n_adults\n",
    "        #Number of adults are updated for generation t+1\n",
    "        genotype_Nf_adult=n_adults[:,1]\n",
    "        genotype_Nm_adult=n_adults[:,2]\n",
    "    end\n",
    "\n",
    "    #Store as dataframes in dictionary for intuitive labelled access \n",
    "    header = Symbol.(genotypes_sym)\n",
    "    output = Dict(\n",
    "        \"genotypenumber_adult_female\" => DataFrame(transpose(n[:,5,:]), header),\n",
    "        \"genotypenumber_adult_male\" => DataFrame(transpose(n[:,6,:]), header),\n",
    "        \"genotypenumber_pupae_female\" => DataFrame(transpose(n[:,3,:]), header),\n",
    "        \"genotypenumber_pupae_male\" => DataFrame(transpose(n[:,4,:]), header),\n",
    "        \"genotypenumber_zygote_female\" => DataFrame(transpose(n[:,1,:]), header),\n",
    "        \"genotypenumber_zygote_male\" => DataFrame(transpose(n[:,2,:]), header),\n",
    "        \"eggnumber\" => DataFrame(transpose(g[:,1,:]), Symbol.(alleles)),\n",
    "        \"spermfreq\" => DataFrame(hcat(transpose(g[:,2,:]),transpose(g[:,3,:])),Symbol.([alleles.*\"X\" ; alleles.*\"Y\"]))\n",
    "    )\n",
    "\n",
    "    return(output)\n",
    "end\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "###### Post simulation processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "code_folding": [
     4,
     85,
     90,
     101,
     121,
     130,
     152,
     161,
     181,
     187
    ]
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Find_correlation (generic function with 1 method)"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Takes in a simulation output and calculates various summary metrics from the numbers of individuals/gametes. \n",
    "Returns a dictionary containing the same simulation output data with additional entries. \n",
    "\"\"\"\n",
    "function Simulation_processing_wrapper(output)\n",
    "        \n",
    "    #Calculate additional values of interest from the time series output\n",
    "    total_Nf = hcat(sum.(eachrow(output[\"genotypenumber_pupae_female\"])),sum.(eachrow(output[\"genotypenumber_adult_female\"])))\n",
    "    total_Nf = DataFrame(total_Nf,[\"pupae\",\"adult\"])\n",
    "    relative_Nf = Find_relative_Nf(total_Nf);\n",
    "\n",
    "    #For all adults (regardless of reproductive fitness)\n",
    "    genotype_freqaf_store = Find_freq_from_no_df(output[\"genotypenumber_adult_female\"])\n",
    "    allele_freqaf = Find_allele_frequencies_from_genotypes_df(genotype_freqaf_store);\n",
    "    genotype_freqam_store = Find_freq_from_no_df(output[\"genotypenumber_adult_male\"])\n",
    "    allele_freqam = Find_allele_frequencies_from_genotypes_df(genotype_freqam_store);\n",
    "    genotype_freqa_store = (genotype_freqaf_store.+genotype_freqam_store)./2;\n",
    "    allele_freqa = Find_allele_frequencies_from_genotypes_df(genotype_freqa_store);\n",
    "   \n",
    "    #For the reproductive adults (regardless of reproductive fitness)\n",
    "    genotype_freqpf_store = Find_freq_from_no_df(output[\"genotypenumber_pupae_female\"])\n",
    "    allele_freqpf = Find_allele_frequencies_from_genotypes_df(genotype_freqpf_store);\n",
    "    genotype_freqpm_store = Find_freq_from_no_df(output[\"genotypenumber_pupae_male\"])\n",
    "    allele_freqpm = Find_allele_frequencies_from_genotypes_df(genotype_freqpm_store);\n",
    "    genotype_freqp_store = (genotype_freqpf_store.+genotype_freqpm_store)./2;\n",
    "    allele_freqp = Find_allele_frequencies_from_genotypes_df(genotype_freqp_store);\n",
    "    \n",
    "    #For the zygotes (regardless of reproductive fitness)\n",
    "    genotype_freqzf_store = Find_freq_from_no_df(output[\"genotypenumber_zygote_female\"])\n",
    "    allele_freqzf = Find_allele_frequencies_from_genotypes_df(genotype_freqzf_store);\n",
    "    genotype_freqzm_store = Find_freq_from_no_df(output[\"genotypenumber_zygote_male\"])\n",
    "    allele_freqzm = Find_allele_frequencies_from_genotypes_df(genotype_freqzm_store);\n",
    "    genotype_freqz_store = (genotype_freqzf_store.+genotype_freqzm_store)./2;\n",
    "    allele_freqz = Find_allele_frequencies_from_genotypes_df(genotype_freqz_store);\n",
    "    \n",
    "    #For the zygotes/all adults (regardless of reproductive fitness)\n",
    "    correlationa = Find_correlation_df(genotypes,genotype_freqa_store)\n",
    "    correlationp = Find_correlation_df(genotypes,genotype_freqp_store)\n",
    "    correlationz = Find_correlation_df(genotypes,genotype_freqz_store)  \n",
    "    \n",
    "    #save in a dictionary\n",
    "    data_dict = Dict(\n",
    "        \n",
    "        \"genotypenumber_adult_female\" => output[\"genotypenumber_adult_female\"],\n",
    "        \"genotypenumber_adult_male\" => output[\"genotypenumber_adult_male\"],\n",
    "        \"genotypenumber_pupae_female\" => output[\"genotypenumber_pupae_female\"],\n",
    "        \"genotypenumber_pupae_male\" => output[\"genotypenumber_pupae_male\"],\n",
    "        \"genotypenumber_zygote_female\" => output[\"genotypenumber_zygote_female\"],\n",
    "        \"genotypenumber_zygote_male\" => output[\"genotypenumber_zygote_male\"],\n",
    "        \n",
    "        \"number_females\" => total_Nf,\n",
    "        \"relative_number_females\" => relative_Nf,\n",
    "        \n",
    "        \"genotypefreq_adult_female\" => genotype_freqaf_store,\n",
    "        \"genotypefreq_adult_male\" => genotype_freqam_store,\n",
    "        \"genotypefreq_adult\" => genotype_freqa_store,\n",
    "        \n",
    "        \"genotypefreq_pupae_female\" => genotype_freqpf_store,\n",
    "        \"genotypefreq_pupae_male\" => genotype_freqpm_store,\n",
    "        \"genotypefreq_pupae\" => genotype_freqp_store,\n",
    "        \n",
    "        \"genotypefreq_zygote_female\" => genotype_freqzf_store,\n",
    "        \"genotypefreq_zygote_male\" => genotype_freqzm_store,\n",
    "        \"genotypefreq_zygote\" => genotype_freqz_store,\n",
    "        \n",
    "        \"allelefreq_adult\" => allele_freqa,\n",
    "        \"allelefreq_pupae\" => allele_freqp,\n",
    "        \"allelefreq_zygote\" => allele_freqz,\n",
    "        \n",
    "        \"allelefreq_adult_female\" => allele_freqaf,\n",
    "        \"allelefreq_pupae_female\" => allele_freqpf,\n",
    "        \"allelefreq_zygote_female\" => allele_freqzf,\n",
    "        \n",
    "        \"allelefreq_adult_male\" => allele_freqam,\n",
    "        \"allelefreq_pupae_male\" => allele_freqpm,\n",
    "        \"allelefreq_zygote_male\" => allele_freqzm,\n",
    "        \n",
    "        \"correlation_adult\" => correlationa,\n",
    "        \"correlation_pupae\" => correlationp,\n",
    "        \"correlation_zygote\" => correlationz\n",
    "    );\n",
    "    return(data_dict)\n",
    "end\n",
    "\n",
    "\"\"\"Calculate the row-wise frequency\"\"\"\n",
    "function Find_freq_from_no_df(input)\n",
    "    input./[sum(row) for row = eachrow(input)]\n",
    "end\n",
    "\n",
    "\"\"\"Calculate the fraction of females relative to generation zero\"\"\"\n",
    "function Find_relative_Nf(Nf_store)\n",
    "    pupae = (Nf_store.pupae)/(Nf_store.pupae[1])\n",
    "    adult = (Nf_store.adult)/(Nf_store.adult[1]);\n",
    "    relative_Nf = DataFrame(pupae = float.(pupae),\n",
    "                    adult=float.(adult))\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Takes in the genotype frequencies of a population at a single time point and \n",
    "calculates the allele frequency.\n",
    "\"\"\"\n",
    "function Find_allele_frequencies_from_genotypes(genotypes,freq_genotype)\n",
    "    allele_freqs = Vector{Float64}()\n",
    "    for allele in ['A','a','α','B','b','β']\n",
    "        allele_freq = 0\n",
    "        for i in 1:length(genotypes)\n",
    "            if (allele in genotypes[i][1]) & (allele in genotypes[i][2])\n",
    "                allele_freq = allele_freq+freq_genotype[i]\n",
    "            elseif (allele in genotypes[i][1]) | (allele in genotypes[i][2])\n",
    "                allele_freq = allele_freq+(freq_genotype[i]/2)\n",
    "            end\n",
    "        end\n",
    "       push!(allele_freqs,allele_freq)\n",
    "    end\n",
    "    return(allele_freqs)\n",
    "end\n",
    "\"\"\"\n",
    "Takes in the genotype frequencies of a population through time (tx45 matrix) and \n",
    "calculates the allele frequency at each timepoint\n",
    "Returns a tx6 dataframe with columns labelled by allele\n",
    "\"\"\"\n",
    "function Find_allele_frequencies_from_genotypes_df(genotype_freq_store)\n",
    "    header = Symbol.(['A','a','α','B','b','β'])\n",
    "    Make_dataframe([Find_allele_frequencies_from_genotypes(genotypes,row) for row = eachrow(genotype_freq_store)],header)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Takes in the genotype frequencies of a population at a single time point and \n",
    "calculated the frequency of each haploid genotype (chromosome).\n",
    "\"\"\"\n",
    "function Find_chromosome_frequencies_from_genotypes(genotypes,freq_genotype)\n",
    "    allele_freqs = []\n",
    "    for allele in alleles\n",
    "        allele_freq = 0\n",
    "        for i in 1:length(genotypes)\n",
    "            if (allele in genotypes[i])\n",
    "                if genotypes[i][1] == genotypes[i][2]\n",
    "                    allele_freq = allele_freq+(freq_genotype[i])\n",
    "                else\n",
    "                    allele_freq = allele_freq+(freq_genotype[i]/2)\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "       append!(allele_freqs,allele_freq)\n",
    "    end\n",
    "    return(allele_freqs)\n",
    "end\n",
    "\"\"\"\n",
    "Takes in the genotype frequencies of a population  through time (tx45 matrix) and \n",
    "calculated the frequency of each haploid genotype (chromosome) at each timepoint.\n",
    "Returns a tx9 dataframe with columns labelled by haploid genotype.\n",
    "\"\"\"\n",
    "function Find_chromosome_frequencies_from_genotypes_df(genotype_freq_store)\n",
    "    header = Symbol.(alleles)\n",
    "    Make_dataframe([Find_chromosome_frequencies_from_genotypes(genotypes,row) for row = eachrow(genotype_freq_store)],header)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Takes in the genotype frequencies of a population at a single time point and \n",
    "calculated the correlation bewteen the constructs a and b.\n",
    "\"\"\"\n",
    "function Find_correlation(genotypes,genotype_freq)\n",
    "    \n",
    "    chrom_freq = Find_chromosome_frequencies_from_genotypes(genotypes,genotype_freq)\n",
    "    ab_freq = chrom_freq[5]\n",
    "    allele_freq = Find_allele_frequencies_from_genotypes(genotypes,genotype_freq)\n",
    "    \n",
    "    a_freq = round(allele_freq[2],digits = 10)\n",
    "    b_freq = round(allele_freq[5],digits = 10)\n",
    "\n",
    "    if ((float(b_freq) > 0.0))\n",
    "        cor = (ab_freq - (a_freq*b_freq))/sqrt(float(a_freq*(1-a_freq)*b_freq*(1-b_freq)))\n",
    "    else\n",
    "        cor = 0.0\n",
    "    end\n",
    "end\n",
    "\"\"\"\n",
    "Takes in the genotype frequencies of a population  through time (tx45 matrix) and \n",
    "calculated the correlation bewteen the constructs a and b at each timepoint.\n",
    "Returns a tx1 dataframe\n",
    "\"\"\"\n",
    "function Find_correlation_df(genotypes,genotype_freq_store)\n",
    "    header = [\"Correlation\"]\n",
    "    Make_dataframe([Find_correlation(genotypes,row) for row = eachrow(genotype_freq_store)],header)\n",
    "end\n",
    "\n",
    "\"\"\"Processes a matrix into a dataframe\"\"\"\n",
    "function Make_dataframe(store, header)\n",
    "    df = DataFrame(transpose(hcat(store...)))\n",
    "    rename!(df, header)\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "###### Input/Output functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "code_folding": [
     3,
     27
    ]
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MDFSXshredder_IO_save (generic function with 1 method)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Writes output dictionary to file, with each entry of the dictionary being written to a different txt file.\n",
    "\"\"\"\n",
    "function Sim_IO_save(file_prefix,\n",
    "                            params,\n",
    "                            data_dict,\n",
    "                            t)\n",
    "    \n",
    "    #Write to info file the parameters used to run the sumulation\n",
    "    fn = file_prefix*\"_info.txt\"\n",
    "    io = open(fn, \"w\");\n",
    "    write(io,\"sim_length \\t\"*string(t)*\"\\n\")\n",
    "    p_label = keys(params)\n",
    "    for p in p_label\n",
    "        write(io,p*\"\\t\"*string(params[p])*\"\\n\")\n",
    "    end\n",
    "    close(io)\n",
    "    \n",
    "    #Write to file the data according to the dictionary label\n",
    "    for i in [keys(data_dict)...]\n",
    "        CSV.write(file_prefix*\"_\"*i*\".csv\", data_dict[i])\n",
    "    end\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Reads output files and stores in dictionary.\n",
    "\"\"\"\n",
    "function Sim_IO_read(file_prefix,\n",
    "                               data_labels)\n",
    "    data_dict = Dict()\n",
    "    for label in data_labels\n",
    "        data_dict[label] = CSV.read(file_prefix*\"_\"*label*\".csv\",DataFrame)\n",
    "    end\n",
    "    return(data_dict)\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [
     3
    ]
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Unpacks the string construct label and assigns locus types, homing dependencies and X-shredder dependency\n",
    "\"\"\"\n",
    "function unpack_construct_label(construct_label)\n",
    "    label_vec = split(construct_label,\"_\")\n",
    "    index=2\n",
    "    LOCUS_A = split(label_vec[index],\"-\")[2]\n",
    "    if LOCUS_A == \"HS\"\n",
    "        LOCUS_A = \"HS\"*\"_\"*label_vec[3]\n",
    "        index = index+1\n",
    "    end\n",
    "    LOCUS_B = split(label_vec[index+1],\"-\")[2]\n",
    "    if LOCUS_B == \"HS\"\n",
    "        LOCUS_B = \"HS\"*\"_\"*label_vec[4]\n",
    "        index = index+1\n",
    "    end\n",
    "    set_A_homing_dependency = split(label_vec[index+2],\"-\")[2]\n",
    "    set_B_homing_dependency = split(label_vec[index+3],\"-\")[2]\n",
    "    set_XS_dependency = split(label_vec[index+4],\"-\")[2]\n",
    "    if set_XS_dependency == \"constitutive\"\n",
    "        set_XS_dependency = set_XS_dependency*\"_\"*label_vec[index+5]\n",
    "    end\n",
    "\n",
    "    return(LOCUS_A,\n",
    "            LOCUS_B,\n",
    "            string(set_A_homing_dependency),\n",
    "            string(set_B_homing_dependency),\n",
    "            string(set_XS_dependency))\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "###### Parameter screen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "code_folding": [
     3,
     23,
     136
    ]
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Parameter_screen_parallel"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Calculates the number of simulations in a 2-dimentional parameter search\n",
    "\"\"\"\n",
    "function Number_sims(B_pars,var_pars)\n",
    "    counter = 0\n",
    "    for i in B_pars\n",
    "        for j in var_pars\n",
    "             counter += 1\n",
    "        end\n",
    "    end\n",
    "    return(counter)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Defines starting population and matrices, runs a single simulation and calculates additional metrics from output\n",
    "for a specific combination of parameters and starting frequency of functional cleavage resistance at the A locus. \n",
    "pcomn is a list of two entries, the first being the resistant frequency of the resistant genotypes,\n",
    "and the second being a list of parameters for the corresponding list of parameter labels (var_symbol_vec).\n",
    "For example, if running a simulation with the o resistance allele at 50%, r=0.2 and m =0.8: \n",
    "    pcomb = [0.5,[0.2,0.8]]\n",
    "    resistant allele = \"α\"\n",
    "    var_symbol_vec = [\"r\",\"m\"]\n",
    "\"\"\"\n",
    "function Parameter_screen_single(pcomb, \n",
    "                                params,\n",
    "                                output_folder,\n",
    "                                var_symbol_vec,\n",
    "                                A_construct,\n",
    "                                B_construct,\n",
    "                                A_homing_dependency,\n",
    "                                B_homing_dependency,\n",
    "                                XS_dependency,\n",
    "                                fitness_effect,\n",
    "                                transgenic_genotype,\n",
    "                                release_freq, \n",
    "                                resistant_allele,\n",
    "                                t)\n",
    "\n",
    "        resistant_freq,var_vec = pcomb\n",
    "\n",
    "        #MAKE STARTING resistant_freq PARAMETER FOLDER\n",
    "        output_standingres_folder = output_folder*\"/Sim_t$(t)_variable_standingres$(resistant_freq)\"\n",
    "        if isdir(output_standingres_folder) == false\n",
    "            mkdir(output_standingres_folder)\n",
    "        end\n",
    "    \n",
    "        #MAKE STARTING variable parameter FOLDER\n",
    "        var_label = \"\"\n",
    "        for i in collect(1:length(var_symbol_vec))\n",
    "            var_label_i = var_symbol_vec[i]\n",
    "            var_label_i = \"_$(var_label_i)$(var_vec[i])\"\n",
    "            var_label = var_label*var_label_i\n",
    "        end\n",
    "        output_loc = output_standingres_folder*\"/Sim_t$(t)_variable_standingres$(resistant_freq)$(var_label)\"\n",
    "            if isdir(output_loc) == false\n",
    "            mkdir(output_loc)\n",
    "        end\n",
    "\n",
    "        #SET VARIABLE PARAMETER\n",
    "        for i in collect(1:length(var_symbol_vec))\n",
    "            params[var_symbol_vec[i]] = var_vec[i]\n",
    "        end\n",
    "    \n",
    "        #Define starting population\n",
    "        starting_population_post_release = release_transgenic_males(\n",
    "            Initiate_prerelease_population_3lifestages(resistant_allele, resistant_freq, params),\n",
    "            Symbol(transgenic_genotype),\n",
    "            release_freq)\n",
    "\n",
    "        #Define matrices\n",
    "        matrices = make_matrices(params,\n",
    "                        A_construct,\n",
    "                        B_construct,\n",
    "                        A_homing_dependency,\n",
    "                        B_homing_dependency,\n",
    "                        XS_dependency);\n",
    "\n",
    "        #simulate time series\n",
    "        output = Simulate_timeseries(\n",
    "                    params,\n",
    "                    matrices,\n",
    "                    fitness_effect,\n",
    "                    t,\n",
    "                    starting_population_post_release)\n",
    "    \n",
    "        unique_file_prefix = output_loc*\"/Sim_t$(t)_variable_standingres$(resistant_freq)$(var_label)\"\n",
    "\n",
    "        total_Nf = hcat(sum.(eachrow(output[\"genotypenumber_pupae_female\"])),sum.(eachrow(output[\"genotypenumber_adult_female\"])))\n",
    "        total_Nf = DataFrame(total_Nf,[\"pupae\",\"adult\"])\n",
    "        relative_Nf = Find_relative_Nf(total_Nf);\n",
    "    \n",
    "        #For larvae\n",
    "        genotype_freqzf_store = Find_freq_from_no_df(output[\"genotypenumber_zygote_female\"])\n",
    "        genotype_freqzm_store = Find_freq_from_no_df(output[\"genotypenumber_zygote_male\"])\n",
    "        genotype_freqz_store = (genotype_freqzf_store.+genotype_freqzm_store)./2;\n",
    "        allele_freqz = Find_allele_frequencies_from_genotypes_df(genotype_freqz_store);\n",
    "    \n",
    "        #For pupae \n",
    "        genotype_freqaf_store = Find_freq_from_no_df(output[\"genotypenumber_pupae_female\"])\n",
    "        genotype_freqam_store = Find_freq_from_no_df(output[\"genotypenumber_pupae_male\"])\n",
    "        genotype_freqa_store = (genotype_freqaf_store.+genotype_freqam_store)./2;\n",
    "        allele_freqa = Find_allele_frequencies_from_genotypes_df(genotype_freqa_store);\n",
    "\n",
    "        #For adults\n",
    "        genotype_freqrf_store = Find_freq_from_no_df(output[\"genotypenumber_adult_female\"])\n",
    "        genotype_freqrm_store = Find_freq_from_no_df(output[\"genotypenumber_adult_male\"])\n",
    "        genotype_freqr_store = (genotype_freqrf_store.+genotype_freqrm_store)./2;\n",
    "        allele_freqr = Find_allele_frequencies_from_genotypes_df(genotype_freqr_store);\n",
    "\n",
    "        data_dict = Dict(\n",
    "            \"number_females\" => total_Nf,\n",
    "            \"relative_number_females\" => relative_Nf,\n",
    "            \"allelefreq_pupae\" => allele_freqa,\n",
    "            \"allelefreq_adult\" => allele_freqr,\n",
    "            \"allelefreq_zygote\" => allele_freqz,\n",
    "        );\n",
    "     #save\n",
    "      Sim_IO_save(unique_file_prefix,\n",
    "                        params,\n",
    "                        data_dict,\n",
    "                        t)\n",
    "\n",
    "      \n",
    "    println(\"Completed sim res$(resistant_freq)$(var_label)\")\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Takes in a range of resistant frequencies (res_pars) and paramaters (var_pars) and runs many simulations in parallel. \n",
    "For example entering \n",
    "    res_pars = [0.5,0.6]\n",
    "    var_pars = [(0.2,0.8),(0.3,0.9)]\n",
    "    var_symbol = [\"r\",\"m\"]\n",
    "Would run 2 simulations\n",
    "    1: res freq = 0.5, r = 0.2, m = 0.8 \n",
    "    2: res freq = 0.6, r = 0.3, m = 0.9 \n",
    "\"\"\"\n",
    "function Parameter_screen_parallel(output_folder,\n",
    "                                        params,\n",
    "                                        res_pars,\n",
    "                                        var_symbol,\n",
    "                                        var_pars,\n",
    "                                        A_construct,\n",
    "                                        B_construct,\n",
    "                                        A_homing_dependency,\n",
    "                                        B_homing_dependency,\n",
    "                                        XS_dependency,\n",
    "                                        fitness_effect,\n",
    "                                        transgenic_genotype,\n",
    "                                        release_freq, \n",
    "                                        resistant_genotype, \n",
    "                                        t)\n",
    "    \n",
    "    param_screen = [(start_res_freq,var) for start_res_freq in res_pars, var in var_pars]\n",
    "    pmap(x -> Parameter_screen_single(x,\n",
    "                                    params,\n",
    "                                    output_folder,\n",
    "                                    var_symbol,\n",
    "                                    A_construct,\n",
    "                                    B_construct,\n",
    "                                    A_homing_dependency,\n",
    "                                    B_homing_dependency,\n",
    "                                    XS_dependency,\n",
    "                                    fitness_effect,\n",
    "                                    transgenic_genotype,\n",
    "                                    release_freq, \n",
    "                                    resistant_genotype, \n",
    "                                    t), \n",
    "                                    param_screen);\n",
    "end\n"
   ]
  }
 ],
 "metadata": {
  "@webio": {
   "lastCommId": null,
   "lastKernelId": null
  },
  "kernelspec": {
   "display_name": "Julia 1.5.2",
   "language": "julia",
   "name": "julia-1.5"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.5.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
